{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98a0d93c-1359-4107-807f-f230b3863708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89dac285-5db9-4261-b4e4-38e0f84db137",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_read = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens_GRCh38_112.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "#tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f9135a3-08a2-46ed-9c66-2622790d4290",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519861</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519862</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519863</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519864</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519865</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3519866 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3519861   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3519862   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3519863   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3519864   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3519865   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                               Tags  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3519861      -       .                                 logic_name=eponine  \n",
       "3519862      +       .                                 logic_name=eponine  \n",
       "3519863      -       .                                 logic_name=eponine  \n",
       "3519864      +       .                                 logic_name=eponine  \n",
       "3519865      -       .                                 logic_name=eponine  \n",
       "\n",
       "[3519866 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Signal', 7: 'Unknown', 8: 'Tags'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64551187-3a83-4376-a7db-f0cc87b190b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'scaffold', 'tRNA'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2f75010d-b1b7-411c-b46c-2591004592f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>940346.0</td>\n",
       "      <td>942173.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000478729;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>945319.0</td>\n",
       "      <td>945562.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000496938;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1593</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>954426.0</td>\n",
       "      <td>959309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000487214;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2582</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1082146.0</td>\n",
       "      <td>1084072.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000464905;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2652</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1083926.0</td>\n",
       "      <td>1116243.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000482816;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519720</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57192634.0</td>\n",
       "      <td>57194566.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711289;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519739</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57207346.0</td>\n",
       "      <td>57208935.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711271;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519774</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57208197.0</td>\n",
       "      <td>57209855.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711275;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519809</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57208519.0</td>\n",
       "      <td>57209909.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711280;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519831</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210591.0</td>\n",
       "      <td>57212074.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711285;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28741 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index  Last_index Value Signal Unknown  \\\n",
       "1432      1  havana  transcript     940346.0    942173.0     .      +       .   \n",
       "1590      1  havana  transcript     945319.0    945562.0     .      -       .   \n",
       "1593      1  havana  transcript     954426.0    959309.0     .      -       .   \n",
       "2582      1  havana  transcript    1082146.0   1084072.0     .      -       .   \n",
       "2652      1  havana  transcript    1083926.0   1116243.0     .      -       .   \n",
       "...      ..     ...         ...          ...         ...   ...    ...     ...   \n",
       "3519720   Y  havana  transcript   57192634.0  57194566.0     .      +       .   \n",
       "3519739   Y  havana  transcript   57207346.0  57208935.0     .      +       .   \n",
       "3519774   Y  havana  transcript   57208197.0  57209855.0     .      +       .   \n",
       "3519809   Y  havana  transcript   57208519.0  57209909.0     .      +       .   \n",
       "3519831   Y  havana  transcript   57210591.0  57212074.0     .      +       .   \n",
       "\n",
       "                                                      Tags  \n",
       "1432     ID=transcript:ENST00000478729;Parent=gene:ENSG...  \n",
       "1590     ID=transcript:ENST00000496938;Parent=gene:ENSG...  \n",
       "1593     ID=transcript:ENST00000487214;Parent=gene:ENSG...  \n",
       "2582     ID=transcript:ENST00000464905;Parent=gene:ENSG...  \n",
       "2652     ID=transcript:ENST00000482816;Parent=gene:ENSG...  \n",
       "...                                                    ...  \n",
       "3519720  ID=transcript:ENST00000711289;Parent=gene:ENSG...  \n",
       "3519739  ID=transcript:ENST00000711271;Parent=gene:ENSG...  \n",
       "3519774  ID=transcript:ENST00000711275;Parent=gene:ENSG...  \n",
       "3519809  ID=transcript:ENST00000711280;Parent=gene:ENSG...  \n",
       "3519831  ID=transcript:ENST00000711285;Parent=gene:ENSG...  \n",
       "\n",
       "[28741 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[tanscripts_df[\"Type\"] == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c35e5d62-69a8-44a0-be89-734bae0489bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['logic_name=eponine']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][3].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c053bad3-d4c7-479f-9b14-c5be43627c1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000478729',\n",
       " 'Parent=gene:ENSG00000187634',\n",
       " 'Name=SAMD11-208',\n",
       " 'biotype=protein_coding_CDS_not_defined',\n",
       " 'transcript_id=ENST00000478729',\n",
       " 'transcript_support_level=5',\n",
       " 'version=1']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][1432].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a52f5986-a36b-41f1-8068-4564c9ccd52a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000641515',\n",
       " 'Parent=gene:ENSG00000186092',\n",
       " 'Name=OR4F5-201',\n",
       " 'biotype=protein_coding',\n",
       " 'ccdsid=CCDS30547.2',\n",
       " 'tag=basic,Ensembl_canonical,MANE_Select,GENCODE Primary',\n",
       " 'transcript_id=ENST00000641515',\n",
       " 'version=2']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][94].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "43e4150a-8eb9-4064-a96e-ebc1d40728f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>65433.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>65520.0</td>\n",
       "      <td>65564.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70009.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>924431.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519661</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57128472.0</td>\n",
       "      <td>57128509.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519675</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57184216.0</td>\n",
       "      <td>57184279.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519694</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57196930.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519696</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>57184226.0</td>\n",
       "      <td>57184263.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519715</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57196373.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387123 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome             Type  First_index  Last_index Value  \\\n",
       "96        1          havana   five_prime_UTR      65419.0     65433.0     .   \n",
       "97        1          havana   five_prime_UTR      65520.0     65564.0     .   \n",
       "102       1          havana  three_prime_UTR      70009.0     71585.0     .   \n",
       "1163      1          havana   five_prime_UTR     923923.0    924431.0     .   \n",
       "1192      1          havana  three_prime_UTR     944154.0    944574.0     .   \n",
       "...      ..             ...              ...          ...         ...   ...   \n",
       "3519661   Y  ensembl_havana  three_prime_UTR   57128472.0  57128509.0     .   \n",
       "3519675   Y  ensembl_havana   five_prime_UTR   57184216.0  57184279.0     .   \n",
       "3519694   Y  ensembl_havana  three_prime_UTR   57196930.0  57197337.0     .   \n",
       "3519696   Y  ensembl_havana   five_prime_UTR   57184226.0  57184263.0     .   \n",
       "3519715   Y  ensembl_havana  three_prime_UTR   57196373.0  57197337.0     .   \n",
       "\n",
       "        Signal Unknown                               Tags  \n",
       "96           +       .  Parent=transcript:ENST00000641515  \n",
       "97           +       .  Parent=transcript:ENST00000641515  \n",
       "102          +       .  Parent=transcript:ENST00000641515  \n",
       "1163         +       .  Parent=transcript:ENST00000616016  \n",
       "1192         +       .  Parent=transcript:ENST00000616016  \n",
       "...        ...     ...                                ...  \n",
       "3519661      +       .  Parent=transcript:ENST00000711265  \n",
       "3519675      +       .  Parent=transcript:ENST00000711286  \n",
       "3519694      +       .  Parent=transcript:ENST00000711286  \n",
       "3519696      +       .  Parent=transcript:ENST00000711287  \n",
       "3519715      +       .  Parent=transcript:ENST00000711287  \n",
       "\n",
       "[387123 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[(tanscripts_df[\"Tags\"].str.contains(\"Parent\", na=False)) & (~tanscripts_df[\"Tags\"].str.contains(\";\", na=False))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b7024b10-b3ea-4db4-93e6-99a0a633c7f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519861</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519862</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519863</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519864</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519865</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>309485 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3519861   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3519862   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3519863   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3519864   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3519865   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                               Tags  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3519861      -       .                                 logic_name=eponine  \n",
       "3519862      +       .                                 logic_name=eponine  \n",
       "3519863      -       .                                 logic_name=eponine  \n",
       "3519864      +       .                                 logic_name=eponine  \n",
       "3519865      -       .                                 logic_name=eponine  \n",
       "\n",
       "[309485 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[~tanscripts_df[\"Tags\"].str.contains(\"Parent\", na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9d0127f-5322-494b-b663-f8612dfd4cb8",
   "metadata": {},
   "source": [
    "## Dividir a coluna em várias colunas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0195a278-4dcd-4a96-ba02-a1e72d548ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def divide_TagsColumn():\n",
    "    DesirableTags = [\"ID\", \"Parent\", \"biotype\"]\n",
    "    IDs = []\n",
    "    Parents = []\n",
    "    biotypes = []\n",
    "    Other_tags = []\n",
    "    for transcript_tags in tanscripts_df[\"Tags\"]:    #for each row in Tags column\n",
    "        otherTags = []\n",
    "        if \"ID=\" not in str(transcript_tags):\n",
    "            IDs.append(np.nan)\n",
    "        if \"Parent=\" not in str(transcript_tags):\n",
    "            Parents.append(np.nan)\n",
    "        if \"biotype=\" not in str(transcript_tags):\n",
    "            biotypes.append(np.nan)\n",
    "        if \";\" in str(transcript_tags):\n",
    "            row_tags = transcript_tags.split(\";\")\n",
    "            for tag in row_tags:\n",
    "                Tag_type = tag.split(\"=\")[0]\n",
    "                if Tag_type == \"ID\":\n",
    "                    IDs.append(tag)\n",
    "                elif Tag_type == \"Parent\":\n",
    "                    Parents.append(tag)\n",
    "                elif Tag_type == \"biotype\":\n",
    "                    biotypes.append(tag)\n",
    "                else:\n",
    "                    otherTags.append(tag)\n",
    "        else:\n",
    "            if \"ID=\" in str(transcript_tags):\n",
    "                IDs.append(np.nan)\n",
    "            elif \"Parent=\" in str(transcript_tags):\n",
    "                Parents.append(np.nan)\n",
    "            elif \"biotype=\" in str(transcript_tags):\n",
    "                biotypes.append(np.nan)\n",
    "            else:\n",
    "                otherTags.append(str(transcript_tags))\n",
    "        Other_tags.append(str(otherTags))\n",
    "    print(len(IDs))\n",
    "    print(len(Parents))\n",
    "    print(\"Parent\" in Other_tags)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"ID\", IDs, True)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Parent\", Parents, True)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Biotype\", biotypes, True)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Other_tags\", str(Other_tags), True)\n",
    "    tanscripts_df.drop(['Tags'], axis=1)\n",
    "    return tanscripts_df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4238bd06-a8e3-437a-a4a8-2ec6f5f65f07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3519866\n",
      "3519866\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "tanscripts_df = divide_TagsColumn()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "28fdbc2b-be77-4dc9-b46c-f88f99fe3efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df = tanscripts_df.drop(\"Tags\", axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ddd1e0fb-c297-4cfa-8ef5-43f26e231e5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Other_tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "      <td>ID=chromosome:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519861</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519862</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519863</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519864</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519865</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3519866 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3519861   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3519862   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3519863   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3519864   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3519865   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                               Tags  \\\n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...   \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg   \n",
       "2            +       .                                 logic_name=eponine   \n",
       "3            -       .                                 logic_name=eponine   \n",
       "4            +       .                                 logic_name=eponine   \n",
       "...        ...     ...                                                ...   \n",
       "3519861      -       .                                 logic_name=eponine   \n",
       "3519862      +       .                                 logic_name=eponine   \n",
       "3519863      -       .                                 logic_name=eponine   \n",
       "3519864      +       .                                 logic_name=eponine   \n",
       "3519865      -       .                                 logic_name=eponine   \n",
       "\n",
       "                      ID Parent Biotype  \\\n",
       "0        ID=chromosome:1    NaN     NaN   \n",
       "1                    NaN    NaN     NaN   \n",
       "2                    NaN    NaN     NaN   \n",
       "3                    NaN    NaN     NaN   \n",
       "4                    NaN    NaN     NaN   \n",
       "...                  ...    ...     ...   \n",
       "3519861              NaN    NaN     NaN   \n",
       "3519862              NaN    NaN     NaN   \n",
       "3519863              NaN    NaN     NaN   \n",
       "3519864              NaN    NaN     NaN   \n",
       "3519865              NaN    NaN     NaN   \n",
       "\n",
       "                                                Other_tags  \n",
       "0        [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "1        [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "2        [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3        [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "4        [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "...                                                    ...  \n",
       "3519861  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519862  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519863  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519864  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519865  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "\n",
       "[3519866 rows x 13 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd5b333-668d-4f62-94f9-a253756f0d38",
   "metadata": {},
   "source": [
    "## As linhas com a tag \"MANE_Plus_Clitical\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e574e850-54f0-40d7-aa29-2d80a8ee462e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mask1 = (\"MANE_Plus_Critical\" in tanscripts_df['Tags']) == 1)# & (gvs_chr20_df['ALT'].str.len() == 1)\n",
    "#tanscripts_df_ManeSelect = tanscripts_df[~tanscripts_df['Tags'].str.contains(\"MANE_Plus_Critical\")]\n",
    "MPC_rows = tanscripts_df['Tags'].str.contains(\"MANE_Plus_Clinical\", na=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03477a94-ef8a-4684-be34-7a0ca6b91304",
   "metadata": {},
   "source": [
    "#### 64 linhas com a tag \"MANE_Plus_Clinical\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e3b4d9fd-de34-4acc-8db0-90031e0da942",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>98689</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>45329242.0</td>\n",
       "      <td>45340440.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000710952;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218853</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>156114711.0</td>\n",
       "      <td>156137866.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000677389;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277374</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>204870862.0</td>\n",
       "      <td>205022822.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000539706;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345935</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>18340709.0</td>\n",
       "      <td>18543557.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000377329;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371005</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>49515198.0</td>\n",
       "      <td>49539121.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000447839;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3345187</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>127612277.0</td>\n",
       "      <td>127692699.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000373302;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3437601</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>53266328.0</td>\n",
       "      <td>53281584.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000639161;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3479523</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>123184278.0</td>\n",
       "      <td>123490915.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000622768;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3488707</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>136147533.0</td>\n",
       "      <td>136211354.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000394155;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3502545</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>154021573.0</td>\n",
       "      <td>154097717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000303391;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index   Last_index Value Signal  \\\n",
       "98689     1          havana  mRNA   45329242.0   45340440.0     .      -   \n",
       "218853    1          havana  mRNA  156114711.0  156137866.0     .      +   \n",
       "277374    1          havana  mRNA  204870862.0  205022822.0     .      +   \n",
       "345935   10  ensembl_havana  mRNA   18340709.0   18543557.0     .      +   \n",
       "371005   10  ensembl_havana  mRNA   49515198.0   49539121.0     .      -   \n",
       "...      ..             ...   ...          ...          ...   ...    ...   \n",
       "3345187   9  ensembl_havana  mRNA  127612277.0  127692699.0     .      +   \n",
       "3437601   X  ensembl_havana  mRNA   53266328.0   53281584.0     .      -   \n",
       "3479523   X  ensembl_havana  mRNA  123184278.0  123490915.0     .      +   \n",
       "3488707   X  ensembl_havana  mRNA  136147533.0  136211354.0     .      +   \n",
       "3502545   X  ensembl_havana  mRNA  154021573.0  154097717.0     .      -   \n",
       "\n",
       "        Unknown                                        Identifiers  \n",
       "98689         .  ID=transcript:ENST00000710952;Parent=gene:ENSG...  \n",
       "218853        .  ID=transcript:ENST00000677389;Parent=gene:ENSG...  \n",
       "277374        .  ID=transcript:ENST00000539706;Parent=gene:ENSG...  \n",
       "345935        .  ID=transcript:ENST00000377329;Parent=gene:ENSG...  \n",
       "371005        .  ID=transcript:ENST00000447839;Parent=gene:ENSG...  \n",
       "...         ...                                                ...  \n",
       "3345187       .  ID=transcript:ENST00000373302;Parent=gene:ENSG...  \n",
       "3437601       .  ID=transcript:ENST00000639161;Parent=gene:ENSG...  \n",
       "3479523       .  ID=transcript:ENST00000622768;Parent=gene:ENSG...  \n",
       "3488707       .  ID=transcript:ENST00000394155;Parent=gene:ENSG...  \n",
       "3502545       .  ID=transcript:ENST00000303391;Parent=gene:ENSG...  \n",
       "\n",
       "[64 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_Mane_Plus_Clinical = tanscripts_df[MPC_rows]\n",
    "tanscripts_df_Mane_Plus_Clinical"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02ea7eb7-a64c-498e-91c5-f646e61ed890",
   "metadata": {},
   "source": [
    "Este metodo de filtragem retira transcritos (anotação \"mRNA\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "df0f7e21-c1dd-4ef8-8c67-66a178a8d292",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['mRNA'], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_Mane_Plus_Clinical[\"Type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "14de4cb3-f370-450b-82c7-0f1e9c189234",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000710952',\n",
       " 'Parent=gene:ENSG00000132781',\n",
       " 'Name=MUTYH-248',\n",
       " 'biotype=protein_coding',\n",
       " 'tag=basic,MANE_Plus_Clinical,GENCODE Primary',\n",
       " 'transcript_id=ENST00000710952',\n",
       " 'version=2']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_Mane_Plus_Clinical[\"Tags\"][98689].split(\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58f59eb9-8f24-4f5e-8946-0cf48e45e25f",
   "metadata": {},
   "source": [
    "## As linhas com anotaçao \"MANE_Select\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8885c4c9-6885-4df0-b83a-89817e3c6468",
   "metadata": {},
   "outputs": [],
   "source": [
    "MS_rows = tanscripts_df['Tags'].str.contains(\"MANE_Select\", na=False)\n",
    "tanscripts_df_ManeSelect = tanscripts_df[MS_rows]\n",
    "#tanscripts_df_noMPC = tanscripts_df[~MPC_rows]   ###usado anteriormente para descartar as linhas com \"MANE_Plus_Clinical\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d96a3374-c7d3-48c1-9e58-ce93ad2366fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Other_tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000641515;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000641515</td>\n",
       "      <td>Parent=gene:ENSG00000186092</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000426406;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000426406</td>\n",
       "      <td>Parent=gene:ENSG00000284733</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000332831;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000332831</td>\n",
       "      <td>Parent=gene:ENSG00000284662</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000616016;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000616016</td>\n",
       "      <td>Parent=gene:ENSG00000187634</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959256.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000327044;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000327044</td>\n",
       "      <td>Parent=gene:ENSG00000188976</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518468</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24618004.0</td>\n",
       "      <td>24639207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382392;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382392</td>\n",
       "      <td>Parent=gene:ENSG00000183795</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518527</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24763069.0</td>\n",
       "      <td>24813393.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382365;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382365</td>\n",
       "      <td>Parent=gene:ENSG00000187191</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518777</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24833919.0</td>\n",
       "      <td>24907040.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000682740;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000682740</td>\n",
       "      <td>Parent=gene:ENSG00000205916</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519057</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25052104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382287;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382287</td>\n",
       "      <td>Parent=gene:ENSG00000185894</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519268</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000306609;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000306609</td>\n",
       "      <td>Parent=gene:ENSG00000172288</td>\n",
       "      <td>biotype=protein_coding</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19226 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "94        1          havana  mRNA      65419.0     71585.0     .      +   \n",
       "396       1  ensembl_havana  mRNA     450740.0    451678.0     .      -   \n",
       "542       1  ensembl_havana  mRNA     685716.0    686654.0     .      -   \n",
       "1162      1          havana  mRNA     923923.0    944574.0     .      +   \n",
       "1525      1  ensembl_havana  mRNA     944203.0    959256.0     .      -   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3518468   Y          havana  mRNA   24618004.0  24639207.0     .      +   \n",
       "3518527   Y  ensembl_havana  mRNA   24763069.0  24813393.0     .      -   \n",
       "3518777   Y          havana  mRNA   24833919.0  24907040.0     .      +   \n",
       "3519057   Y          havana  mRNA   25030901.0  25052104.0     .      -   \n",
       "3519268   Y  ensembl_havana  mRNA   25622117.0  25624902.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \\\n",
       "94            .  ID=transcript:ENST00000641515;Parent=gene:ENSG...   \n",
       "396           .  ID=transcript:ENST00000426406;Parent=gene:ENSG...   \n",
       "542           .  ID=transcript:ENST00000332831;Parent=gene:ENSG...   \n",
       "1162          .  ID=transcript:ENST00000616016;Parent=gene:ENSG...   \n",
       "1525          .  ID=transcript:ENST00000327044;Parent=gene:ENSG...   \n",
       "...         ...                                                ...   \n",
       "3518468       .  ID=transcript:ENST00000382392;Parent=gene:ENSG...   \n",
       "3518527       .  ID=transcript:ENST00000382365;Parent=gene:ENSG...   \n",
       "3518777       .  ID=transcript:ENST00000682740;Parent=gene:ENSG...   \n",
       "3519057       .  ID=transcript:ENST00000382287;Parent=gene:ENSG...   \n",
       "3519268       .  ID=transcript:ENST00000306609;Parent=gene:ENSG...   \n",
       "\n",
       "                                    ID                       Parent  \\\n",
       "94       ID=transcript:ENST00000641515  Parent=gene:ENSG00000186092   \n",
       "396      ID=transcript:ENST00000426406  Parent=gene:ENSG00000284733   \n",
       "542      ID=transcript:ENST00000332831  Parent=gene:ENSG00000284662   \n",
       "1162     ID=transcript:ENST00000616016  Parent=gene:ENSG00000187634   \n",
       "1525     ID=transcript:ENST00000327044  Parent=gene:ENSG00000188976   \n",
       "...                                ...                          ...   \n",
       "3518468  ID=transcript:ENST00000382392  Parent=gene:ENSG00000183795   \n",
       "3518527  ID=transcript:ENST00000382365  Parent=gene:ENSG00000187191   \n",
       "3518777  ID=transcript:ENST00000682740  Parent=gene:ENSG00000205916   \n",
       "3519057  ID=transcript:ENST00000382287  Parent=gene:ENSG00000185894   \n",
       "3519268  ID=transcript:ENST00000306609  Parent=gene:ENSG00000172288   \n",
       "\n",
       "                        Biotype  \\\n",
       "94       biotype=protein_coding   \n",
       "396      biotype=protein_coding   \n",
       "542      biotype=protein_coding   \n",
       "1162     biotype=protein_coding   \n",
       "1525     biotype=protein_coding   \n",
       "...                         ...   \n",
       "3518468  biotype=protein_coding   \n",
       "3518527  biotype=protein_coding   \n",
       "3518777  biotype=protein_coding   \n",
       "3519057  biotype=protein_coding   \n",
       "3519268  biotype=protein_coding   \n",
       "\n",
       "                                                Other_tags  \n",
       "94       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "396      [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "542      [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "1162     [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "1525     [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "...                                                    ...  \n",
       "3518468  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3518527  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3518777  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519057  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519268  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "\n",
       "[19226 rows x 13 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f1203622-c631-480c-82ad-a941a6673941",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['mRNA'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect[\"Type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "50506876-e5cc-4818-ae95-296242f54760",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000641515',\n",
       " 'Parent=gene:ENSG00000186092',\n",
       " 'Name=OR4F5-201',\n",
       " 'biotype=protein_coding',\n",
       " 'ccdsid=CCDS30547.2',\n",
       " 'tag=basic,Ensembl_canonical,MANE_Select,GENCODE Primary',\n",
       " 'transcript_id=ENST00000641515',\n",
       " 'version=2']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect[\"Tags\"][94].split(\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23d5bf67-b9f6-4e53-b5da-60b7fe75ec71",
   "metadata": {},
   "source": [
    "Como espectavel, as linhas correspondem a transcritos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ef8d7245-d6be-4015-bce8-07cb7093450f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "#genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "#CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "#UTR3 = tanscripts_df.loc[tanscripts_df['Type'] == \"three_prime_UTR\"]\n",
    "#UTR5 = tanscripts_df.loc[tanscripts_df['Type'] == \"five_prime_UTR\"]\n",
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "67c11abe-f72b-4b96-8f88-1556a029e124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Other_tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000450305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000450305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519839</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212184.0</td>\n",
       "      <td>57213125.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519840</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213204.0</td>\n",
       "      <td>57213357.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519841</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213526.0</td>\n",
       "      <td>57213602.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519842</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213880.0</td>\n",
       "      <td>57213964.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519843</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214350.0</td>\n",
       "      <td>57214397.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1668828 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome  Type  First_index  Last_index Value Signal Unknown  \\\n",
       "18        1  havana  exon      11869.0     12227.0     .      +       .   \n",
       "19        1  havana  exon      12613.0     12721.0     .      +       .   \n",
       "20        1  havana  exon      13221.0     14409.0     .      +       .   \n",
       "24        1  havana  exon      12010.0     12057.0     .      +       .   \n",
       "25        1  havana  exon      12179.0     12227.0     .      +       .   \n",
       "...      ..     ...   ...          ...         ...   ...    ...     ...   \n",
       "3519839   Y  havana  exon   57212184.0  57213125.0     .      -       .   \n",
       "3519840   Y  havana  exon   57213204.0  57213357.0     .      -       .   \n",
       "3519841   Y  havana  exon   57213526.0  57213602.0     .      -       .   \n",
       "3519842   Y  havana  exon   57213880.0  57213964.0     .      -       .   \n",
       "3519843   Y  havana  exon   57214350.0  57214397.0     .      -       .   \n",
       "\n",
       "                                                      Tags   ID  \\\n",
       "18       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "19       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "20       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "24       Parent=transcript:ENST00000450305;Name=ENSE000...  NaN   \n",
       "25       Parent=transcript:ENST00000450305;Name=ENSE000...  NaN   \n",
       "...                                                    ...  ...   \n",
       "3519839  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519840  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519841  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519842  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519843  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "\n",
       "                                    Parent Biotype  \\\n",
       "18       Parent=transcript:ENST00000456328     NaN   \n",
       "19       Parent=transcript:ENST00000456328     NaN   \n",
       "20       Parent=transcript:ENST00000456328     NaN   \n",
       "24       Parent=transcript:ENST00000450305     NaN   \n",
       "25       Parent=transcript:ENST00000450305     NaN   \n",
       "...                                    ...     ...   \n",
       "3519839  Parent=transcript:ENST00000711270     NaN   \n",
       "3519840  Parent=transcript:ENST00000711270     NaN   \n",
       "3519841  Parent=transcript:ENST00000711270     NaN   \n",
       "3519842  Parent=transcript:ENST00000711270     NaN   \n",
       "3519843  Parent=transcript:ENST00000711270     NaN   \n",
       "\n",
       "                                                Other_tags  \n",
       "18       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "19       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "20       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "24       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "25       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "...                                                    ...  \n",
       "3519839  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519840  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519841  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519842  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519843  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "\n",
       "[1668828 rows x 13 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d8c4a8de-862a-482a-8b7e-cb7f0f8a2e21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Parent=transcript:ENST00000456328',\n",
       " 'Name=ENSE00002234944',\n",
       " 'constitutive=1',\n",
       " 'ensembl_end_phase=-1',\n",
       " 'ensembl_phase=-1',\n",
       " 'exon_id=ENSE00002234944',\n",
       " 'rank=1',\n",
       " 'version=1']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons[\"Tags\"][18].split(\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c7670e-1353-489d-b8bd-b18ff90ef881",
   "metadata": {},
   "source": [
    "### Para procurar transcrito a transcrito verificar os elementos que tenham a anotação: 'Parent=transcript:.............'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "356364ba-3553-4aa3-9f8d-413306f12449",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RegionsByTranscript():\n",
    "    counter = 0\n",
    "    for row in tanscripts_df_ManeSelect.index:\n",
    "        #print(\"transcripto:\", tanscripts_df_ManeSelect.values[row])\n",
    "        ID_transcript = tanscripts_df_ManeSelect['ID'][row]\n",
    "        transcript = ID_transcript.split(\"=\")[1]\n",
    "        Parent_Tag = \"Parent=\" + transcript\n",
    "        #Parent=transcript:.............'\n",
    "        transcriptRegions_df = tanscripts_df.loc[tanscripts_df[\"Parent\"] == Parent_Tag]\n",
    "        if counter == 0:\n",
    "            #print(\"sou a primeira iteração do novo df\")\n",
    "            MainTranscripts_df = transcriptRegions_df\n",
    "            counter += 1\n",
    "        else:\n",
    "            #print(\"index: \", row)\n",
    "            MainTranscripts_df = pd.concat([MainTranscripts_df, transcriptRegions_df])\n",
    "    return MainTranscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c76ad2-851f-49d1-9d74-980390eeaec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_transcripts_df = RegionsByTranscript()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "280646e8-71cc-4dec-bc4d-47ad13d8da60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Other_tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>65433.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520.0</td>\n",
       "      <td>65573.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565.0</td>\n",
       "      <td>65573.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000493376</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037.0</td>\n",
       "      <td>70008.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000493376</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>69037.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000641515;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000641515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519076</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25052074.0</td>\n",
       "      <td>25052104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000382287;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000382287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519270</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519271</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519272</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624527.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519273</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>393850 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "95        1          havana  exon      65419.0     65433.0     .      +   \n",
       "98        1          havana  exon      65520.0     65573.0     .      +   \n",
       "99        1          havana   CDS      65565.0     65573.0     .      +   \n",
       "100       1          havana   CDS      69037.0     70008.0     .      +   \n",
       "101       1          havana  exon      69037.0     71585.0     .      +   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3519076   Y          havana  exon   25052074.0  25052104.0     .      -   \n",
       "3519270   Y  ensembl_havana  exon   25622117.0  25624034.0     .      +   \n",
       "3519271   Y  ensembl_havana   CDS   25622443.0  25624034.0     .      +   \n",
       "3519272   Y  ensembl_havana   CDS   25624455.0  25624527.0     .      +   \n",
       "3519273   Y  ensembl_havana  exon   25624455.0  25624902.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \\\n",
       "95            .  Parent=transcript:ENST00000641515;Name=ENSE000...   \n",
       "98            .  Parent=transcript:ENST00000641515;Name=ENSE000...   \n",
       "99            0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...   \n",
       "100           0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...   \n",
       "101           .  Parent=transcript:ENST00000641515;Name=ENSE000...   \n",
       "...         ...                                                ...   \n",
       "3519076       .  Parent=transcript:ENST00000382287;Name=ENSE000...   \n",
       "3519270       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "3519271       0  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519272       1  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519273       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "\n",
       "                             ID                             Parent Biotype  \\\n",
       "95                          NaN  Parent=transcript:ENST00000641515     NaN   \n",
       "98                          NaN  Parent=transcript:ENST00000641515     NaN   \n",
       "99       ID=CDS:ENSP00000493376  Parent=transcript:ENST00000641515     NaN   \n",
       "100      ID=CDS:ENSP00000493376  Parent=transcript:ENST00000641515     NaN   \n",
       "101                         NaN  Parent=transcript:ENST00000641515     NaN   \n",
       "...                         ...                                ...     ...   \n",
       "3519076                     NaN  Parent=transcript:ENST00000382287     NaN   \n",
       "3519270                     NaN  Parent=transcript:ENST00000306609     NaN   \n",
       "3519271  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609     NaN   \n",
       "3519272  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609     NaN   \n",
       "3519273                     NaN  Parent=transcript:ENST00000306609     NaN   \n",
       "\n",
       "                                                Other_tags  \n",
       "95       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "98       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "99       [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "100      [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "101      [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "...                                                    ...  \n",
       "3519076  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519270  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519271  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519272  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "3519273  [\"['Alias=CM000663.2,chr1,NC_000001.11']\", \"['...  \n",
       "\n",
       "[393850 rows x 13 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_transcripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d991d040-64a4-4520-a63d-5b4ff72580f1",
   "metadata": {},
   "source": [
    "De todos os elementos são agrupados os elementos que pertencem aos transcritos principais \"Mane_Select\".\n",
    "A densidade de cada tipo de elemeto é feita considerando o intervalo desde o indice inicial mais baixo até ao indice final de maior valor, sendo portanto uma densidade conservativa.\n",
    "Relembrar forma de verificar a densidade dos intrões"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8db5c514-3218-471d-a3c0-c49041eeccee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49bbd252-0fed-472e-a11c-f5eba8d3f97c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#UTR5_overlap = UTR5.loc[UTR5[\"Tags\"] == UTR5[\"Tags\"][3516301]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83106f9b-07cf-4dea-8603-d7c41f78e66c",
   "metadata": {},
   "source": [
    "## Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cd7ce5a1-dd7c-445d-8388-e423544af64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_dict_toDF(dictionary):\n",
    "    transcripts_matrix = []\n",
    "    concat_list = []\n",
    "    for value in dictionary.values():\n",
    "        concat_list.append(value[0])\n",
    "    return pd.concat(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "86fb3cf3-3746-4a7b-ae73-8dafdbf4be72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcripts_NoOverlaps(element_type = \"CDS\"):\n",
    "    element_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    elements_dict = {}\n",
    "    #elements_dict.update({\"header\": element_df.columns.tolist()})\n",
    "    count = 0\n",
    "    for row in element_df.index:\n",
    "        Tags = element_df[\"Tags\"][row]\n",
    "        transcript_id = Tags.split(\";\")[0]\n",
    "        transcript_length = element_df[\"Last_index\"][row] - element_df[\"First_index\"][row]\n",
    "        if transcript_id not in elements_dict.keys():\n",
    "        #if elements_dict.get(transcript_id) == None:\n",
    "            #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "            elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        else:\n",
    "            #print(\"resolving overlap...\")\n",
    "            if transcript_length > elements_dict[transcript_id][1]:\n",
    "                #elements_dict.update({transcript_id: (element_df.values[row], transcript_length)})\n",
    "                elements_dict.update({transcript_id: (element_df.iloc[[count]], transcript_length)})\n",
    "        count+=1\n",
    "    return transcripts_dict_toDF(elements_dict)\n",
    "    #transcripts_matrix = []\n",
    "    #for value in elements_dict.values():\n",
    "    #    transcripts_matrix.append(value[0])\n",
    "    #return elements_dict\n",
    "    #return pd.DataFrame(elements_dict.values())\n",
    "    #return pd.DataFrame(transcripts_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "867e6833-f85a-473a-add0-e59b583c5f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_matrix(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    The matrix representing all intances of a sequence type available from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        #print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            chr_array[int(chr_sequence[\"First_index\"][row]) : int(chr_sequence[\"Last_index\"][row])] += 1\n",
    "                \n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed94b2e9-f714-4ecf-96bc-ea1f3cf8e92b",
   "metadata": {},
   "source": [
    "### Fazer a função para apenas um cromossoma de cada vez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1c2d7876-334e-4c10-ab98-db71272c3f1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Função para considerar todo o intervalo entre o primeiro indice até ao ultimo indice dos transcritos\n",
    "def element_chr_array(chromosomes_df, sequence_df, chr_n):\n",
    "    \"\"\"\n",
    "    The matrix representing all intances first instance until the last instance of the sequence type from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    chr_n: String only number ex: \"1\"\n",
    "    \"\"\"\n",
    "    print(\"searching cytobands in chromossome: \", chr_n)\n",
    "    chromosome = chromosomes_df.loc[chromosomes_df[\"Chr\"] == chr_n]\n",
    "    chr_size = int(chromosome[\"Last_index\"])\n",
    "    chr_sequence = sequence_df.loc[sequence_df['Chr'] == chr_n]\n",
    "    chr_array = np.zeros((chr_size), dtype=int)\n",
    "    #elements region\n",
    "    sequence_indexes = []\n",
    "    for row in chr_sequence.index:\n",
    "        sequence_indexes.append(int(chr_sequence[\"First_index\"][row]))\n",
    "        sequence_indexes.append(int(chr_sequence[\"Last_index\"][row]))\n",
    "        sequence_indexes = [min(sequence_indexes), max(sequence_indexes)]\n",
    "    print(\"element's initial and final indexes:\", sequence_indexes)\n",
    "    chr_array[sequence_indexes[0] : sequence_indexes[1]] += 1\n",
    "    #print(chr_array[sequence_indexes[0]-20 : sequence_indexes[1]]+20)\n",
    "    return chr_array\n",
    "\n",
    "\n",
    "#Juntar os arrays talvez não é boa ideia, memória"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "89545755-91b4-44e2-afe6-e1f8bea46e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity_chr(cytobands_df, chr_array):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    chr_densities = []\n",
    "    chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for row in chr_cytobands.index:\n",
    "        cytoband_array = chr_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "        dr_length = np.count_nonzero(cytoband_array)\n",
    "        cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "        density = dr_length/cytoband_length\n",
    "        chr_densities.append(round(density, 10))\n",
    "    return chr_densities   \n",
    "    #cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "f169403a-f4f1-4fce-9e84-2035ffc51651",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity_chr(cytobands_df, sequence_array, exons_array, chr_n):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == \"Chr\" + str(chr_n)]\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for row in chr_cytobands.index:\n",
    "        region_cytoband_array = sequence_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "        exon_cytoband_array = exons_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            \n",
    "        region_indexes = np.where(region_cytoband_array != 0)[0].tolist()\n",
    "        exon_indexes = np.where(exon_cytoband_array != 0)[0].tolist()\n",
    "        exon_length = len(np.intersect1d(region_indexes, exon_indexes))    #indices que estam na região e são exões\n",
    "        intron_length = len(np.setdiff1d(region_indexes, exon_indexes))      #indices que estam na região mas não são exões\n",
    "            \n",
    "        cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "        exon_density = exon_length/cytoband_length\n",
    "        intron_density = intron_length/cytoband_length\n",
    "        exon_densities.append(round(exon_density, 10))\n",
    "        intron_densities.append(round(intron_density, 10))   #max 6    \n",
    "    return exon_densities, intron_densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0382a318-f581-4ba7-bc60-aae7889c145b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr_process(chromosomes_df, chr_n, cytobands_df):\n",
    "    #for chr in chromosomes_df[\"Chr\"]:\n",
    "    sequence_dfs = [CDSs, five_prime_UTR, three_prime_UTR, exons]\n",
    "    sequence_types = [\"CDS\", \"five_prime_UTR\", \"three_prime_UTR\", \"exon\"]\n",
    "    ### criar já o array do chromossoma correspondente à presença de exons\n",
    "    chr_exons_array = element_chr_array(chromosomes_df, exons, chr_n)\n",
    "    ### Para cada um dos 3 tipos de sequencias calcular as densidades necessárias\n",
    "    ###CDS\n",
    "    chr_cds_array = element_chr_array(chromosomes_df, cds_df, chr_n)\n",
    "    chr_cds_dens = calc_elementDensity_chr(cytobands_df, chr_array)\n",
    "    cds_exon_dens, cds_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_array, chr_exons_array)\n",
    "    ###five_prime_UTR\n",
    "    chr_utr5_array = element_chr_array(chromosomes_df, utr5_df, chr_n)\n",
    "    chr_utr5_dens = calc_elementDensity_chr(cytobands_df, chr_array)\n",
    "    utr5_exon_dens, utr5_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_array, chr_exons_array)\n",
    "    ###three_prime_UTR\n",
    "    chr_utr3_array = element_chr_array(chromosomes_df, utr3_df, chr_n)\n",
    "    chr_utr3_dens = calc_elementDensity_chr(cytobands_df, chr_array)\n",
    "    utr3_exon_dens, utr3_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_array, chr_exons_array)\n",
    "    ###exon\n",
    "    chr_exon_dens = calc_elementDensity_chr(cytobands_df, chr_exons_array)\n",
    "    ###Guardar as variáveis em arrays separados dentro ou fora da função\n",
    "    return chr_cds_dens, cds_exon_dens, cds_intron_dens, chr_utr5_dens, utr5_exon_dens, utr5_intron_dens, chr_utr3_dens, utr3_exon_dens, utr3_intron_dens, chr_exon_dens\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e866ec-f64d-4aa4-a273-6a13347fb3eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_allChr(chromosomes_df, chr_n, cytobands_df):\n",
    "    densities = []\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):\n",
    "        chr_process(chromosomes_df, chr_n, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f67eeca-ae66-4f7a-b67f-27273aa83aa9",
   "metadata": {},
   "source": [
    "Funções de calculo de densidade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0053bc01-6bdd-46e1-9592-41e0f5a2599b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity(cytobands_df, region_matrix, col_name=\"Element_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    densities = []\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):             \n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            dr_length = np.count_nonzero(cytoband_array)\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            density = dr_length/cytoband_length\n",
    "            densities.append(round(density, 10))\n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e6e7d8a-9c5d-4c16-95ba-720845fc909c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity(cytobands_df, region_matrix, exons_matrix, col_name1=\"exon_density\", col_name2=\"intron_density\"):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):\n",
    "        #print(\"Calculating densities in chromosome \", cytobands_df[\"Chromossome\"].unique()[chrom], \" ...\")\n",
    "        chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == cytobands_df[\"Chromossome\"].unique()[chrom]]\n",
    "        #Calculo da densidade para cada cytoband\n",
    "        for row in chr_cytobands.index:\n",
    "            ############################ Resolver problema, o conjunto dos exons contem todo o conjunto dos CDSs ###############################\n",
    "            region_cytoband_array = region_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            exon_cytoband_array = exons_matrix[chrom][int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "            #print(\"cytoband: \", chr_cytobands[\"Citoband\"][row])\n",
    "            #print(\"array_regiao: \", region_cytoband_array)\n",
    "            #print(\"exon_regiao: \", exon_cytoband_array)\n",
    "            region_indexes = np.where(region_cytoband_array != 0)[0].tolist()\n",
    "            exon_indexes = np.where(exon_cytoband_array != 0)[0].tolist()\n",
    "            #print(\"number of elements of region: \", len(region_indexes))\n",
    "            #print(\"number of exons in region: \", len(np.intersect1d(region_indexes, exon_indexes)))\n",
    "            #print(\"exons in CDS region: \", np.intersect1d(region_indexes, exon_indexes))\n",
    "            #print(\"introns in CDS region: \", np.setdiff1d(region_indexes, exon_indexes))\n",
    "            exon_length = len(np.intersect1d(region_indexes, exon_indexes))    #indices que estam na região e são exões\n",
    "            intron_length = len(np.setdiff1d(region_indexes, exon_indexes))      #indices que estam na região mas não são exões\n",
    "            #for region_element in region_indexes:\n",
    "            #    if region_element in exon_indexes:\n",
    "            #        exon_length+=1\n",
    "            #    elif region_element not in exon_indexes:\n",
    "            #        intron_length+=1\n",
    "            cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "            exon_density = exon_length/cytoband_length\n",
    "            intron_density = intron_length/cytoband_length\n",
    "            exon_densities.append(round(exon_density, 10))\n",
    "            intron_densities.append(round(intron_density, 10))   #max 6\n",
    "            \n",
    "        \n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name1, exon_densities, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), col_name2, intron_densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93fd572c-e059-4592-8147-cbeb4f16de06",
   "metadata": {},
   "outputs": [],
   "source": [
    "#array1 = [1, 2, 3, 6, 7, 8]\n",
    "#array2 = [2, 3, 4, 7, 8, 9]\n",
    "#print(np.intersect1d(array1, array2))\n",
    "#print(np.setdiff1d(array1, array2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "84e048d0-2f72-453e-bdb7-d9007893a020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    df_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        df_cols += col + \"\\t\"\n",
    "    df_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(df_cols))\n",
    "    for row in df.values:\n",
    "        df_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            df_row += str(col) + \"\\t\"\n",
    "        df_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(df_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1e8cdd-042e-4071-9737-69169c01ee2d",
   "metadata": {},
   "source": [
    "# Code running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fd457d08-3d9c-4ad7-b34a-e4cdd5aa389d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  \n",
       "0              0.087516   0.006384   0.084476     0.326328  \n",
       "1              0.084077   0.013902   0.079223     0.281556  \n",
       "2              0.056236   0.012909   0.051421     0.275410  \n",
       "3              0.049082   0.013229   0.042217     0.404844  \n",
       "4              0.055881   0.011072   0.048928     0.404047  \n",
       "..                  ...        ...        ...          ...  \n",
       "858            0.036956   0.014211   0.029356     0.015158  \n",
       "859            0.061662   0.015902   0.052790     0.012965  \n",
       "860            0.082234   0.005789   0.078347     0.006441  \n",
       "861            0.026278   0.009164   0.020721     0.001822  \n",
       "862            0.005530   0.001643   0.004251     0.000594  \n",
       "\n",
       "[863 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoband_read = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoband_write = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr_dg_MANESelect.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoband_read, sep=\"\\t\")#, header=None))\n",
    "#cytobands_df2 = pd.read_csv(cytoband_write, sep=\"\\t\")\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5377fb6f-0db3-40d0-9c18-1632ee187249",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "083cfb80-3439-436b-abfc-8426f5ed1dc5",
   "metadata": {},
   "source": [
    "## Criação de variáveis globais necessárias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0fa9158f-d079-4c6e-bedd-6e8072a100ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Para o dataset sem filtragem   ################################################\n",
    "chromosomes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "#exon_matrix = element_matrix(chromosomes, exons)\n",
    "\n",
    "############################ Para o dataset com filtragem pela tag #########################################\n",
    "#chromosomes_noMPC = tanscripts_df_noMPC.loc[tanscripts_df_noMPC['Type'] == \"chromosome\"]\n",
    "#exons_noMPC = tanscripts_df_noMPC.loc[tanscripts_df_noMPC['Type'] == \"exon\"]\n",
    "#exon_matrix_noMPC = element_matrix(chromosomes_noMPC, exons_noMPC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "916001ef-75b6-4cd9-aa21-3fdcdb1d4fac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328201</th>\n",
       "      <td>10</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133797422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467767</th>\n",
       "      <td>11</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>135086622.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673000</th>\n",
       "      <td>12</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133275309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862634</th>\n",
       "      <td>13</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>114364328.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924501</th>\n",
       "      <td>14</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>107043718.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048326</th>\n",
       "      <td>15</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101991189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171301</th>\n",
       "      <td>16</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90338345.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329042</th>\n",
       "      <td>17</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83257441.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536644</th>\n",
       "      <td>18</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80373285.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598313</th>\n",
       "      <td>19</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58617616.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791801</th>\n",
       "      <td>2</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242193529.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063156</th>\n",
       "      <td>20</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64444167.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145930</th>\n",
       "      <td>21</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46709983.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186784</th>\n",
       "      <td>22</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50818468.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2261287</th>\n",
       "      <td>3</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>198295559.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484781</th>\n",
       "      <td>4</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>190214555.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628956</th>\n",
       "      <td>5</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>181538259.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786252</th>\n",
       "      <td>6</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170805979.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2952608</th>\n",
       "      <td>7</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>159345973.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3120542</th>\n",
       "      <td>8</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145138636.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248106</th>\n",
       "      <td>9</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>138394717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387542</th>\n",
       "      <td>MT</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387706</th>\n",
       "      <td>X</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>156040895.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3509168</th>\n",
       "      <td>Y</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>57227415.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index   Last_index Value Signal  \\\n",
       "0         1  GRCh38  chromosome          1.0  248956422.0     .      .   \n",
       "328201   10  GRCh38  chromosome          1.0  133797422.0     .      .   \n",
       "467767   11  GRCh38  chromosome          1.0  135086622.0     .      .   \n",
       "673000   12  GRCh38  chromosome          1.0  133275309.0     .      .   \n",
       "862634   13  GRCh38  chromosome          1.0  114364328.0     .      .   \n",
       "924501   14  GRCh38  chromosome          1.0  107043718.0     .      .   \n",
       "1048326  15  GRCh38  chromosome          1.0  101991189.0     .      .   \n",
       "1171301  16  GRCh38  chromosome          1.0   90338345.0     .      .   \n",
       "1329042  17  GRCh38  chromosome          1.0   83257441.0     .      .   \n",
       "1536644  18  GRCh38  chromosome          1.0   80373285.0     .      .   \n",
       "1598313  19  GRCh38  chromosome          1.0   58617616.0     .      .   \n",
       "1791801   2  GRCh38  chromosome          1.0  242193529.0     .      .   \n",
       "2063156  20  GRCh38  chromosome          1.0   64444167.0     .      .   \n",
       "2145930  21  GRCh38  chromosome          1.0   46709983.0     .      .   \n",
       "2186784  22  GRCh38  chromosome          1.0   50818468.0     .      .   \n",
       "2261287   3  GRCh38  chromosome          1.0  198295559.0     .      .   \n",
       "2484781   4  GRCh38  chromosome          1.0  190214555.0     .      .   \n",
       "2628956   5  GRCh38  chromosome          1.0  181538259.0     .      .   \n",
       "2786252   6  GRCh38  chromosome          1.0  170805979.0     .      .   \n",
       "2952608   7  GRCh38  chromosome          1.0  159345973.0     .      .   \n",
       "3120542   8  GRCh38  chromosome          1.0  145138636.0     .      .   \n",
       "3248106   9  GRCh38  chromosome          1.0  138394717.0     .      .   \n",
       "3387542  MT  GRCh38  chromosome          1.0      16569.0     .      .   \n",
       "3387706   X  GRCh38  chromosome          1.0  156040895.0     .      .   \n",
       "3509168   Y  GRCh38  chromosome          1.0   57227415.0     .      .   \n",
       "\n",
       "        Unknown                                        Identifiers  \n",
       "0             .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "328201        .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  \n",
       "467767        .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  \n",
       "673000        .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  \n",
       "862634        .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  \n",
       "924501        .  ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...  \n",
       "1048326       .  ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...  \n",
       "1171301       .  ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...  \n",
       "1329042       .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...  \n",
       "1536644       .  ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...  \n",
       "1598313       .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...  \n",
       "1791801       .  ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...  \n",
       "2063156       .  ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...  \n",
       "2145930       .  ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...  \n",
       "2186784       .  ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...  \n",
       "2261287       .  ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...  \n",
       "2484781       .  ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...  \n",
       "2628956       .  ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...  \n",
       "2786252       .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...  \n",
       "2952608       .  ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...  \n",
       "3120542       .  ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...  \n",
       "3248106       .  ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...  \n",
       "3387542       .   ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1  \n",
       "3387706       .  ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...  \n",
       "3509168       .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a22db2-57e3-470c-afa9-2031e8b044a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "indices do elemento inicial e final: [11869, 248937043]\n",
      "searching citobands in chromossome:  10\n",
      "indices do elemento inicial e final: [14061, 133778699]\n",
      "searching citobands in chromossome:  11\n",
      "indices do elemento inicial e final: [75780, 135075908]\n",
      "searching citobands in chromossome:  12\n",
      "indices do elemento inicial e final: [12310, 133238549]\n",
      "searching citobands in chromossome:  13\n",
      "indices do elemento inicial e final: [18174010, 114346637]\n",
      "searching citobands in chromossome:  14\n",
      "indices do elemento inicial e final: [16057472, 106879812]\n",
      "searching citobands in chromossome:  15\n",
      "indices do elemento inicial e final: [19878555, 101979093]\n",
      "searching citobands in chromossome:  16\n",
      "indices do elemento inicial e final: [11555, 90222851]\n",
      "searching citobands in chromossome:  17\n",
      "indices do elemento inicial e final: [64099, 83240804]\n",
      "searching citobands in chromossome:  18\n",
      "indices do elemento inicial e final: [11103, 80247514]\n",
      "searching citobands in chromossome:  19\n",
      "indices do elemento inicial e final: [60951, 58605223]\n"
     ]
    }
   ],
   "source": [
    "############################ Abordagem conservadora, regiões mais vastas ##################################\n",
    "exon_matrix = element_matrix_conservative(chromosomes, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "16adb942-48c6-4bda-8d19-2a5425552610",
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Para o dataset com filtragem de transcritos ###################################################\n",
    "#exons_noOverlaps_df = transcripts_NoOverlaps(\"exon\")\n",
    "#exon_matrix_noOvrlp = element_matrix(chromosomes, exons_noOverlaps_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "630588ea-ca59-42bb-8713-6f74e47dd2ad",
   "metadata": {},
   "source": [
    "## Sem filtragem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3c1d4022-9ac9-4907-97ad-6ee7b43491af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_transcriptDens(cytobands_df, element_type, tanscripts_df, chromosomes_df): #element_type = \"CDS\"\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    sub_trascripts_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    print(\"building matriz...\")\n",
    "    transcripts_matrix = element_matrix(chromosomes_df, sub_trascripts_df)\n",
    "    colname = element_type+\"s_density\"\n",
    "    print(\"doing calculations...\")\n",
    "    calc_elementDensity(cytobands_df, transcripts_matrix, colname)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "07516778-0141-486b-814d-ad58bdb085ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_Trcpts_IExons(cytobands_df, element_type , tanscripts_df, chromosomes_df, exon_matrix): #element_type = \"CDS\"\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    sub_trascripts_df = tanscripts_df.loc[tanscripts_df['Type'] == element_type]\n",
    "    exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "    print(\"building matriz...\")\n",
    "    transcripts_matrix = element_matrix(chromosomes_df, sub_trascripts_df)\n",
    "    exon_matrix = element_matrix(chromosomes_df, exons)\n",
    "    print(\"doing calculations...\")\n",
    "    colname1 = element_type+\"Exon_density\"\n",
    "    colname2 = element_type+\"Intron_density\"\n",
    "    calc_ExonIntronDensity(cytobands_df, transcripts_matrix, exon_matrix, colname1, colname2)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c4dbff3-8457-4e99-b1a6-aca386e290e2",
   "metadata": {},
   "source": [
    "## Com filtragem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6fb2715-301c-4988-acbf-e2f877c85b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_transcriptDens_noOverlps(cytobands_df, element_type = \"CDS\", chromosomes_df = chromosomes):\n",
    "    chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_noOverlaps_df = transcripts_NoOverlaps(element_type)\n",
    "    print(\"building matriz...\")\n",
    "    transcript_matrix = element_matrix(chromosomes_df, element_noOverlaps_df)\n",
    "    colname = element_type+\"s_density\"\n",
    "    print(\"doing calculations...\")\n",
    "    calc_elementDensity(cytobands_df, transcript_matrix, colname)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68b2db56-016e-4314-a28d-405156aa373a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def study_Trcpts_IExons_noOverlps(cytobands_df, element_type = \"CDS\", chromosomes_df = chromosomes, exon_matrix_noOvrlp = exon_matrix_noOvrlp):\n",
    "    #chromosomes_df = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "    element_noOverlaps_df = transcripts_NoOverlaps(element_type)\n",
    "    #exons_dict = transcripts_NoOverlaps(\"exon\")\n",
    "    #exons_noOverlaps_df = transcripts_dict_toDF(exons_dict)\n",
    "    print(\"a fazer a matriz...\")\n",
    "    transcript_matrix = element_matrix(chromosomes_df, element_noOverlaps_df)\n",
    "    #exon_matrix = element_matrix(chromosmes, exons_noOverlaps_df)\n",
    "    colname1 = element_type+\"Intron_density\"\n",
    "    colname2 = element_type+\"Exon_density\"\n",
    "    print(\"a fazer os calculos...\")\n",
    "    calc_ExonIntronDensity(cytobands_df, transcript_matrix, exon_matrix_noOvrlp, colname1, colname2)\n",
    "    writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01879f35-00d3-4754-89a5-1b34a9447e01",
   "metadata": {},
   "source": [
    "## Com Nova filtragem é aplicar as funções sem filtragem no dataframe já filtrado: tanscripts_df_ManeSelect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9495e9-9702-4571-a78f-9d4835b82301",
   "metadata": {},
   "source": [
    "## CDS density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b2e61635-55f7-4429-9bca-4478b22c7104",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_transcriptDens(cytobands_df, \"CDS\", tanscripts_df, chromosomes)\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"CDS\")\n",
    "study_transcriptDens(cytobands_df, \"CDS\", tanscripts_df_noMPC, chromosomes_noMPC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d690d7f-d2aa-453e-b132-1ac6e85dcf68",
   "metadata": {},
   "source": [
    "### CDSs' Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d2701f6-bff2-4b7e-ab92-d701c1bed997",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"CDS\", tanscripts_df, chromosomes, exon_matrix)\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"CDS\")\n",
    "study_Trcpts_IExons(cytobands_df, \"CDS\", tanscripts_df_noMPC, chromosomes_noMPC, exon_matrix_noMPC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9b6d07a5-655d-44d4-a2ad-8ea1b2f48ad1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  \n",
       "0           0.037748                0.0  \n",
       "1           0.015817                0.0  \n",
       "2           0.024363                0.0  \n",
       "3           0.012765                0.0  \n",
       "4           0.029924                0.0  \n",
       "..               ...                ...  \n",
       "858         0.003879                0.0  \n",
       "859         0.002446                0.0  \n",
       "860         0.004806                0.0  \n",
       "861         0.002666                0.0  \n",
       "862         0.000130                0.0  \n",
       "\n",
       "[863 rows x 13 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b149728d-101f-4b84-9fa8-a77b5686e219",
   "metadata": {},
   "source": [
    "## 5 Prime UTR density studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2778db1a-3cd0-46d2-961c-4a3b807334a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_transcriptDens(cytobands_df, \"five_prime_UTR\", tanscripts_df, chromosomes))\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"five_prime_UTR\")\n",
    "study_transcriptDens(cytobands_df, \"five_prime_UTR\", tanscripts_df_noMPC, chromosomes_noMPC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1bd2ce3-76ac-481a-a4af-20733256d243",
   "metadata": {},
   "source": [
    "### 5 Prime UTR Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c2c4b5e9-f175-4756-9342-b832d4317a4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"five_prime_UTR\", tanscripts_df, chromosomes, exon_matrix)\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"five_prime_UTR\")\n",
    "study_Trcpts_IExons(cytobands_df, \"five_prime_UTR\", tanscripts_df_noMPC, chromosomes_noMPC, exon_matrix_noMPC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "184ef8f3-92ec-4ed3-9b38-8ed4b3d0fb50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  five_prime_UTRs_density  \\\n",
       "0           0.037748                0.0                 0.012022   \n",
       "1           0.015817                0.0                 0.003561   \n",
       "2           0.024363                0.0                 0.006496   \n",
       "3           0.012765                0.0                 0.004600   \n",
       "4           0.029924                0.0                 0.005499   \n",
       "..               ...                ...                      ...   \n",
       "858         0.003879                0.0                 0.001027   \n",
       "859         0.002446                0.0                 0.000346   \n",
       "860         0.004806                0.0                 0.000675   \n",
       "861         0.002666                0.0                 0.000838   \n",
       "862         0.000130                0.0                 0.000006   \n",
       "\n",
       "     five_prime_UTRExon_density  five_prime_UTRIntron_density  \n",
       "0                      0.012022                           0.0  \n",
       "1                      0.003561                           0.0  \n",
       "2                      0.006496                           0.0  \n",
       "3                      0.004600                           0.0  \n",
       "4                      0.005499                           0.0  \n",
       "..                          ...                           ...  \n",
       "858                    0.001027                           0.0  \n",
       "859                    0.000346                           0.0  \n",
       "860                    0.000675                           0.0  \n",
       "861                    0.000838                           0.0  \n",
       "862                    0.000006                           0.0  \n",
       "\n",
       "[863 rows x 16 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51f8d5e-d87f-47bc-89ef-60478ed00610",
   "metadata": {},
   "source": [
    "## Prime 3 UTR density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2b70b50a-9eae-4fa6-932b-9ce572f7fd78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_transcriptDens(cytobands_df, \"three_prime_UTR\", tanscripts_df, chromosomes))\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"three_prime_UTR\")\n",
    "study_transcriptDens(cytobands_df, \"three_prime_UTR\", tanscripts_df_noMPC, chromosomes_noMPC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d99a6f99-a946-4123-8f99-8984c9b011a0",
   "metadata": {},
   "source": [
    "## 3 Prime UTR Exon and Intron densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a6f4ede0-3475-415d-995e-5096c30268f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_Trcpts_IExons(cytobands_df, \"three_prime_UTR\", tanscripts_df, chromosomes, exon_matrix)\n",
    "#study_Trcpts_IExons_noOverlps(cytobands_df, \"three_prime_UTR\")\n",
    "study_Trcpts_IExons(cytobands_df, \"three_prime_UTR\", tanscripts_df_noMPC, chromosomes_noMPC, exon_matrix_noMPC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "383b40f1-7696-41e8-bb07-3dceca86bd09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  five_prime_UTRs_density  \\\n",
       "0           0.037748                0.0                 0.012022   \n",
       "1           0.015817                0.0                 0.003561   \n",
       "2           0.024363                0.0                 0.006496   \n",
       "3           0.012765                0.0                 0.004600   \n",
       "4           0.029924                0.0                 0.005499   \n",
       "..               ...                ...                      ...   \n",
       "858         0.003879                0.0                 0.001027   \n",
       "859         0.002446                0.0                 0.000346   \n",
       "860         0.004806                0.0                 0.000675   \n",
       "861         0.002666                0.0                 0.000838   \n",
       "862         0.000130                0.0                 0.000006   \n",
       "\n",
       "     five_prime_UTRExon_density  five_prime_UTRIntron_density  \\\n",
       "0                      0.012022                           0.0   \n",
       "1                      0.003561                           0.0   \n",
       "2                      0.006496                           0.0   \n",
       "3                      0.004600                           0.0   \n",
       "4                      0.005499                           0.0   \n",
       "..                          ...                           ...   \n",
       "858                    0.001027                           0.0   \n",
       "859                    0.000346                           0.0   \n",
       "860                    0.000675                           0.0   \n",
       "861                    0.000838                           0.0   \n",
       "862                    0.000006                           0.0   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRExon_density  \\\n",
       "0                    0.032887                     0.032887   \n",
       "1                    0.022181                     0.022181   \n",
       "2                    0.033723                     0.033723   \n",
       "3                    0.017843                     0.017843   \n",
       "4                    0.035758                     0.035758   \n",
       "..                        ...                          ...   \n",
       "858                  0.002867                     0.002867   \n",
       "859                  0.001032                     0.001032   \n",
       "860                  0.003568                     0.003568   \n",
       "861                  0.002204                     0.002204   \n",
       "862                  0.000104                     0.000104   \n",
       "\n",
       "     three_prime_UTRIntron_density  \n",
       "0                              0.0  \n",
       "1                              0.0  \n",
       "2                              0.0  \n",
       "3                              0.0  \n",
       "4                              0.0  \n",
       "..                             ...  \n",
       "858                            0.0  \n",
       "859                            0.0  \n",
       "860                            0.0  \n",
       "861                            0.0  \n",
       "862                            0.0  \n",
       "\n",
       "[863 rows x 19 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc99fd3-b968-4579-9ce9-c0ceec7f6925",
   "metadata": {},
   "source": [
    "## Exons density study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b8e17263-abd9-4634-b815-4ef21afd7bf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building matriz...\n",
      "doing calculations...\n"
     ]
    }
   ],
   "source": [
    "#study_transcriptDens(cytobands_df, \"exon\")\n",
    "#study_transcriptDens_noOverlps(cytobands_df, \"exon\")\n",
    "study_transcriptDens(cytobands_df, \"exon\", tanscripts_df_noMPC, chromosomes_noMPC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71d7563-10f1-4782-a714-2dc899e78b2d",
   "metadata": {},
   "source": [
    "## Resultados finais no dataset de cytobands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c86fb72-ef32-4537-bf93-baf93c91d71b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2665ba-fab3-4863-bc61-428452c2a574",
   "metadata": {},
   "source": [
    "## Resultados depois de filtrar transcritos MANE_Plus_Critical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d902ad05-249f-4cd0-b698-a41e53d254b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "      <th>exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.078555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.082023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.092777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  five_prime_UTRs_density  \\\n",
       "0           0.037748                0.0                 0.012022   \n",
       "1           0.015817                0.0                 0.003561   \n",
       "2           0.024363                0.0                 0.006496   \n",
       "3           0.012765                0.0                 0.004600   \n",
       "4           0.029924                0.0                 0.005499   \n",
       "..               ...                ...                      ...   \n",
       "858         0.003879                0.0                 0.001027   \n",
       "859         0.002446                0.0                 0.000346   \n",
       "860         0.004806                0.0                 0.000675   \n",
       "861         0.002666                0.0                 0.000838   \n",
       "862         0.000130                0.0                 0.000006   \n",
       "\n",
       "     five_prime_UTRExon_density  five_prime_UTRIntron_density  \\\n",
       "0                      0.012022                           0.0   \n",
       "1                      0.003561                           0.0   \n",
       "2                      0.006496                           0.0   \n",
       "3                      0.004600                           0.0   \n",
       "4                      0.005499                           0.0   \n",
       "..                          ...                           ...   \n",
       "858                    0.001027                           0.0   \n",
       "859                    0.000346                           0.0   \n",
       "860                    0.000675                           0.0   \n",
       "861                    0.000838                           0.0   \n",
       "862                    0.000006                           0.0   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRExon_density  \\\n",
       "0                    0.032887                     0.032887   \n",
       "1                    0.022181                     0.022181   \n",
       "2                    0.033723                     0.033723   \n",
       "3                    0.017843                     0.017843   \n",
       "4                    0.035758                     0.035758   \n",
       "..                        ...                          ...   \n",
       "858                  0.002867                     0.002867   \n",
       "859                  0.001032                     0.001032   \n",
       "860                  0.003568                     0.003568   \n",
       "861                  0.002204                     0.002204   \n",
       "862                  0.000104                     0.000104   \n",
       "\n",
       "     three_prime_UTRIntron_density  exons_density  \n",
       "0                              0.0       0.170647  \n",
       "1                              0.0       0.078555  \n",
       "2                              0.0       0.082023  \n",
       "3                              0.0       0.047006  \n",
       "4                              0.0       0.092777  \n",
       "..                             ...            ...  \n",
       "858                            0.0       0.015587  \n",
       "859                            0.0       0.045282  \n",
       "860                            0.0       0.035870  \n",
       "861                            0.0       0.033834  \n",
       "862                            0.0       0.000574  \n",
       "\n",
       "[863 rows x 20 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e597745-bcfe-4359-be42-3b8d639cc3f6",
   "metadata": {},
   "source": [
    "## Resultados depois de filtrar apenas as região de maior tamanho de cada transcrito  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "520d1578-d4d3-4b78-9768-d098f23a045f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.016467</td>\n",
       "      <td>0.013392</td>\n",
       "      <td>0.003076</td>\n",
       "      <td>0.010664</td>\n",
       "      <td>0.005135</td>\n",
       "      <td>0.005529</td>\n",
       "      <td>0.026310</td>\n",
       "      <td>0.025853</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>0.133749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>0.004490</td>\n",
       "      <td>0.001576</td>\n",
       "      <td>0.003274</td>\n",
       "      <td>0.001761</td>\n",
       "      <td>0.001512</td>\n",
       "      <td>0.019776</td>\n",
       "      <td>0.019406</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>0.060012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.010266</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0.003362</td>\n",
       "      <td>0.005659</td>\n",
       "      <td>0.002073</td>\n",
       "      <td>0.003586</td>\n",
       "      <td>0.025848</td>\n",
       "      <td>0.025715</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.056062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.005996</td>\n",
       "      <td>0.004821</td>\n",
       "      <td>0.001175</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.001499</td>\n",
       "      <td>0.015875</td>\n",
       "      <td>0.015746</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.034280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.009799</td>\n",
       "      <td>0.006866</td>\n",
       "      <td>0.002934</td>\n",
       "      <td>0.004954</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.002777</td>\n",
       "      <td>0.029553</td>\n",
       "      <td>0.028516</td>\n",
       "      <td>0.001037</td>\n",
       "      <td>0.066018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>0.000290</td>\n",
       "      <td>0.000927</td>\n",
       "      <td>0.000344</td>\n",
       "      <td>0.000583</td>\n",
       "      <td>0.002534</td>\n",
       "      <td>0.002459</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.009953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.033020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>0.000802</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.000540</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.000503</td>\n",
       "      <td>0.002733</td>\n",
       "      <td>0.002733</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.001369</td>\n",
       "      <td>0.001294</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.000308</td>\n",
       "      <td>0.002016</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.016467   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.006066   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.010266   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.005996   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.009799   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.000917   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002263   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.001096   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.001369   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000050   \n",
       "\n",
       "     CDSIntron_density  CDSExon_density  five_prime_UTRs_density  \\\n",
       "0             0.013392         0.003076                 0.010664   \n",
       "1             0.004490         0.001576                 0.003274   \n",
       "2             0.006903         0.003362                 0.005659   \n",
       "3             0.004821         0.001175                 0.003569   \n",
       "4             0.006866         0.002934                 0.004954   \n",
       "..                 ...              ...                      ...   \n",
       "858           0.000627         0.000290                 0.000927   \n",
       "859           0.002263         0.000000                 0.000346   \n",
       "860           0.000802         0.000294                 0.000540   \n",
       "861           0.001294         0.000075                 0.000652   \n",
       "862           0.000043         0.000007                 0.000005   \n",
       "\n",
       "     five_prime_UTRIntron_density  five_prime_UTRExon_density  \\\n",
       "0                        0.005135                    0.005529   \n",
       "1                        0.001761                    0.001512   \n",
       "2                        0.002073                    0.003586   \n",
       "3                        0.002070                    0.001499   \n",
       "4                        0.002176                    0.002777   \n",
       "..                            ...                         ...   \n",
       "858                      0.000344                    0.000583   \n",
       "859                      0.000338                    0.000008   \n",
       "860                      0.000037                    0.000503   \n",
       "861                      0.000345                    0.000308   \n",
       "862                      0.000000                    0.000005   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRIntron_density  \\\n",
       "0                    0.026310                       0.025853   \n",
       "1                    0.019776                       0.019406   \n",
       "2                    0.025848                       0.025715   \n",
       "3                    0.015875                       0.015746   \n",
       "4                    0.029553                       0.028516   \n",
       "..                        ...                            ...   \n",
       "858                  0.002534                       0.002459   \n",
       "859                  0.000937                       0.000635   \n",
       "860                  0.002733                       0.002733   \n",
       "861                  0.002016                       0.001749   \n",
       "862                  0.000094                       0.000094   \n",
       "\n",
       "     three_prime_UTRExon_density  exons_density  \n",
       "0                       0.000456       0.133749  \n",
       "1                       0.000370       0.060012  \n",
       "2                       0.000133       0.056062  \n",
       "3                       0.000129       0.034280  \n",
       "4                       0.001037       0.066018  \n",
       "..                           ...            ...  \n",
       "858                     0.000075       0.009953  \n",
       "859                     0.000302       0.033020  \n",
       "860                     0.000000       0.021283  \n",
       "861                     0.000267       0.022278  \n",
       "862                     0.000000       0.000460  \n",
       "\n",
       "[863 rows x 20 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "510a3d3b-6634-4ee4-a067-2fa62b571fe3",
   "metadata": {},
   "source": [
    "## Resultados sem usar filtros, o conjunto dos exões cobre completamente as regiões estudadas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "238d1646-ee79-45ce-b69b-99bd9992f480",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDSs_density</th>\n",
       "      <th>CDSExon_density</th>\n",
       "      <th>CDSIntron_density</th>\n",
       "      <th>five_prime_UTRs_density</th>\n",
       "      <th>five_prime_UTRExon_density</th>\n",
       "      <th>five_prime_UTRIntron_density</th>\n",
       "      <th>three_prime_UTRs_density</th>\n",
       "      <th>three_prime_UTRExon_density</th>\n",
       "      <th>three_prime_UTRIntron_density</th>\n",
       "      <th>exons_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.037748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.012022</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.032887</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.015817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.003561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.022181</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.078555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.024363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.033723</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.082023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.012765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.004600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.017843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.029924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.035758</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.092777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.003879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.001027</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.001032</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.045282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.004806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.000675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.002666</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.000838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDSs_density  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328      0.037748   \n",
       "1              0.084077   0.013902   0.079223     0.281556      0.015817   \n",
       "2              0.056236   0.012909   0.051421     0.275410      0.024363   \n",
       "3              0.049082   0.013229   0.042217     0.404844      0.012765   \n",
       "4              0.055881   0.011072   0.048928     0.404047      0.029924   \n",
       "..                  ...        ...        ...          ...           ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158      0.003879   \n",
       "859            0.061662   0.015902   0.052790     0.012965      0.002446   \n",
       "860            0.082234   0.005789   0.078347     0.006441      0.004806   \n",
       "861            0.026278   0.009164   0.020721     0.001822      0.002666   \n",
       "862            0.005530   0.001643   0.004251     0.000594      0.000130   \n",
       "\n",
       "     CDSExon_density  CDSIntron_density  five_prime_UTRs_density  \\\n",
       "0           0.037748                0.0                 0.012022   \n",
       "1           0.015817                0.0                 0.003561   \n",
       "2           0.024363                0.0                 0.006496   \n",
       "3           0.012765                0.0                 0.004600   \n",
       "4           0.029924                0.0                 0.005499   \n",
       "..               ...                ...                      ...   \n",
       "858         0.003879                0.0                 0.001027   \n",
       "859         0.002446                0.0                 0.000346   \n",
       "860         0.004806                0.0                 0.000675   \n",
       "861         0.002666                0.0                 0.000838   \n",
       "862         0.000130                0.0                 0.000006   \n",
       "\n",
       "     five_prime_UTRExon_density  five_prime_UTRIntron_density  \\\n",
       "0                      0.012022                           0.0   \n",
       "1                      0.003561                           0.0   \n",
       "2                      0.006496                           0.0   \n",
       "3                      0.004600                           0.0   \n",
       "4                      0.005499                           0.0   \n",
       "..                          ...                           ...   \n",
       "858                    0.001027                           0.0   \n",
       "859                    0.000346                           0.0   \n",
       "860                    0.000675                           0.0   \n",
       "861                    0.000838                           0.0   \n",
       "862                    0.000006                           0.0   \n",
       "\n",
       "     three_prime_UTRs_density  three_prime_UTRExon_density  \\\n",
       "0                    0.032887                     0.032887   \n",
       "1                    0.022181                     0.022181   \n",
       "2                    0.033723                     0.033723   \n",
       "3                    0.017843                     0.017843   \n",
       "4                    0.035758                     0.035758   \n",
       "..                        ...                          ...   \n",
       "858                  0.002867                     0.002867   \n",
       "859                  0.001032                     0.001032   \n",
       "860                  0.003568                     0.003568   \n",
       "861                  0.002204                     0.002204   \n",
       "862                  0.000104                     0.000104   \n",
       "\n",
       "     three_prime_UTRIntron_density  exons_density  \n",
       "0                              0.0       0.170647  \n",
       "1                              0.0       0.078555  \n",
       "2                              0.0       0.082023  \n",
       "3                              0.0       0.047006  \n",
       "4                              0.0       0.092777  \n",
       "..                             ...            ...  \n",
       "858                            0.0       0.015587  \n",
       "859                            0.0       0.045282  \n",
       "860                            0.0       0.035870  \n",
       "861                            0.0       0.033834  \n",
       "862                            0.0       0.000574  \n",
       "\n",
       "[863 rows x 20 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9d2061bc-6f75-4a28-8156-baf3060b9db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cytobands_df.values[cytobands_df.index[70]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fc6cd1d0-454c-4354-9c18-bd9354568385",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sumOf_columns(df, colums=[\"CDSIntron_density\", \"UTR5Intron_density\", \"UTR3Intron_density\"], newColName = \"Introns_density\"):\n",
    "    \"\"\"\n",
    "    colums is a list of string\n",
    "    \"\"\"\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    new_densites = []\n",
    "    for row in df.index:\n",
    "        density =0\n",
    "        for column in colums:\n",
    "            density += df[column][row]\n",
    "        new_densites.append(round(density, 3))\n",
    "        \n",
    "    df.insert(len(df.columns), newColName, new_densites, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931407ff-fdfa-4df4-8ea6-cb2ab76afe52",
   "metadata": {},
   "outputs": [],
   "source": [
    "sumOf_columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c8bf14c9-0b59-42bc-8b73-eec9e7cb7e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(cytoPath_write, cytobands_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "586f8eb0-e8fd-4241-8e14-6da5756f8d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2212fb36-b602-4e07-be2a-5892afb6ada6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=CDS:ENSP00000493376;Parent=transcript:ENST00000641515;protein_id=ENSP00000493376'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDSs[\"Tags\"][99]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "129fc67e-4dbf-4f56-ab08-50f83017e28b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
