{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e2750fd-1444-44ff-ad62-32c9c926fade",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib as mpl\n",
    "import copy\n",
    "from matplotlib.colors import LinearSegmentedColormap, ListedColormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9cb64c83-1bb4-4bd7-af2c-b23d77db8975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>size</th>\n",
       "      <th>cytoband</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>208887</td>\n",
       "      <td>209430</td>\n",
       "      <td>0.76</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000209;connected_gene=TBC1D...</td>\n",
       "      <td>543</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>226800</td>\n",
       "      <td>228800</td>\n",
       "      <td>0.71</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000227;connected_gene=DEFB1...</td>\n",
       "      <td>2000</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>266199</td>\n",
       "      <td>267649</td>\n",
       "      <td>1.72</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000266;connected_gene=ENSG0...</td>\n",
       "      <td>1450</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>279139</td>\n",
       "      <td>279609</td>\n",
       "      <td>0.51</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000280;connected_gene=C20or...</td>\n",
       "      <td>470</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>289399</td>\n",
       "      <td>292542</td>\n",
       "      <td>2.12</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000289;connected_gene=C20or...</td>\n",
       "      <td>3143</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419015</th>\n",
       "      <td>chr19</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>49645261</td>\n",
       "      <td>49646700</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH19J049645;connected_gene=HSALN...</td>\n",
       "      <td>1439</td>\n",
       "      <td>q13.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419016</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17222391</td>\n",
       "      <td>17224411</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017222;connected_gene=FLCN;...</td>\n",
       "      <td>2020</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419017</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17220946</td>\n",
       "      <td>17221921</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017220;connected_gene=NT5M;...</td>\n",
       "      <td>975</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419018</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17215651</td>\n",
       "      <td>17216481</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017216;connected_gene=FLCN;...</td>\n",
       "      <td>830</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419019</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>64315535</td>\n",
       "      <td>64319253</td>\n",
       "      <td>0.70</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J064315;connected_gene=HSALN...</td>\n",
       "      <td>3718</td>\n",
       "      <td>q13.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>419020 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chrom      source       feature name     start       end  score  \\\n",
       "0       chr20  GeneHancer           Enhancer    208887    209430   0.76   \n",
       "1       chr20  GeneHancer  Promoter/Enhancer    226800    228800   0.71   \n",
       "2       chr20  GeneHancer  Promoter/Enhancer    266199    267649   1.72   \n",
       "3       chr20  GeneHancer  Promoter/Enhancer    279139    279609   0.51   \n",
       "4       chr20  GeneHancer  Promoter/Enhancer    289399    292542   2.12   \n",
       "...       ...         ...                ...       ...       ...    ...   \n",
       "419015  chr19  GeneHancer           Enhancer  49645261  49646700   0.14   \n",
       "419016  chr17  GeneHancer           Enhancer  17222391  17224411   0.26   \n",
       "419017  chr17  GeneHancer           Enhancer  17220946  17221921   0.14   \n",
       "419018  chr17  GeneHancer           Enhancer  17215651  17216481   0.26   \n",
       "419019  chr20  GeneHancer           Enhancer  64315535  64319253   0.70   \n",
       "\n",
       "       strand frame                                         attributes  size  \\\n",
       "0           .     .  genehancer_id=GH20J000209;connected_gene=TBC1D...   543   \n",
       "1           .     .  genehancer_id=GH20J000227;connected_gene=DEFB1...  2000   \n",
       "2           .     .  genehancer_id=GH20J000266;connected_gene=ENSG0...  1450   \n",
       "3           .     .  genehancer_id=GH20J000280;connected_gene=C20or...   470   \n",
       "4           .     .  genehancer_id=GH20J000289;connected_gene=C20or...  3143   \n",
       "...       ...   ...                                                ...   ...   \n",
       "419015      .     .  genehancer_id=GH19J049645;connected_gene=HSALN...  1439   \n",
       "419016      .     .  genehancer_id=GH17J017222;connected_gene=FLCN;...  2020   \n",
       "419017      .     .  genehancer_id=GH17J017220;connected_gene=NT5M;...   975   \n",
       "419018      .     .  genehancer_id=GH17J017216;connected_gene=FLCN;...   830   \n",
       "419019      .     .  genehancer_id=GH20J064315;connected_gene=HSALN...  3718   \n",
       "\n",
       "       cytoband  \n",
       "0           p13  \n",
       "1           p13  \n",
       "2           p13  \n",
       "3           p13  \n",
       "4           p13  \n",
       "...         ...  \n",
       "419015   q13.33  \n",
       "419016    p11.2  \n",
       "419017    p11.2  \n",
       "419018    p11.2  \n",
       "419019   q13.33  \n",
       "\n",
       "[419020 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_genes_read = \"/home/androx/Documents/trabalho/datasets/Disease_genes/genehancer/GeneHancer_v5.20_processed.gff\"\n",
    "disease_genes_write =  \"/home/androx/Documents/trabalho/datasets/Disease_genes/genehancer/GeneHancer_v5.20_complete.gff\"\n",
    "dg_df = pd.read_csv(disease_genes_read, sep=\"\\t\")\n",
    "\n",
    "dg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8152b78-1884-4beb-bdd3-20f642bb62da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2552245"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_list = []\n",
    "for row in dg_df[\"attributes\"]:\n",
    "    row_elements = row.split(\";\")\n",
    "    for i in range (1, len(row_elements)-1, 2):\n",
    "        gene_symbol = row_elements[i].split(\"=\")[1]\n",
    "        genes_list.append(gene_symbol)\n",
    "\n",
    "len(genes_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "372b2400-34f8-4309-b48c-e3535b82ac27",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    string = \"\"\n",
    "    for gene_symbol in genes_list:\n",
    "        string += \" \" + gene_symbol\n",
    "    f.write(string)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f2cb19a4-717b-4c76-86cb-b8d2e93085c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symbols_path = \"/home/androx/Documents/trabalho/datasets/Disease_genes/genehancer/DGsymbols.txt\"\n",
    "writeNewtxt(gene_symbols_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6638bd05-791c-45f6-84ad-94a242b28024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>initial_alias</th>\n",
       "      <th>converted_alias</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>namespace</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TBC1D20</td>\n",
       "      <td>ENSG00000125875</td>\n",
       "      <td>TBC1D20</td>\n",
       "      <td>TBC1 domain family member 20 [Source:HGNC Symb...</td>\n",
       "      <td>ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PIR-34165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DEFB128</td>\n",
       "      <td>ENSG00000185982</td>\n",
       "      <td>DEFB128</td>\n",
       "      <td>defensin beta 128 [Source:HGNC Symbol;Acc:HGNC...</td>\n",
       "      <td>ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEFB129</td>\n",
       "      <td>ENSG00000125903</td>\n",
       "      <td>DEFB129</td>\n",
       "      <td>defensin beta 129 [Source:HGNC Symbol;Acc:HGNC...</td>\n",
       "      <td>ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HSALNG0128009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573611</th>\n",
       "      <td>HSALNG0131740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573612</th>\n",
       "      <td>PIR-39425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573613</th>\n",
       "      <td>HSALNG0131742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573614</th>\n",
       "      <td>HSALNG0131741</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573615</th>\n",
       "      <td>PCMTD2</td>\n",
       "      <td>ENSG00000203880</td>\n",
       "      <td>PCMTD2</td>\n",
       "      <td>protein-L-isoaspartate (D-aspartate) O-methylt...</td>\n",
       "      <td>ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2573616 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         initial_alias  converted_alias     name  \\\n",
       "0              TBC1D20  ENSG00000125875  TBC1D20   \n",
       "1            PIR-34165              NaN      NaN   \n",
       "2              DEFB128  ENSG00000185982  DEFB128   \n",
       "3              DEFB129  ENSG00000125903  DEFB129   \n",
       "4        HSALNG0128009              NaN      NaN   \n",
       "...                ...              ...      ...   \n",
       "2573611  HSALNG0131740              NaN      NaN   \n",
       "2573612      PIR-39425              NaN      NaN   \n",
       "2573613  HSALNG0131742              NaN      NaN   \n",
       "2573614  HSALNG0131741              NaN      NaN   \n",
       "2573615         PCMTD2  ENSG00000203880   PCMTD2   \n",
       "\n",
       "                                               description  \\\n",
       "0        TBC1 domain family member 20 [Source:HGNC Symb...   \n",
       "1                                                      NaN   \n",
       "2        defensin beta 128 [Source:HGNC Symbol;Acc:HGNC...   \n",
       "3        defensin beta 129 [Source:HGNC Symbol;Acc:HGNC...   \n",
       "4                                                      NaN   \n",
       "...                                                    ...   \n",
       "2573611                                                NaN   \n",
       "2573612                                                NaN   \n",
       "2573613                                                NaN   \n",
       "2573614                                                NaN   \n",
       "2573615  protein-L-isoaspartate (D-aspartate) O-methylt...   \n",
       "\n",
       "                                             namespace  \n",
       "0        ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE  \n",
       "1                                                  NaN  \n",
       "2        ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE  \n",
       "3        ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE  \n",
       "4                                                  NaN  \n",
       "...                                                ...  \n",
       "2573611                                            NaN  \n",
       "2573612                                            NaN  \n",
       "2573613                                            NaN  \n",
       "2573614                                            NaN  \n",
       "2573615  ENTREZGENE,GENECARDS,HGNC,UNIPROT_GN,WIKIGENE  \n",
       "\n",
       "[2573616 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idConvert_read= \"/home/androx/Documents/trabalho/datasets/Disease_genes/genehancer/gProfiler_hsapiens_01-07-2024_17-15-03.csv\"\n",
    "idConvert_df = pd.read_csv(idConvert_read, sep=\",\")\n",
    "\n",
    "idConvert_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "85495d88-1062-4ebc-be2a-97a9026fd5d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_newCol():\n",
    "    ENG_ids = []\n",
    "    converted_line_n = 0\n",
    "    for row in dg_df[\"attributes\"]:\n",
    "        ENG_ids_row = \"\"\n",
    "        for i in range (row.count(\"gene=\")):\n",
    "            dg = idConvert_df[\"converted_alias\"][converted_line_n]\n",
    "            ENG_ids_row += str(dg) + \", \"\n",
    "            converted_line_n += 1\n",
    "        ENG_ids.append(str(ENG_ids_row[:-2]))\n",
    "    return ENG_ids\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b641967c-bc40-42fb-99c7-1ddb293a1361",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    str_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        str_cols += col + \"\\t\"\n",
    "    str_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(str_cols))\n",
    "    for row in df.values:\n",
    "        str_row = \"\"\n",
    "        for col in row[:-1]:\n",
    "            str_row += str(col) + \"\\t\"\n",
    "        str_row += str(row[-1])+ \"\\n\"\n",
    "        f.write(str_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8bdafcfd-cd47-4d3a-9b5c-6d46b9c42524",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ENGids = create_newCol()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ee35a8ad-19af-44ae-b5d6-d944791ab003",
   "metadata": {},
   "outputs": [],
   "source": [
    "dg_df.insert(len(dg_df.columns), \"connected_genes_ENGids\", gene_ENGids, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "783bfdcc-29d4-421e-9e08-8d90a0d67656",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>size</th>\n",
       "      <th>cytoband</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>208887</td>\n",
       "      <td>209430</td>\n",
       "      <td>0.76</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000209;connected_gene=TBC1D...</td>\n",
       "      <td>543</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>226800</td>\n",
       "      <td>228800</td>\n",
       "      <td>0.71</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000227;connected_gene=DEFB1...</td>\n",
       "      <td>2000</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>266199</td>\n",
       "      <td>267649</td>\n",
       "      <td>1.72</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000266;connected_gene=ENSG0...</td>\n",
       "      <td>1450</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>279139</td>\n",
       "      <td>279609</td>\n",
       "      <td>0.51</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000280;connected_gene=C20or...</td>\n",
       "      <td>470</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>289399</td>\n",
       "      <td>292542</td>\n",
       "      <td>2.12</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000289;connected_gene=C20or...</td>\n",
       "      <td>3143</td>\n",
       "      <td>p13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419015</th>\n",
       "      <td>chr19</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>49645261</td>\n",
       "      <td>49646700</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH19J049645;connected_gene=HSALN...</td>\n",
       "      <td>1439</td>\n",
       "      <td>q13.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419016</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17222391</td>\n",
       "      <td>17224411</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017222;connected_gene=FLCN;...</td>\n",
       "      <td>2020</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419017</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17220946</td>\n",
       "      <td>17221921</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017220;connected_gene=NT5M;...</td>\n",
       "      <td>975</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419018</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17215651</td>\n",
       "      <td>17216481</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017216;connected_gene=FLCN;...</td>\n",
       "      <td>830</td>\n",
       "      <td>p11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419019</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>64315535</td>\n",
       "      <td>64319253</td>\n",
       "      <td>0.70</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J064315;connected_gene=HSALN...</td>\n",
       "      <td>3718</td>\n",
       "      <td>q13.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>419020 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chrom      source       feature name     start       end  score  \\\n",
       "0       chr20  GeneHancer           Enhancer    208887    209430   0.76   \n",
       "1       chr20  GeneHancer  Promoter/Enhancer    226800    228800   0.71   \n",
       "2       chr20  GeneHancer  Promoter/Enhancer    266199    267649   1.72   \n",
       "3       chr20  GeneHancer  Promoter/Enhancer    279139    279609   0.51   \n",
       "4       chr20  GeneHancer  Promoter/Enhancer    289399    292542   2.12   \n",
       "...       ...         ...                ...       ...       ...    ...   \n",
       "419015  chr19  GeneHancer           Enhancer  49645261  49646700   0.14   \n",
       "419016  chr17  GeneHancer           Enhancer  17222391  17224411   0.26   \n",
       "419017  chr17  GeneHancer           Enhancer  17220946  17221921   0.14   \n",
       "419018  chr17  GeneHancer           Enhancer  17215651  17216481   0.26   \n",
       "419019  chr20  GeneHancer           Enhancer  64315535  64319253   0.70   \n",
       "\n",
       "       strand frame                                         attributes  size  \\\n",
       "0           .     .  genehancer_id=GH20J000209;connected_gene=TBC1D...   543   \n",
       "1           .     .  genehancer_id=GH20J000227;connected_gene=DEFB1...  2000   \n",
       "2           .     .  genehancer_id=GH20J000266;connected_gene=ENSG0...  1450   \n",
       "3           .     .  genehancer_id=GH20J000280;connected_gene=C20or...   470   \n",
       "4           .     .  genehancer_id=GH20J000289;connected_gene=C20or...  3143   \n",
       "...       ...   ...                                                ...   ...   \n",
       "419015      .     .  genehancer_id=GH19J049645;connected_gene=HSALN...  1439   \n",
       "419016      .     .  genehancer_id=GH17J017222;connected_gene=FLCN;...  2020   \n",
       "419017      .     .  genehancer_id=GH17J017220;connected_gene=NT5M;...   975   \n",
       "419018      .     .  genehancer_id=GH17J017216;connected_gene=FLCN;...   830   \n",
       "419019      .     .  genehancer_id=GH20J064315;connected_gene=HSALN...  3718   \n",
       "\n",
       "       cytoband  \n",
       "0           p13  \n",
       "1           p13  \n",
       "2           p13  \n",
       "3           p13  \n",
       "4           p13  \n",
       "...         ...  \n",
       "419015   q13.33  \n",
       "419016    p11.2  \n",
       "419017    p11.2  \n",
       "419018    p11.2  \n",
       "419019   q13.33  \n",
       "\n",
       "[419020 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg_df.drop(columns=['connected_genes_ENGids'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "42efb8bc-af2f-4e0a-9cfc-0f4a9af4b0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>size</th>\n",
       "      <th>cytoband</th>\n",
       "      <th>connected_genes_ENGids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>208887</td>\n",
       "      <td>209430</td>\n",
       "      <td>0.76</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000209;connected_gene=TBC1D...</td>\n",
       "      <td>543</td>\n",
       "      <td>p13</td>\n",
       "      <td>ENSG00000125875, nan, ENSG00000185982, ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>226800</td>\n",
       "      <td>228800</td>\n",
       "      <td>0.71</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000227;connected_gene=DEFB1...</td>\n",
       "      <td>2000</td>\n",
       "      <td>p13</td>\n",
       "      <td>ENSG00000125903, nan, ENSG00000186458, ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>266199</td>\n",
       "      <td>267649</td>\n",
       "      <td>1.72</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000266;connected_gene=ENSG0...</td>\n",
       "      <td>1450</td>\n",
       "      <td>p13</td>\n",
       "      <td>ENSG00000272874, nan, ENSG00000247315, ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>279139</td>\n",
       "      <td>279609</td>\n",
       "      <td>0.51</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000280;connected_gene=C20or...</td>\n",
       "      <td>470</td>\n",
       "      <td>p13</td>\n",
       "      <td>ENSG00000196476, nan, ENSG00000186458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Promoter/Enhancer</td>\n",
       "      <td>289399</td>\n",
       "      <td>292542</td>\n",
       "      <td>2.12</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J000289;connected_gene=C20or...</td>\n",
       "      <td>3143</td>\n",
       "      <td>p13</td>\n",
       "      <td>ENSG00000196476, ENSG00000215388, ENSG00000125...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419015</th>\n",
       "      <td>chr19</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>49645261</td>\n",
       "      <td>49646700</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH19J049645;connected_gene=HSALN...</td>\n",
       "      <td>1439</td>\n",
       "      <td>q13.33</td>\n",
       "      <td>ENSG00000170185, ENSG00000250326, ENSG00000293...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419016</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17222391</td>\n",
       "      <td>17224411</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017222;connected_gene=FLCN;...</td>\n",
       "      <td>2020</td>\n",
       "      <td>p11.2</td>\n",
       "      <td>ENSG00000153147, nan, nan, ENSG00000184144, EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419017</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17220946</td>\n",
       "      <td>17221921</td>\n",
       "      <td>0.14</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017220;connected_gene=NT5M;...</td>\n",
       "      <td>975</td>\n",
       "      <td>p11.2</td>\n",
       "      <td>ENSG00000276691, nan, ENSG00000061273, ENSG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419018</th>\n",
       "      <td>chr17</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>17215651</td>\n",
       "      <td>17216481</td>\n",
       "      <td>0.26</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH17J017216;connected_gene=FLCN;...</td>\n",
       "      <td>830</td>\n",
       "      <td>p11.2</td>\n",
       "      <td>ENSG00000268069, ENSG00000079337, ENSG00000229...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419019</th>\n",
       "      <td>chr20</td>\n",
       "      <td>GeneHancer</td>\n",
       "      <td>Enhancer</td>\n",
       "      <td>64315535</td>\n",
       "      <td>64319253</td>\n",
       "      <td>0.70</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>genehancer_id=GH20J064315;connected_gene=HSALN...</td>\n",
       "      <td>3718</td>\n",
       "      <td>q13.33</td>\n",
       "      <td>ENSG00000135972, nan, ENSG00000163468, ENSG000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>419020 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chrom      source       feature name     start       end  score  \\\n",
       "0       chr20  GeneHancer           Enhancer    208887    209430   0.76   \n",
       "1       chr20  GeneHancer  Promoter/Enhancer    226800    228800   0.71   \n",
       "2       chr20  GeneHancer  Promoter/Enhancer    266199    267649   1.72   \n",
       "3       chr20  GeneHancer  Promoter/Enhancer    279139    279609   0.51   \n",
       "4       chr20  GeneHancer  Promoter/Enhancer    289399    292542   2.12   \n",
       "...       ...         ...                ...       ...       ...    ...   \n",
       "419015  chr19  GeneHancer           Enhancer  49645261  49646700   0.14   \n",
       "419016  chr17  GeneHancer           Enhancer  17222391  17224411   0.26   \n",
       "419017  chr17  GeneHancer           Enhancer  17220946  17221921   0.14   \n",
       "419018  chr17  GeneHancer           Enhancer  17215651  17216481   0.26   \n",
       "419019  chr20  GeneHancer           Enhancer  64315535  64319253   0.70   \n",
       "\n",
       "       strand frame                                         attributes  size  \\\n",
       "0           .     .  genehancer_id=GH20J000209;connected_gene=TBC1D...   543   \n",
       "1           .     .  genehancer_id=GH20J000227;connected_gene=DEFB1...  2000   \n",
       "2           .     .  genehancer_id=GH20J000266;connected_gene=ENSG0...  1450   \n",
       "3           .     .  genehancer_id=GH20J000280;connected_gene=C20or...   470   \n",
       "4           .     .  genehancer_id=GH20J000289;connected_gene=C20or...  3143   \n",
       "...       ...   ...                                                ...   ...   \n",
       "419015      .     .  genehancer_id=GH19J049645;connected_gene=HSALN...  1439   \n",
       "419016      .     .  genehancer_id=GH17J017222;connected_gene=FLCN;...  2020   \n",
       "419017      .     .  genehancer_id=GH17J017220;connected_gene=NT5M;...   975   \n",
       "419018      .     .  genehancer_id=GH17J017216;connected_gene=FLCN;...   830   \n",
       "419019      .     .  genehancer_id=GH20J064315;connected_gene=HSALN...  3718   \n",
       "\n",
       "       cytoband                             connected_genes_ENGids  \n",
       "0           p13  ENSG00000125875, nan, ENSG00000185982, ENSG000...  \n",
       "1           p13  ENSG00000125903, nan, ENSG00000186458, ENSG000...  \n",
       "2           p13  ENSG00000272874, nan, ENSG00000247315, ENSG000...  \n",
       "3           p13              ENSG00000196476, nan, ENSG00000186458  \n",
       "4           p13  ENSG00000196476, ENSG00000215388, ENSG00000125...  \n",
       "...         ...                                                ...  \n",
       "419015   q13.33  ENSG00000170185, ENSG00000250326, ENSG00000293...  \n",
       "419016    p11.2  ENSG00000153147, nan, nan, ENSG00000184144, EN...  \n",
       "419017    p11.2  ENSG00000276691, nan, ENSG00000061273, ENSG000...  \n",
       "419018    p11.2  ENSG00000268069, ENSG00000079337, ENSG00000229...  \n",
       "419019   q13.33  ENSG00000135972, nan, ENSG00000163468, ENSG000...  \n",
       "\n",
       "[419020 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f387b629-4d4d-43c5-bb6c-987a4ff7865c",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(disease_genes_write, dg_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8cf90f0a-7dbf-4fc8-8fc9-2bb9d833ca8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0       1                  2           3            4        5  6  7  \\\n",
       "0        1  GRCh38         chromosome         1.0  248956422.0        .  .  .   \n",
       "1        1       .  biological_region     10469.0      11240.0  1.3e+03  .  .   \n",
       "2        1       .  biological_region     10650.0      10657.0    0.999  +  .   \n",
       "3        1       .  biological_region     10655.0      10657.0    0.999  -  .   \n",
       "4        1       .  biological_region     10678.0      10687.0    0.999  +  .   \n",
       "...     ..     ...                ...         ...          ...      ... .. ..   \n",
       "3516515  Y       .  biological_region  57215804.0   57215813.0    0.999  -  .   \n",
       "3516516  Y       .  biological_region  57215832.0   57215840.0    0.999  +  .   \n",
       "3516517  Y       .  biological_region  57215833.0   57215842.0    0.999  -  .   \n",
       "3516518  Y       .  biological_region  57215862.0   57215867.0    0.999  +  .   \n",
       "3516519  Y       .  biological_region  57215862.0   57215870.0    0.999  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1                 external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2                                       logic_name=eponine  \n",
       "3                                       logic_name=eponine  \n",
       "4                                       logic_name=eponine  \n",
       "...                                                    ...  \n",
       "3516515                                 logic_name=eponine  \n",
       "3516516                                 logic_name=eponine  \n",
       "3516517                                 logic_name=eponine  \n",
       "3516518                                 logic_name=eponine  \n",
       "3516519                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_read= \"/home/androx/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens.GRCh38.112.chr.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57ae9efc-c125-42c9-9e6d-82f419f13797",
   "metadata": {},
   "source": [
    "#remover todas as linhas com \"#\" do dataset NAO necssario"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9058d4cd-22c4-42ff-acfb-ec397a50065d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3516520 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3516515   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3516516   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3516517   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3516518   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3516519   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Unknown1 Unknown2                                        Identifiers  \n",
       "0              .        .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1              .        .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2              +        .                                 logic_name=eponine  \n",
       "3              -        .                                 logic_name=eponine  \n",
       "4              +        .                                 logic_name=eponine  \n",
       "...          ...      ...                                                ...  \n",
       "3516515        -        .                                 logic_name=eponine  \n",
       "3516516        +        .                                 logic_name=eponine  \n",
       "3516517        -        .                                 logic_name=eponine  \n",
       "3516518        +        .                                 logic_name=eponine  \n",
       "3516519        -        .                                 logic_name=eponine  \n",
       "\n",
       "[3516520 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Unknown1', 7: 'Unknown2', 8: 'Identifiers'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2dd7b6fc-202a-4eec-b301-3594ab9ef9e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "21613be0-5d59-4718-ac31-44193c80a751",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328201</th>\n",
       "      <td>10</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133797422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467767</th>\n",
       "      <td>11</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>135086622.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673000</th>\n",
       "      <td>12</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133275309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862634</th>\n",
       "      <td>13</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>114364328.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924501</th>\n",
       "      <td>14</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>107043718.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048326</th>\n",
       "      <td>15</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101991189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171301</th>\n",
       "      <td>16</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90338345.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329042</th>\n",
       "      <td>17</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83257441.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536644</th>\n",
       "      <td>18</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80373285.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598313</th>\n",
       "      <td>19</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58617616.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791801</th>\n",
       "      <td>2</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242193529.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063156</th>\n",
       "      <td>20</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64444167.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145930</th>\n",
       "      <td>21</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46709983.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186784</th>\n",
       "      <td>22</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50818468.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2261287</th>\n",
       "      <td>3</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>198295559.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484781</th>\n",
       "      <td>4</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>190214555.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628956</th>\n",
       "      <td>5</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>181538259.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786252</th>\n",
       "      <td>6</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170805979.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2952608</th>\n",
       "      <td>7</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>159345973.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3120542</th>\n",
       "      <td>8</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145138636.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248106</th>\n",
       "      <td>9</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>138394717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384196</th>\n",
       "      <td>MT</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384360</th>\n",
       "      <td>X</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>156040895.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3505822</th>\n",
       "      <td>Y</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>57227415.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index   Last_index Value Unknown1  \\\n",
       "0         1  GRCh38  chromosome          1.0  248956422.0     .        .   \n",
       "328201   10  GRCh38  chromosome          1.0  133797422.0     .        .   \n",
       "467767   11  GRCh38  chromosome          1.0  135086622.0     .        .   \n",
       "673000   12  GRCh38  chromosome          1.0  133275309.0     .        .   \n",
       "862634   13  GRCh38  chromosome          1.0  114364328.0     .        .   \n",
       "924501   14  GRCh38  chromosome          1.0  107043718.0     .        .   \n",
       "1048326  15  GRCh38  chromosome          1.0  101991189.0     .        .   \n",
       "1171301  16  GRCh38  chromosome          1.0   90338345.0     .        .   \n",
       "1329042  17  GRCh38  chromosome          1.0   83257441.0     .        .   \n",
       "1536644  18  GRCh38  chromosome          1.0   80373285.0     .        .   \n",
       "1598313  19  GRCh38  chromosome          1.0   58617616.0     .        .   \n",
       "1791801   2  GRCh38  chromosome          1.0  242193529.0     .        .   \n",
       "2063156  20  GRCh38  chromosome          1.0   64444167.0     .        .   \n",
       "2145930  21  GRCh38  chromosome          1.0   46709983.0     .        .   \n",
       "2186784  22  GRCh38  chromosome          1.0   50818468.0     .        .   \n",
       "2261287   3  GRCh38  chromosome          1.0  198295559.0     .        .   \n",
       "2484781   4  GRCh38  chromosome          1.0  190214555.0     .        .   \n",
       "2628956   5  GRCh38  chromosome          1.0  181538259.0     .        .   \n",
       "2786252   6  GRCh38  chromosome          1.0  170805979.0     .        .   \n",
       "2952608   7  GRCh38  chromosome          1.0  159345973.0     .        .   \n",
       "3120542   8  GRCh38  chromosome          1.0  145138636.0     .        .   \n",
       "3248106   9  GRCh38  chromosome          1.0  138394717.0     .        .   \n",
       "3384196  MT  GRCh38  chromosome          1.0      16569.0     .        .   \n",
       "3384360   X  GRCh38  chromosome          1.0  156040895.0     .        .   \n",
       "3505822   Y  GRCh38  chromosome          1.0   57227415.0     .        .   \n",
       "\n",
       "        Unknown2                                        Identifiers  \n",
       "0              .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "328201         .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  \n",
       "467767         .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  \n",
       "673000         .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  \n",
       "862634         .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  \n",
       "924501         .  ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...  \n",
       "1048326        .  ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...  \n",
       "1171301        .  ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...  \n",
       "1329042        .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...  \n",
       "1536644        .  ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...  \n",
       "1598313        .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...  \n",
       "1791801        .  ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...  \n",
       "2063156        .  ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...  \n",
       "2145930        .  ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...  \n",
       "2186784        .  ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...  \n",
       "2261287        .  ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...  \n",
       "2484781        .  ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...  \n",
       "2628956        .  ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...  \n",
       "2786252        .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...  \n",
       "2952608        .  ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...  \n",
       "3120542        .  ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...  \n",
       "3248106        .  ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...  \n",
       "3384196        .   ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1  \n",
       "3384360        .  ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...  \n",
       "3505822        .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosmes = tanscripts_df.loc[tanscripts_df['Type'] == \"chromosome\"]\n",
    "chromosmes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8ad98867-40bb-42d2-af17-609cb6c07bdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944575.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3515710</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25062548.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3515921</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516228</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57067865.0</td>\n",
       "      <td>57130289.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516327</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57184216.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516392</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57207346.0</td>\n",
       "      <td>57212230.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21547 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Unknown1  \\\n",
       "93        1  ensembl_havana  gene      65419.0     71585.0     .        +   \n",
       "395       1  ensembl_havana  gene     450740.0    451678.0     .        -   \n",
       "541       1  ensembl_havana  gene     685716.0    686654.0     .        -   \n",
       "1161      1  ensembl_havana  gene     923923.0    944575.0     .        +   \n",
       "1524      1  ensembl_havana  gene     944203.0    959309.0     .        -   \n",
       "...      ..             ...   ...          ...         ...   ...      ...   \n",
       "3515710   Y  ensembl_havana  gene   25030901.0  25062548.0     .        -   \n",
       "3515921   Y  ensembl_havana  gene   25622117.0  25624902.0     .        +   \n",
       "3516228   Y  ensembl_havana  gene   57067865.0  57130289.0     .        +   \n",
       "3516327   Y  ensembl_havana  gene   57184216.0  57197337.0     .        +   \n",
       "3516392   Y  ensembl_havana  gene   57207346.0  57212230.0     .        +   \n",
       "\n",
       "        Unknown2                                        Identifiers  \n",
       "93             .  ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...  \n",
       "395            .  ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...  \n",
       "541            .  ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...  \n",
       "1161           .  ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...  \n",
       "1524           .  ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...  \n",
       "...          ...                                                ...  \n",
       "3515710        .  ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...  \n",
       "3515921        .  ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...  \n",
       "3516228        .  ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...  \n",
       "3516327        .  ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...  \n",
       "3516392        .  ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...  \n",
       "\n",
       "[21547 rows x 9 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = tanscripts_df.loc[tanscripts_df['Type'] == \"gene\"]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d98a9c35-2ba2-4ef7-a2de-9de93a6d5b03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "93         ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...\n",
       "395        ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...\n",
       "541        ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...\n",
       "1161       ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...\n",
       "1524       ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...\n",
       "                                 ...                        \n",
       "3515710    ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...\n",
       "3515921    ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...\n",
       "3516228    ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...\n",
       "3516327    ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...\n",
       "3516392    ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...\n",
       "Name: Identifiers, Length: 21547, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[\"Identifiers\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "94d0ae5d-1f64-4445-9cc9-1226fdb22e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565.0</td>\n",
       "      <td>65573.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037.0</td>\n",
       "      <td>70008.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000493376;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000409316;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000329982;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1165</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432.0</td>\n",
       "      <td>924948.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000478421;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516411</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209532.0</td>\n",
       "      <td>57209733.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516413</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57209822.0</td>\n",
       "      <td>57209980.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516415</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57210640.0</td>\n",
       "      <td>57210792.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516417</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211552.0</td>\n",
       "      <td>57211620.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516419</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57211761.0</td>\n",
       "      <td>57211894.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=CDS:ENSP00000518618;Parent=transcript:ENST0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900145 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome Type  First_index  Last_index Value Unknown1  \\\n",
       "99        1          havana  CDS      65565.0     65573.0     .        +   \n",
       "100       1          havana  CDS      69037.0     70008.0     .        +   \n",
       "398       1  ensembl_havana  CDS     450740.0    451678.0     .        -   \n",
       "544       1  ensembl_havana  CDS     685716.0    686654.0     .        -   \n",
       "1165      1          havana  CDS     924432.0    924948.0     .        +   \n",
       "...      ..             ...  ...          ...         ...   ...      ...   \n",
       "3516411   Y          havana  CDS   57209532.0  57209733.0     .        +   \n",
       "3516413   Y          havana  CDS   57209822.0  57209980.0     .        +   \n",
       "3516415   Y          havana  CDS   57210640.0  57210792.0     .        +   \n",
       "3516417   Y          havana  CDS   57211552.0  57211620.0     .        +   \n",
       "3516419   Y          havana  CDS   57211761.0  57211894.0     .        +   \n",
       "\n",
       "        Unknown2                                        Identifiers  \n",
       "99             0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "100            0  ID=CDS:ENSP00000493376;Parent=transcript:ENST0...  \n",
       "398            0  ID=CDS:ENSP00000409316;Parent=transcript:ENST0...  \n",
       "544            0  ID=CDS:ENSP00000329982;Parent=transcript:ENST0...  \n",
       "1165           0  ID=CDS:ENSP00000478421;Parent=transcript:ENST0...  \n",
       "...          ...                                                ...  \n",
       "3516411        0  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516413        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516415        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516417        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "3516419        2  ID=CDS:ENSP00000518618;Parent=transcript:ENST0...  \n",
       "\n",
       "[900145 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDSs = tanscripts_df.loc[tanscripts_df['Type'] == \"CDS\"]\n",
    "CDSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "707bb31a-7f65-4228-a706-cd1523c2cad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516493</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212184.0</td>\n",
       "      <td>57213125.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516494</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213204.0</td>\n",
       "      <td>57213357.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516495</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213526.0</td>\n",
       "      <td>57213602.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516496</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213880.0</td>\n",
       "      <td>57213964.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516497</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214350.0</td>\n",
       "      <td>57214397.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1668627 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome  Type  First_index  Last_index Value Unknown1 Unknown2  \\\n",
       "18        1  havana  exon      11869.0     12227.0     .        +        .   \n",
       "19        1  havana  exon      12613.0     12721.0     .        +        .   \n",
       "20        1  havana  exon      13221.0     14409.0     .        +        .   \n",
       "24        1  havana  exon      12010.0     12057.0     .        +        .   \n",
       "25        1  havana  exon      12179.0     12227.0     .        +        .   \n",
       "...      ..     ...   ...          ...         ...   ...      ...      ...   \n",
       "3516493   Y  havana  exon   57212184.0  57213125.0     .        -        .   \n",
       "3516494   Y  havana  exon   57213204.0  57213357.0     .        -        .   \n",
       "3516495   Y  havana  exon   57213526.0  57213602.0     .        -        .   \n",
       "3516496   Y  havana  exon   57213880.0  57213964.0     .        -        .   \n",
       "3516497   Y  havana  exon   57214350.0  57214397.0     .        -        .   \n",
       "\n",
       "                                               Identifiers  \n",
       "18       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "19       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "20       Parent=transcript:ENST00000456328;Name=ENSE000...  \n",
       "24       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "25       Parent=transcript:ENST00000450305;Name=ENSE000...  \n",
       "...                                                    ...  \n",
       "3516493  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516494  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516495  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516496  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "3516497  Parent=transcript:ENST00000711270;Name=ENSE000...  \n",
       "\n",
       "[1668627 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = tanscripts_df.loc[tanscripts_df['Type'] == \"exon\"]\n",
    "exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "59fe2151-79d7-4387-a213-45b9b6091fab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Unknown1</th>\n",
       "      <th>Unknown2</th>\n",
       "      <th>Identifiers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10681.0</td>\n",
       "      <td>10688.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516515</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516516</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516517</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516518</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3516519</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180392 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr Genome               Type  First_index  Last_index    Value  \\\n",
       "1         1      .  biological_region      10469.0     11240.0  1.3e+03   \n",
       "2         1      .  biological_region      10650.0     10657.0    0.999   \n",
       "3         1      .  biological_region      10655.0     10657.0    0.999   \n",
       "4         1      .  biological_region      10678.0     10687.0    0.999   \n",
       "5         1      .  biological_region      10681.0     10688.0    0.999   \n",
       "...      ..    ...                ...          ...         ...      ...   \n",
       "3516515   Y      .  biological_region   57215804.0  57215813.0    0.999   \n",
       "3516516   Y      .  biological_region   57215832.0  57215840.0    0.999   \n",
       "3516517   Y      .  biological_region   57215833.0  57215842.0    0.999   \n",
       "3516518   Y      .  biological_region   57215862.0  57215867.0    0.999   \n",
       "3516519   Y      .  biological_region   57215862.0  57215870.0    0.999   \n",
       "\n",
       "        Unknown1 Unknown2                               Identifiers  \n",
       "1              .        .  external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2              +        .                        logic_name=eponine  \n",
       "3              -        .                        logic_name=eponine  \n",
       "4              +        .                        logic_name=eponine  \n",
       "5              -        .                        logic_name=eponine  \n",
       "...          ...      ...                                       ...  \n",
       "3516515        -        .                        logic_name=eponine  \n",
       "3516516        +        .                        logic_name=eponine  \n",
       "3516517        -        .                        logic_name=eponine  \n",
       "3516518        +        .                        logic_name=eponine  \n",
       "3516519        -        .                        logic_name=eponine  \n",
       "\n",
       "[180392 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = tanscripts_df.loc[tanscripts_df['Type'] == \"biological_region\"]\n",
    "regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b6f379f9-7c46-4b22-8e2e-f2d0462c4ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0e444e03-75bc-4cfb-8c26-65b782a35b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nao é util para já agregar todos od conjuntos, mais apropriado fazer um mapa de cada coisa\n",
    "\n",
    "def calc_exons_introns(chromosomes_df, sequence_df, exons_df, genes_df, regions_df):\n",
    "    \"\"\"interval_df = Genes or CDSs\n",
    "    +2⁰ = 1: is exon\n",
    "    +2¹ = 2: is part of a region\n",
    "    +2² = 4: is part of CDS\n",
    "    +2³ = 8: is part of gene\n",
    "    \"\"\"\n",
    "    Chrmosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chromosome_elements = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chromosome_exons = exons_df.loc[exons_df['Chr'] == chromosome]\n",
    "        chr_genes = genes_df.loc[genes_df['Chr'] == chromosome]\n",
    "        chr_regions = regions_df.loc[regions_df['Chr'] == chromosome]\n",
    "        element_introns = np.zeros((chr_size), dtype=int)\n",
    "        #genes region\n",
    "        for row in chr_genes.index:\n",
    "            for i in range (int(chr_genes[\"First_index\"][row]), int(chr_genes[\"Last_index\"][row])):\n",
    "                element_introns[i]+=8\n",
    "        #CDS region\n",
    "        for row in chromosome_elements.index:\n",
    "            for i in range (int(chromosome_elements[\"First_index\"][row]), int(chromosome_elements[\"Last_index\"][row])):\n",
    "                element_introns[i]+=4\n",
    "        #other regions\n",
    "        for row in chr_regions.index:\n",
    "            for i in range (int(chr_regions[\"First_index\"][row]), int(chr_regions[\"Last_index\"][row])):\n",
    "                element_introns[i]+=2\n",
    "        #exons region\n",
    "        for row in chromosome_exons.index:\n",
    "            for i in range (int(chromosome_exons[\"First_index\"][row]), int(chromosome_exons[\"Last_index\"][row])):\n",
    "                element_introns[i]+=1\n",
    "        \n",
    "        Chrmosomes_matrix.append(element_introns)\n",
    "    return Chrmosomes_matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2137d3c0-cade-4228-982a-3961802f04cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_group(chromosomes_df, sequence_df):\n",
    "    \"\"\"\n",
    "    sequence_df = Genes or CDSs, exons or other_regions\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_sequence = sequence_df.loc[sequence_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        for row in chr_sequence.index:\n",
    "            for i in range (int(chr_sequence[\"First_index\"][row]), int(chr_sequence[\"Last_index\"][row])):\n",
    "                chr_array[i]+=1\n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3dfa3b89-ab5d-4b62-8df5-22fefd903785",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0])]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_matrix = calc_group(chromosmes, exons)\n",
    "exons_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e80393b4-abc9-485a-bab3-5eb66a750207",
   "metadata": {},
   "outputs": [],
   "source": [
    "#otherRegions_matrix = calc_group(chromosmes, regions)\n",
    "#otherRegions_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8df025bc-68de-4b73-b378-929666676ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_introns(chromosomes_df, CDS_df, exons_df):\n",
    "    \"\"\"\n",
    "    sequence_df = Genes or CDSs, exons or other_regions\n",
    "    if CDS element: +2\n",
    "    if exon: +1\n",
    "    introns = 2\n",
    "    exons = 3\n",
    "    \"\"\"\n",
    "    Chromosomes_matrix = []\n",
    "    for line in chromosomes_df.index:\n",
    "        chromosome = chromosomes_df[\"Chr\"][line]\n",
    "        print(\"searching citobands in chromossome: \", chromosome)\n",
    "        chr_size = int(chromosomes_df[\"Last_index\"][line])\n",
    "        chr_CDS = CDS_df.loc[CDS_df['Chr'] == chromosome]\n",
    "        chromosome_exons = exons_df.loc[exons_df['Chr'] == chromosome]\n",
    "        chr_array = np.zeros((chr_size), dtype=int)\n",
    "        #elements region\n",
    "        #CDS region\n",
    "        for row in chr_CDS.index:\n",
    "            for i in range (int(chr_CDS[\"First_index\"][row]), int(chr_CDS[\"Last_index\"][row])):\n",
    "                chr_array[i]+=2\n",
    "        #exons region\n",
    "        for row in chromosome_exons.index:\n",
    "            for i in range (int(chromosome_exons[\"First_index\"][row]), int(chromosome_exons[\"Last_index\"][row])):\n",
    "                chr_array[i]+=1\n",
    "        \n",
    "        Chromosomes_matrix.append(chr_array)\n",
    "    return Chromosomes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9948cf7b-ed3c-44ea-8e84-06f5ecc0bbeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching citobands in chromossome:  1\n",
      "searching citobands in chromossome:  10\n",
      "searching citobands in chromossome:  11\n",
      "searching citobands in chromossome:  12\n",
      "searching citobands in chromossome:  13\n",
      "searching citobands in chromossome:  14\n",
      "searching citobands in chromossome:  15\n",
      "searching citobands in chromossome:  16\n",
      "searching citobands in chromossome:  17\n",
      "searching citobands in chromossome:  18\n",
      "searching citobands in chromossome:  19\n",
      "searching citobands in chromossome:  2\n",
      "searching citobands in chromossome:  20\n",
      "searching citobands in chromossome:  21\n",
      "searching citobands in chromossome:  22\n",
      "searching citobands in chromossome:  3\n",
      "searching citobands in chromossome:  4\n",
      "searching citobands in chromossome:  5\n",
      "searching citobands in chromossome:  6\n",
      "searching citobands in chromossome:  7\n",
      "searching citobands in chromossome:  8\n",
      "searching citobands in chromossome:  9\n",
      "searching citobands in chromossome:  MT\n",
      "searching citobands in chromossome:  X\n",
      "searching citobands in chromossome:  Y\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0]),\n",
       " array([0, 0, 0, ..., 0, 0, 0])]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns_matrix = calc_introns(chromosmes, CDSs, exons)\n",
    "introns_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e8f6e29-a64d-4c6d-88a2-7727c0ab6baa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
