{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e63572f-9367-4ba6-b9c0-c58ecbd9634a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a73f5f4-a487-41fc-be9c-6ef6e9b69f1d",
   "metadata": {},
   "source": [
    "## Studies of the transcripts datafile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34b10306-22bc-48fb-8574-2c7975349522",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_read = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens_GRCh38_112.gff3\"\n",
    "tanscripts_df = pd.read_csv(tanscripts_read, sep=\"\\t\", header=None)\n",
    "#tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "72f884b7-6777-43a5-adb4-9ffd074ab759",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519861</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519862</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519863</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519864</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519865</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3519866 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3519861   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3519862   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3519863   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3519864   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3519865   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                               Tags  \n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +       .                                 logic_name=eponine  \n",
       "3            -       .                                 logic_name=eponine  \n",
       "4            +       .                                 logic_name=eponine  \n",
       "...        ...     ...                                                ...  \n",
       "3519861      -       .                                 logic_name=eponine  \n",
       "3519862      +       .                                 logic_name=eponine  \n",
       "3519863      -       .                                 logic_name=eponine  \n",
       "3519864      +       .                                 logic_name=eponine  \n",
       "3519865      -       .                                 logic_name=eponine  \n",
       "\n",
       "[3519866 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df = tanscripts_df.rename(columns={0: 'Chr', 1: 'Genome', 2: 'Type', 3: 'First_index', 4: 'Last_index', 5: 'Value', 6: 'Signal', 7: 'Unknown', 8: 'Tags'})\n",
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94d21770-0525-4163-afe9-dfa3bc4ff8c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon',\n",
       "       nan, 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene',\n",
       "       'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA',\n",
       "       'transcript', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA',\n",
       "       'processed_transcript', 'V_gene_segment', 'D_gene_segment',\n",
       "       'J_gene_segment', 'C_gene_segment', 'scaffold', 'tRNA'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dfaf4b9c-d367-4733-be02-fa343cf91f37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>940346.0</td>\n",
       "      <td>942173.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000478729;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>945319.0</td>\n",
       "      <td>945562.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000496938;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1593</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>954426.0</td>\n",
       "      <td>959309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000487214;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2582</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1082146.0</td>\n",
       "      <td>1084072.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000464905;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2652</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1083926.0</td>\n",
       "      <td>1116243.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000482816;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519720</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57192634.0</td>\n",
       "      <td>57194566.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711289;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519739</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57207346.0</td>\n",
       "      <td>57208935.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711271;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519774</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57208197.0</td>\n",
       "      <td>57209855.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711275;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519809</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57208519.0</td>\n",
       "      <td>57209909.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711280;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519831</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57210591.0</td>\n",
       "      <td>57212074.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000711285;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28741 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index  Last_index Value Signal Unknown  \\\n",
       "1432      1  havana  transcript     940346.0    942173.0     .      +       .   \n",
       "1590      1  havana  transcript     945319.0    945562.0     .      -       .   \n",
       "1593      1  havana  transcript     954426.0    959309.0     .      -       .   \n",
       "2582      1  havana  transcript    1082146.0   1084072.0     .      -       .   \n",
       "2652      1  havana  transcript    1083926.0   1116243.0     .      -       .   \n",
       "...      ..     ...         ...          ...         ...   ...    ...     ...   \n",
       "3519720   Y  havana  transcript   57192634.0  57194566.0     .      +       .   \n",
       "3519739   Y  havana  transcript   57207346.0  57208935.0     .      +       .   \n",
       "3519774   Y  havana  transcript   57208197.0  57209855.0     .      +       .   \n",
       "3519809   Y  havana  transcript   57208519.0  57209909.0     .      +       .   \n",
       "3519831   Y  havana  transcript   57210591.0  57212074.0     .      +       .   \n",
       "\n",
       "                                                      Tags  \n",
       "1432     ID=transcript:ENST00000478729;Parent=gene:ENSG...  \n",
       "1590     ID=transcript:ENST00000496938;Parent=gene:ENSG...  \n",
       "1593     ID=transcript:ENST00000487214;Parent=gene:ENSG...  \n",
       "2582     ID=transcript:ENST00000464905;Parent=gene:ENSG...  \n",
       "2652     ID=transcript:ENST00000482816;Parent=gene:ENSG...  \n",
       "...                                                    ...  \n",
       "3519720  ID=transcript:ENST00000711289;Parent=gene:ENSG...  \n",
       "3519739  ID=transcript:ENST00000711271;Parent=gene:ENSG...  \n",
       "3519774  ID=transcript:ENST00000711275;Parent=gene:ENSG...  \n",
       "3519809  ID=transcript:ENST00000711280;Parent=gene:ENSG...  \n",
       "3519831  ID=transcript:ENST00000711285;Parent=gene:ENSG...  \n",
       "\n",
       "[28741 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[tanscripts_df[\"Type\"] == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "743856aa-8245-4174-8b81-7114dd63fcd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944575.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519056</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25062548.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519267</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519574</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57067865.0</td>\n",
       "      <td>57130289.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519673</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57184216.0</td>\n",
       "      <td>57197337.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519738</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>57207346.0</td>\n",
       "      <td>57212230.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21571 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "93        1  ensembl_havana  gene      65419.0     71585.0     .      +   \n",
       "395       1  ensembl_havana  gene     450740.0    451678.0     .      -   \n",
       "541       1  ensembl_havana  gene     685716.0    686654.0     .      -   \n",
       "1161      1  ensembl_havana  gene     923923.0    944575.0     .      +   \n",
       "1524      1  ensembl_havana  gene     944203.0    959309.0     .      -   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3519056   Y  ensembl_havana  gene   25030901.0  25062548.0     .      -   \n",
       "3519267   Y  ensembl_havana  gene   25622117.0  25624902.0     .      +   \n",
       "3519574   Y  ensembl_havana  gene   57067865.0  57130289.0     .      +   \n",
       "3519673   Y  ensembl_havana  gene   57184216.0  57197337.0     .      +   \n",
       "3519738   Y  ensembl_havana  gene   57207346.0  57212230.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \n",
       "93            .  ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...  \n",
       "395           .  ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...  \n",
       "541           .  ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...  \n",
       "1161          .  ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...  \n",
       "1524          .  ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...  \n",
       "...         ...                                                ...  \n",
       "3519056       .  ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...  \n",
       "3519267       .  ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...  \n",
       "3519574       .  ID=gene:ENSG00000292366;Name=VAMP7;biotype=pro...  \n",
       "3519673       .  ID=gene:ENSG00000292373;Name=IL9R;biotype=prot...  \n",
       "3519738       .  ID=gene:ENSG00000292372;Name=WASH6P;biotype=pr...  \n",
       "\n",
       "[21571 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[tanscripts_df[\"Type\"] == 'gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b15f11b-1261-4a62-828f-ee2f22b4ebbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['logic_name=eponine']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][3].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1bbd2430-4208-43d9-a382-3dc9cd0f6c5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000478729',\n",
       " 'Parent=gene:ENSG00000187634',\n",
       " 'Name=SAMD11-208',\n",
       " 'biotype=protein_coding_CDS_not_defined',\n",
       " 'transcript_id=ENST00000478729',\n",
       " 'transcript_support_level=5',\n",
       " 'version=1']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][1432].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e9de9fd8-00e5-4258-90a8-448f719d641a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=gene:ENSG00000186092',\n",
       " 'Name=OR4F5',\n",
       " 'biotype=protein_coding',\n",
       " 'description=olfactory receptor family 4 subfamily F member 5 [Source:HGNC Symbol%3BAcc:HGNC:14825]',\n",
       " 'gene_id=ENSG00000186092',\n",
       " 'logic_name=ensembl_havana_gene_homo_sapiens',\n",
       " 'version=7']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df[\"Tags\"][93].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16997ae5-b75e-4565-986e-970a49da4266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328201</th>\n",
       "      <td>10</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133797422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467767</th>\n",
       "      <td>11</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>135086622.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673000</th>\n",
       "      <td>12</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133275309.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862634</th>\n",
       "      <td>13</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>114364328.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924501</th>\n",
       "      <td>14</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>107043718.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048326</th>\n",
       "      <td>15</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>101991189.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171301</th>\n",
       "      <td>16</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90338345.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329042</th>\n",
       "      <td>17</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>83257441.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536644</th>\n",
       "      <td>18</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80373285.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598313</th>\n",
       "      <td>19</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58617616.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1791801</th>\n",
       "      <td>2</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242193529.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063156</th>\n",
       "      <td>20</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64444167.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145930</th>\n",
       "      <td>21</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46709983.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2186784</th>\n",
       "      <td>22</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50818468.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2261287</th>\n",
       "      <td>3</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>198295559.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2484781</th>\n",
       "      <td>4</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>190214555.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2628956</th>\n",
       "      <td>5</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>181538259.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786252</th>\n",
       "      <td>6</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170805979.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2952608</th>\n",
       "      <td>7</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>159345973.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3120542</th>\n",
       "      <td>8</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145138636.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248106</th>\n",
       "      <td>9</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>138394717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387542</th>\n",
       "      <td>MT</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16569.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387706</th>\n",
       "      <td>X</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>156040895.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3509168</th>\n",
       "      <td>Y</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>57227415.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome        Type  First_index   Last_index Value Signal  \\\n",
       "0         1  GRCh38  chromosome          1.0  248956422.0     .      .   \n",
       "328201   10  GRCh38  chromosome          1.0  133797422.0     .      .   \n",
       "467767   11  GRCh38  chromosome          1.0  135086622.0     .      .   \n",
       "673000   12  GRCh38  chromosome          1.0  133275309.0     .      .   \n",
       "862634   13  GRCh38  chromosome          1.0  114364328.0     .      .   \n",
       "924501   14  GRCh38  chromosome          1.0  107043718.0     .      .   \n",
       "1048326  15  GRCh38  chromosome          1.0  101991189.0     .      .   \n",
       "1171301  16  GRCh38  chromosome          1.0   90338345.0     .      .   \n",
       "1329042  17  GRCh38  chromosome          1.0   83257441.0     .      .   \n",
       "1536644  18  GRCh38  chromosome          1.0   80373285.0     .      .   \n",
       "1598313  19  GRCh38  chromosome          1.0   58617616.0     .      .   \n",
       "1791801   2  GRCh38  chromosome          1.0  242193529.0     .      .   \n",
       "2063156  20  GRCh38  chromosome          1.0   64444167.0     .      .   \n",
       "2145930  21  GRCh38  chromosome          1.0   46709983.0     .      .   \n",
       "2186784  22  GRCh38  chromosome          1.0   50818468.0     .      .   \n",
       "2261287   3  GRCh38  chromosome          1.0  198295559.0     .      .   \n",
       "2484781   4  GRCh38  chromosome          1.0  190214555.0     .      .   \n",
       "2628956   5  GRCh38  chromosome          1.0  181538259.0     .      .   \n",
       "2786252   6  GRCh38  chromosome          1.0  170805979.0     .      .   \n",
       "2952608   7  GRCh38  chromosome          1.0  159345973.0     .      .   \n",
       "3120542   8  GRCh38  chromosome          1.0  145138636.0     .      .   \n",
       "3248106   9  GRCh38  chromosome          1.0  138394717.0     .      .   \n",
       "3387542  MT  GRCh38  chromosome          1.0      16569.0     .      .   \n",
       "3387706   X  GRCh38  chromosome          1.0  156040895.0     .      .   \n",
       "3509168   Y  GRCh38  chromosome          1.0   57227415.0     .      .   \n",
       "\n",
       "        Unknown                                               Tags  \n",
       "0             .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "328201        .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  \n",
       "467767        .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  \n",
       "673000        .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  \n",
       "862634        .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  \n",
       "924501        .  ID=chromosome:14;Alias=CM000676.2,chr14,NC_000...  \n",
       "1048326       .  ID=chromosome:15;Alias=CM000677.2,chr15,NC_000...  \n",
       "1171301       .  ID=chromosome:16;Alias=CM000678.2,chr16,NC_000...  \n",
       "1329042       .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...  \n",
       "1536644       .  ID=chromosome:18;Alias=CM000680.2,chr18,NC_000...  \n",
       "1598313       .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...  \n",
       "1791801       .  ID=chromosome:2;Alias=CM000664.2,chr2,NC_00000...  \n",
       "2063156       .  ID=chromosome:20;Alias=CM000682.2,chr20,NC_000...  \n",
       "2145930       .  ID=chromosome:21;Alias=CM000683.2,chr21,NC_000...  \n",
       "2186784       .  ID=chromosome:22;Alias=CM000684.2,chr22,NC_000...  \n",
       "2261287       .  ID=chromosome:3;Alias=CM000665.2,chr3,NC_00000...  \n",
       "2484781       .  ID=chromosome:4;Alias=CM000666.2,chr4,NC_00000...  \n",
       "2628956       .  ID=chromosome:5;Alias=CM000667.2,chr5,NC_00000...  \n",
       "2786252       .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...  \n",
       "2952608       .  ID=chromosome:7;Alias=CM000669.2,chr7,NC_00000...  \n",
       "3120542       .  ID=chromosome:8;Alias=CM000670.2,chr8,NC_00000...  \n",
       "3248106       .  ID=chromosome:9;Alias=CM000671.2,chr9,NC_00000...  \n",
       "3387542       .   ID=chromosome:MT;Alias=chrM,J01415.2,NC_012920.1  \n",
       "3387706       .  ID=chromosome:X;Alias=CM000685.2,chrX,NC_00002...  \n",
       "3509168       .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosomes = tanscripts_df.loc[tanscripts_df[\"Type\"] == 'chromosome']\n",
    "chromosomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "012d28f9-3d6f-4874-b56e-2670b34f2d86",
   "metadata": {},
   "source": [
    "As linhas que contêm o Tag \"Parent\" e apenas essa tag."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3d49f44-e817-413f-b2f9-31f35022c13f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df.loc[(tanscripts_df[\"Tags\"].str.contains(\"Parent\", na=False)) & (~tanscripts_df[\"Tags\"].str.contains(\";\", na=False))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bf3fe1a-2a6d-4ced-a866-2a60e3a4d9bb",
   "metadata": {},
   "source": [
    "As linhas que não contêm a tag \"Parent\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8100ef1b-c707-4cc7-aafe-f18b1cd0b8a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df.loc[~tanscripts_df[\"Tags\"].str.contains(\"Parent\", na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ff3790-de87-43e6-bb90-541bc3e33221",
   "metadata": {},
   "source": [
    "## Dividir a coluna Tags em várias colunas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eb4b1d0a-c6c7-4d25-ba35-36bfd1083102",
   "metadata": {},
   "outputs": [],
   "source": [
    "def divide_TagsColumn():\n",
    "    DesirableTags = [\"ID\", \"Parent\"]#, \"biotype\"]\n",
    "    IDs = []\n",
    "    parents = []\n",
    "    #biotypes = []\n",
    "    names = []\n",
    "    for transcript_tags in tanscripts_df[\"Tags\"]:    #for each row in Tags column\n",
    "        if \";\" in str(transcript_tags):\n",
    "            present_tags = []\n",
    "            row_tags = transcript_tags.split(\";\")\n",
    "            for tag in row_tags:\n",
    "                Tag_type = tag.split(\"=\")[0]\n",
    "                if Tag_type == \"ID\":\n",
    "                    IDs.append(tag)\n",
    "                    present_tags.append(\"id\")\n",
    "                elif Tag_type == \"Parent\":\n",
    "                    parents.append(tag)\n",
    "                    present_tags.append(\"parent\")\n",
    "                elif Tag_type == \"Name\":\n",
    "                    names.append(str(tag).split(\"=\")[1])\n",
    "                    present_tags.append(\"Name\")\n",
    "            if \"id\" not in present_tags:\n",
    "                IDs.append(np.nan)\n",
    "            if \"parent\" not in present_tags:\n",
    "                parents.append(np.nan)\n",
    "            if \"Name\" not in present_tags:\n",
    "                names.append(np.nan)\n",
    "\n",
    "        elif \";\" not in str(transcript_tags):\n",
    "            if \"ID=\" in str(transcript_tags):\n",
    "                IDs.append(str(transcript_tags))\n",
    "                parents.append(np.nan)\n",
    "                names.append(np.nan)\n",
    "            elif \"Parent=\" in str(transcript_tags):\n",
    "                IDs.append(np.nan)\n",
    "                parents.append(str(transcript_tags))\n",
    "                names.append(np.nan)\n",
    "            elif \"Name=\" in str(transcript_tags):\n",
    "                IDs.append(np.nan)\n",
    "                parents.append(np.nan)\n",
    "                names.append(str(transcript_tags.split(\"=\")[1]))\n",
    "            else:\n",
    "                IDs.append(np.nan)\n",
    "                parents.append(np.nan)\n",
    "                names.append(np.nan)\n",
    "\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"ID\", IDs, True)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Parent\", parents, True)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Name\", names, True)\n",
    "    tanscripts_df.drop(['Tags'], axis=1)\n",
    "    return tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c47458a0-fc6c-4d31-8757-7dc0f9f5acef",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df = divide_TagsColumn()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5a1ae314-777c-4f01-bdd4-47ef88a84c6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "      <td>ID=chromosome:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469.0</td>\n",
       "      <td>11240.0</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655.0</td>\n",
       "      <td>10657.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678.0</td>\n",
       "      <td>10687.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519861</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215804.0</td>\n",
       "      <td>57215813.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519862</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215832.0</td>\n",
       "      <td>57215840.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519863</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215833.0</td>\n",
       "      <td>57215842.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519864</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215867.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519865</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>57215862.0</td>\n",
       "      <td>57215870.0</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3519866 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome               Type  First_index   Last_index    Value  \\\n",
       "0         1  GRCh38         chromosome          1.0  248956422.0        .   \n",
       "1         1       .  biological_region      10469.0      11240.0  1.3e+03   \n",
       "2         1       .  biological_region      10650.0      10657.0    0.999   \n",
       "3         1       .  biological_region      10655.0      10657.0    0.999   \n",
       "4         1       .  biological_region      10678.0      10687.0    0.999   \n",
       "...      ..     ...                ...          ...          ...      ...   \n",
       "3519861   Y       .  biological_region   57215804.0   57215813.0    0.999   \n",
       "3519862   Y       .  biological_region   57215832.0   57215840.0    0.999   \n",
       "3519863   Y       .  biological_region   57215833.0   57215842.0    0.999   \n",
       "3519864   Y       .  biological_region   57215862.0   57215867.0    0.999   \n",
       "3519865   Y       .  biological_region   57215862.0   57215870.0    0.999   \n",
       "\n",
       "        Signal Unknown                                               Tags  \\\n",
       "0            .       .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...   \n",
       "1            .       .           external_name=oe %3D 0.79;logic_name=cpg   \n",
       "2            +       .                                 logic_name=eponine   \n",
       "3            -       .                                 logic_name=eponine   \n",
       "4            +       .                                 logic_name=eponine   \n",
       "...        ...     ...                                                ...   \n",
       "3519861      -       .                                 logic_name=eponine   \n",
       "3519862      +       .                                 logic_name=eponine   \n",
       "3519863      -       .                                 logic_name=eponine   \n",
       "3519864      +       .                                 logic_name=eponine   \n",
       "3519865      -       .                                 logic_name=eponine   \n",
       "\n",
       "                      ID Parent Name  \n",
       "0        ID=chromosome:1    NaN  NaN  \n",
       "1                    NaN    NaN  NaN  \n",
       "2                    NaN    NaN  NaN  \n",
       "3                    NaN    NaN  NaN  \n",
       "4                    NaN    NaN  NaN  \n",
       "...                  ...    ...  ...  \n",
       "3519861              NaN    NaN  NaN  \n",
       "3519862              NaN    NaN  NaN  \n",
       "3519863              NaN    NaN  NaN  \n",
       "3519864              NaN    NaN  NaN  \n",
       "3519865              NaN    NaN  NaN  \n",
       "\n",
       "[3519866 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "455327b1-d621-4edf-926f-7e6723245590",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>ENSE00002234944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>ENSE00003582793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000456328;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000456328</td>\n",
       "      <td>ENSE00002312635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010.0</td>\n",
       "      <td>12057.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000450305</td>\n",
       "      <td>ENSE00001948541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000450305;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000450305</td>\n",
       "      <td>ENSE00001671638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519839</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212184.0</td>\n",
       "      <td>57213125.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>ENSE00004015127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519840</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213204.0</td>\n",
       "      <td>57213357.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>ENSE00004015126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519841</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213526.0</td>\n",
       "      <td>57213602.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>ENSE00004015125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519842</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213880.0</td>\n",
       "      <td>57213964.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>ENSE00004015124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519843</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214350.0</td>\n",
       "      <td>57214397.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000711270;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000711270</td>\n",
       "      <td>ENSE00004015123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1668828 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr  Genome  Type  First_index  Last_index Value Signal Unknown  \\\n",
       "18        1  havana  exon      11869.0     12227.0     .      +       .   \n",
       "19        1  havana  exon      12613.0     12721.0     .      +       .   \n",
       "20        1  havana  exon      13221.0     14409.0     .      +       .   \n",
       "24        1  havana  exon      12010.0     12057.0     .      +       .   \n",
       "25        1  havana  exon      12179.0     12227.0     .      +       .   \n",
       "...      ..     ...   ...          ...         ...   ...    ...     ...   \n",
       "3519839   Y  havana  exon   57212184.0  57213125.0     .      -       .   \n",
       "3519840   Y  havana  exon   57213204.0  57213357.0     .      -       .   \n",
       "3519841   Y  havana  exon   57213526.0  57213602.0     .      -       .   \n",
       "3519842   Y  havana  exon   57213880.0  57213964.0     .      -       .   \n",
       "3519843   Y  havana  exon   57214350.0  57214397.0     .      -       .   \n",
       "\n",
       "                                                      Tags   ID  \\\n",
       "18       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "19       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "20       Parent=transcript:ENST00000456328;Name=ENSE000...  NaN   \n",
       "24       Parent=transcript:ENST00000450305;Name=ENSE000...  NaN   \n",
       "25       Parent=transcript:ENST00000450305;Name=ENSE000...  NaN   \n",
       "...                                                    ...  ...   \n",
       "3519839  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519840  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519841  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519842  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "3519843  Parent=transcript:ENST00000711270;Name=ENSE000...  NaN   \n",
       "\n",
       "                                    Parent             Name  \n",
       "18       Parent=transcript:ENST00000456328  ENSE00002234944  \n",
       "19       Parent=transcript:ENST00000456328  ENSE00003582793  \n",
       "20       Parent=transcript:ENST00000456328  ENSE00002312635  \n",
       "24       Parent=transcript:ENST00000450305  ENSE00001948541  \n",
       "25       Parent=transcript:ENST00000450305  ENSE00001671638  \n",
       "...                                    ...              ...  \n",
       "3519839  Parent=transcript:ENST00000711270  ENSE00004015127  \n",
       "3519840  Parent=transcript:ENST00000711270  ENSE00004015126  \n",
       "3519841  Parent=transcript:ENST00000711270  ENSE00004015125  \n",
       "3519842  Parent=transcript:ENST00000711270  ENSE00004015124  \n",
       "3519843  Parent=transcript:ENST00000711270  ENSE00004015123  \n",
       "\n",
       "[1668828 rows x 12 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df.loc[tanscripts_df[\"Type\"] == 'exon']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8158e8b7-4ea5-4b9e-9d67-3bf48d4eda1e",
   "metadata": {},
   "source": [
    "## As linhas com a tag de anotaçao \"MANE_Select\", ou seja os transcritos principais a procurar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b583c375-9626-4fb4-83ba-f589a293e14c",
   "metadata": {},
   "outputs": [],
   "source": [
    "MS_rows = tanscripts_df['Tags'].str.contains(\"MANE_Select\", na=False)\n",
    "tanscripts_df_ManeSelect = tanscripts_df[MS_rows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7c73ed5e-c9fc-4f1b-89a7-8a962159be4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000641515;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000641515</td>\n",
       "      <td>Parent=gene:ENSG00000186092</td>\n",
       "      <td>OR4F5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000426406;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000426406</td>\n",
       "      <td>Parent=gene:ENSG00000284733</td>\n",
       "      <td>OR4F29-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000332831;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000332831</td>\n",
       "      <td>Parent=gene:ENSG00000284662</td>\n",
       "      <td>OR4F16-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000616016;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000616016</td>\n",
       "      <td>Parent=gene:ENSG00000187634</td>\n",
       "      <td>SAMD11-209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959256.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000327044;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000327044</td>\n",
       "      <td>Parent=gene:ENSG00000188976</td>\n",
       "      <td>NOC2L-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518468</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24618004.0</td>\n",
       "      <td>24639207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382392;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382392</td>\n",
       "      <td>Parent=gene:ENSG00000183795</td>\n",
       "      <td>BPY2B-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518527</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24763069.0</td>\n",
       "      <td>24813393.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382365;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382365</td>\n",
       "      <td>Parent=gene:ENSG00000187191</td>\n",
       "      <td>DAZ3-202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518777</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24833919.0</td>\n",
       "      <td>24907040.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000682740;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000682740</td>\n",
       "      <td>Parent=gene:ENSG00000205916</td>\n",
       "      <td>DAZ4-208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519057</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25052104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382287;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382287</td>\n",
       "      <td>Parent=gene:ENSG00000185894</td>\n",
       "      <td>BPY2C-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519268</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000306609;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000306609</td>\n",
       "      <td>Parent=gene:ENSG00000172288</td>\n",
       "      <td>CDY1-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19226 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "94        1          havana  mRNA      65419.0     71585.0     .      +   \n",
       "396       1  ensembl_havana  mRNA     450740.0    451678.0     .      -   \n",
       "542       1  ensembl_havana  mRNA     685716.0    686654.0     .      -   \n",
       "1162      1          havana  mRNA     923923.0    944574.0     .      +   \n",
       "1525      1  ensembl_havana  mRNA     944203.0    959256.0     .      -   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3518468   Y          havana  mRNA   24618004.0  24639207.0     .      +   \n",
       "3518527   Y  ensembl_havana  mRNA   24763069.0  24813393.0     .      -   \n",
       "3518777   Y          havana  mRNA   24833919.0  24907040.0     .      +   \n",
       "3519057   Y          havana  mRNA   25030901.0  25052104.0     .      -   \n",
       "3519268   Y  ensembl_havana  mRNA   25622117.0  25624902.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \\\n",
       "94            .  ID=transcript:ENST00000641515;Parent=gene:ENSG...   \n",
       "396           .  ID=transcript:ENST00000426406;Parent=gene:ENSG...   \n",
       "542           .  ID=transcript:ENST00000332831;Parent=gene:ENSG...   \n",
       "1162          .  ID=transcript:ENST00000616016;Parent=gene:ENSG...   \n",
       "1525          .  ID=transcript:ENST00000327044;Parent=gene:ENSG...   \n",
       "...         ...                                                ...   \n",
       "3518468       .  ID=transcript:ENST00000382392;Parent=gene:ENSG...   \n",
       "3518527       .  ID=transcript:ENST00000382365;Parent=gene:ENSG...   \n",
       "3518777       .  ID=transcript:ENST00000682740;Parent=gene:ENSG...   \n",
       "3519057       .  ID=transcript:ENST00000382287;Parent=gene:ENSG...   \n",
       "3519268       .  ID=transcript:ENST00000306609;Parent=gene:ENSG...   \n",
       "\n",
       "                                    ID                       Parent  \\\n",
       "94       ID=transcript:ENST00000641515  Parent=gene:ENSG00000186092   \n",
       "396      ID=transcript:ENST00000426406  Parent=gene:ENSG00000284733   \n",
       "542      ID=transcript:ENST00000332831  Parent=gene:ENSG00000284662   \n",
       "1162     ID=transcript:ENST00000616016  Parent=gene:ENSG00000187634   \n",
       "1525     ID=transcript:ENST00000327044  Parent=gene:ENSG00000188976   \n",
       "...                                ...                          ...   \n",
       "3518468  ID=transcript:ENST00000382392  Parent=gene:ENSG00000183795   \n",
       "3518527  ID=transcript:ENST00000382365  Parent=gene:ENSG00000187191   \n",
       "3518777  ID=transcript:ENST00000682740  Parent=gene:ENSG00000205916   \n",
       "3519057  ID=transcript:ENST00000382287  Parent=gene:ENSG00000185894   \n",
       "3519268  ID=transcript:ENST00000306609  Parent=gene:ENSG00000172288   \n",
       "\n",
       "               Name  \n",
       "94        OR4F5-201  \n",
       "396      OR4F29-201  \n",
       "542      OR4F16-201  \n",
       "1162     SAMD11-209  \n",
       "1525      NOC2L-201  \n",
       "...             ...  \n",
       "3518468   BPY2B-201  \n",
       "3518527    DAZ3-202  \n",
       "3518777    DAZ4-208  \n",
       "3519057   BPY2C-201  \n",
       "3519268    CDY1-201  \n",
       "\n",
       "[19226 rows x 12 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0256fcf4-cc2a-42c5-83b0-c6c0e1d07608",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_Filteredcolumn(tanscripts_df):\n",
    "    GeneSymbols = []\n",
    "    for Name in tanscripts_df[\"Name\"]:\n",
    "        GeneSymbol = str(Name).split(\"-\")[0]\n",
    "        GeneSymbols.append(GeneSymbol)\n",
    "    tanscripts_df.insert(len(tanscripts_df.columns), \"Gene Symbol\", GeneSymbols, True)\n",
    "    return tanscripts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "44124abe-cf03-4a3a-94a5-bf164a4fe16e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df_gn = add_Filteredcolumn(tanscripts_df_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2ffdf112-5859-43a4-8430-83eac5a98771",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000641515;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000641515</td>\n",
       "      <td>Parent=gene:ENSG00000186092</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000426406;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000426406</td>\n",
       "      <td>Parent=gene:ENSG00000284733</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000332831;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000332831</td>\n",
       "      <td>Parent=gene:ENSG00000284662</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000616016;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000616016</td>\n",
       "      <td>Parent=gene:ENSG00000187634</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959256.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000327044;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000327044</td>\n",
       "      <td>Parent=gene:ENSG00000188976</td>\n",
       "      <td>NOC2L-201</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518468</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24618004.0</td>\n",
       "      <td>24639207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382392;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382392</td>\n",
       "      <td>Parent=gene:ENSG00000183795</td>\n",
       "      <td>BPY2B-201</td>\n",
       "      <td>BPY2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518527</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24763069.0</td>\n",
       "      <td>24813393.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382365;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382365</td>\n",
       "      <td>Parent=gene:ENSG00000187191</td>\n",
       "      <td>DAZ3-202</td>\n",
       "      <td>DAZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518777</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24833919.0</td>\n",
       "      <td>24907040.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000682740;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000682740</td>\n",
       "      <td>Parent=gene:ENSG00000205916</td>\n",
       "      <td>DAZ4-208</td>\n",
       "      <td>DAZ4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519057</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25052104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382287;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382287</td>\n",
       "      <td>Parent=gene:ENSG00000185894</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>BPY2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519268</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000306609;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000306609</td>\n",
       "      <td>Parent=gene:ENSG00000172288</td>\n",
       "      <td>CDY1-201</td>\n",
       "      <td>CDY1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19226 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "94        1          havana  mRNA      65419.0     71585.0     .      +   \n",
       "396       1  ensembl_havana  mRNA     450740.0    451678.0     .      -   \n",
       "542       1  ensembl_havana  mRNA     685716.0    686654.0     .      -   \n",
       "1162      1          havana  mRNA     923923.0    944574.0     .      +   \n",
       "1525      1  ensembl_havana  mRNA     944203.0    959256.0     .      -   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3518468   Y          havana  mRNA   24618004.0  24639207.0     .      +   \n",
       "3518527   Y  ensembl_havana  mRNA   24763069.0  24813393.0     .      -   \n",
       "3518777   Y          havana  mRNA   24833919.0  24907040.0     .      +   \n",
       "3519057   Y          havana  mRNA   25030901.0  25052104.0     .      -   \n",
       "3519268   Y  ensembl_havana  mRNA   25622117.0  25624902.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \\\n",
       "94            .  ID=transcript:ENST00000641515;Parent=gene:ENSG...   \n",
       "396           .  ID=transcript:ENST00000426406;Parent=gene:ENSG...   \n",
       "542           .  ID=transcript:ENST00000332831;Parent=gene:ENSG...   \n",
       "1162          .  ID=transcript:ENST00000616016;Parent=gene:ENSG...   \n",
       "1525          .  ID=transcript:ENST00000327044;Parent=gene:ENSG...   \n",
       "...         ...                                                ...   \n",
       "3518468       .  ID=transcript:ENST00000382392;Parent=gene:ENSG...   \n",
       "3518527       .  ID=transcript:ENST00000382365;Parent=gene:ENSG...   \n",
       "3518777       .  ID=transcript:ENST00000682740;Parent=gene:ENSG...   \n",
       "3519057       .  ID=transcript:ENST00000382287;Parent=gene:ENSG...   \n",
       "3519268       .  ID=transcript:ENST00000306609;Parent=gene:ENSG...   \n",
       "\n",
       "                                    ID                       Parent  \\\n",
       "94       ID=transcript:ENST00000641515  Parent=gene:ENSG00000186092   \n",
       "396      ID=transcript:ENST00000426406  Parent=gene:ENSG00000284733   \n",
       "542      ID=transcript:ENST00000332831  Parent=gene:ENSG00000284662   \n",
       "1162     ID=transcript:ENST00000616016  Parent=gene:ENSG00000187634   \n",
       "1525     ID=transcript:ENST00000327044  Parent=gene:ENSG00000188976   \n",
       "...                                ...                          ...   \n",
       "3518468  ID=transcript:ENST00000382392  Parent=gene:ENSG00000183795   \n",
       "3518527  ID=transcript:ENST00000382365  Parent=gene:ENSG00000187191   \n",
       "3518777  ID=transcript:ENST00000682740  Parent=gene:ENSG00000205916   \n",
       "3519057  ID=transcript:ENST00000382287  Parent=gene:ENSG00000185894   \n",
       "3519268  ID=transcript:ENST00000306609  Parent=gene:ENSG00000172288   \n",
       "\n",
       "               Name Gene Symbol  \n",
       "94        OR4F5-201       OR4F5  \n",
       "396      OR4F29-201      OR4F29  \n",
       "542      OR4F16-201      OR4F16  \n",
       "1162     SAMD11-209      SAMD11  \n",
       "1525      NOC2L-201       NOC2L  \n",
       "...             ...         ...  \n",
       "3518468   BPY2B-201       BPY2B  \n",
       "3518527    DAZ3-202        DAZ3  \n",
       "3518777    DAZ4-208        DAZ4  \n",
       "3519057   BPY2C-201       BPY2C  \n",
       "3519268    CDY1-201        CDY1  \n",
       "\n",
       "[19226 rows x 13 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "5ce10542-8fac-48ee-a832-73f8fb09a13f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#trscpts_GN = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/transcrpts_geneNames.gff3\"\n",
    "#writeNewtxt(trscpts_GN, tanscripts_df_ManeSelect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f408b98-baa5-4a1d-bb76-0eecb2f99973",
   "metadata": {},
   "outputs": [],
   "source": [
    "#trscpts_GN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9016b817-8a6e-471c-a557-a66df0e063c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ID=transcript:ENST00000641515',\n",
       " 'Parent=gene:ENSG00000186092',\n",
       " 'Name=OR4F5-201',\n",
       " 'biotype=protein_coding',\n",
       " 'ccdsid=CCDS30547.2',\n",
       " 'tag=basic,Ensembl_canonical,MANE_Select,GENCODE Primary',\n",
       " 'transcript_id=ENST00000641515',\n",
       " 'version=2']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect[\"Tags\"][94].split(\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73278e36-00ac-411c-9427-35eee73b325b",
   "metadata": {},
   "source": [
    "## Filtrar os transcritos que correspondem a genes relacionados a doenças"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "325176ec-6f0d-4963-9bc8-f30533e98f63",
   "metadata": {},
   "outputs": [],
   "source": [
    "DGs_read = \"/home/andre/Documents/trabalho/datasets/Disease_genes/GeneHancer/GDs_GeneCards_treated_final.txt\"\n",
    "DGs_df = pd.read_csv(DGs_read, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f5ed8c5a-5603-4c43-8cee-08b9ca7a8e75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Locus</th>\n",
       "      <th>MalaCards Disorders (sorted by relevance)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>Alpha-1-B Glycoprotein</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>Voyeurism; Nephrotic Syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 Complementation Factor</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>Lynch Syndrome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>Alpha-2-Macroglobulin</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>Alpha-2-Macroglobulin Deficiency; Alzheimer&amp;#3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M Antisense RNA 1</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>Gastric Cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>Alpha-2-Macroglobulin Like 1</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>Otitis Media; Brugada Syndrome 6; Baraitser-Wi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22140</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>Zinc Finger X-Linked Duplicated A</td>\n",
       "      <td>Xp11.21</td>\n",
       "      <td>Osteopathia Striata With Cranial Sclerosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22141</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>Zinc Finger X-Linked Duplicated B</td>\n",
       "      <td>Xp11.21</td>\n",
       "      <td>Oculopharyngodistal Myopathy 1; Osteopathia St...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22142</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>ZXD Family Zinc Finger C</td>\n",
       "      <td>3q21.3</td>\n",
       "      <td>Moyamoya Disease 1; Microcephalic Osteodysplas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22143</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>Zyg-11 Family Member B, Cell Cycle Regulator</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>Craniofacial Microsomia 1; Cardiomyopathy, Fam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22144</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>Zyxin</td>\n",
       "      <td>7q34</td>\n",
       "      <td>Lipomatosis, Familial Multiple; Ewing Sarcoma;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22145 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Symbol                                     Gene Name     Locus  \\\n",
       "0            A1BG                        Alpha-1-B Glycoprotein  19q13.43   \n",
       "1            A1CF                APOBEC1 Complementation Factor  10q11.23   \n",
       "2             A2M                         Alpha-2-Macroglobulin  12p13.31   \n",
       "3         A2M-AS1                           A2M Antisense RNA 1  12p13.31   \n",
       "4           A2ML1                  Alpha-2-Macroglobulin Like 1  12p13.31   \n",
       "...           ...                                           ...       ...   \n",
       "22140        ZXDA             Zinc Finger X-Linked Duplicated A   Xp11.21   \n",
       "22141        ZXDB             Zinc Finger X-Linked Duplicated B   Xp11.21   \n",
       "22142        ZXDC                      ZXD Family Zinc Finger C    3q21.3   \n",
       "22143      ZYG11B  Zyg-11 Family Member B, Cell Cycle Regulator    1p32.3   \n",
       "22144         ZYX                                         Zyxin      7q34   \n",
       "\n",
       "               MalaCards Disorders (sorted by relevance)  \n",
       "0                         Voyeurism; Nephrotic Syndrome   \n",
       "1                                      Lynch Syndrome 2   \n",
       "2      Alpha-2-Macroglobulin Deficiency; Alzheimer&#3...  \n",
       "3                                        Gastric Cancer   \n",
       "4      Otitis Media; Brugada Syndrome 6; Baraitser-Wi...  \n",
       "...                                                  ...  \n",
       "22140        Osteopathia Striata With Cranial Sclerosis   \n",
       "22141  Oculopharyngodistal Myopathy 1; Osteopathia St...  \n",
       "22142  Moyamoya Disease 1; Microcephalic Osteodysplas...  \n",
       "22143  Craniofacial Microsomia 1; Cardiomyopathy, Fam...  \n",
       "22144  Lipomatosis, Familial Multiple; Ewing Sarcoma;...  \n",
       "\n",
       "[22145 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DGs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1e22c329-3b9f-428d-aa49-06e48ebd8dfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.arrays.numpy_.NumpyExtensionArray"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(DGs_df[\"Gene Name\"].array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7caf8c86-9091-4599-9bc4-ffbde0f5c6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df_ManeSelect[\"Name\"] in DGs_df[\"Gene Name\"].array\n",
    "#tanscripts_df_ManeSelect_DGs = tanscripts_df_ManeSelect.loc[tanscripts_df_ManeSelect[\"Name\"] in DGs_df[\"Gene Name\"].array]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b18e4d2a-cf1a-4b5c-8278-97132061cdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "tanscripts_df_ManeSelect_DGs = tanscripts_df_gn[tanscripts_df_gn['Gene Symbol'].isin(DGs_df['Gene Symbol'])]\n",
    "#tanscripts_df_ManeSelect_DGs = tanscripts_df_ManeSelect[tanscripts_df_ManeSelect['Name'].apply(lambda name: DGs_df['Gene Symbol'].str.contains(str(name), case=False).any())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "124791d7-42a7-41bd-9858-fa6efab87c26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000426406;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000426406</td>\n",
       "      <td>Parent=gene:ENSG00000284733</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000332831;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000332831</td>\n",
       "      <td>Parent=gene:ENSG00000284662</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>944574.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000616016;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000616016</td>\n",
       "      <td>Parent=gene:ENSG00000187634</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>944203.0</td>\n",
       "      <td>959256.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000327044;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000327044</td>\n",
       "      <td>Parent=gene:ENSG00000188976</td>\n",
       "      <td>NOC2L-201</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1622</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>960584.0</td>\n",
       "      <td>965719.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000338591;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000338591</td>\n",
       "      <td>Parent=gene:ENSG00000187961</td>\n",
       "      <td>KLHL17-201</td>\n",
       "      <td>KLHL17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518468</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24618004.0</td>\n",
       "      <td>24639207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382392;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382392</td>\n",
       "      <td>Parent=gene:ENSG00000183795</td>\n",
       "      <td>BPY2B-201</td>\n",
       "      <td>BPY2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518527</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24763069.0</td>\n",
       "      <td>24813393.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382365;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382365</td>\n",
       "      <td>Parent=gene:ENSG00000187191</td>\n",
       "      <td>DAZ3-202</td>\n",
       "      <td>DAZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3518777</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24833919.0</td>\n",
       "      <td>24907040.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000682740;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000682740</td>\n",
       "      <td>Parent=gene:ENSG00000205916</td>\n",
       "      <td>DAZ4-208</td>\n",
       "      <td>DAZ4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519057</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25030901.0</td>\n",
       "      <td>25052104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000382287;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000382287</td>\n",
       "      <td>Parent=gene:ENSG00000185894</td>\n",
       "      <td>BPY2C-201</td>\n",
       "      <td>BPY2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519268</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000306609;Parent=gene:ENSG...</td>\n",
       "      <td>ID=transcript:ENST00000306609</td>\n",
       "      <td>Parent=gene:ENSG00000172288</td>\n",
       "      <td>CDY1-201</td>\n",
       "      <td>CDY1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16735 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome  Type  First_index  Last_index Value Signal  \\\n",
       "396       1  ensembl_havana  mRNA     450740.0    451678.0     .      -   \n",
       "542       1  ensembl_havana  mRNA     685716.0    686654.0     .      -   \n",
       "1162      1          havana  mRNA     923923.0    944574.0     .      +   \n",
       "1525      1  ensembl_havana  mRNA     944203.0    959256.0     .      -   \n",
       "1622      1  ensembl_havana  mRNA     960584.0    965719.0     .      +   \n",
       "...      ..             ...   ...          ...         ...   ...    ...   \n",
       "3518468   Y          havana  mRNA   24618004.0  24639207.0     .      +   \n",
       "3518527   Y  ensembl_havana  mRNA   24763069.0  24813393.0     .      -   \n",
       "3518777   Y          havana  mRNA   24833919.0  24907040.0     .      +   \n",
       "3519057   Y          havana  mRNA   25030901.0  25052104.0     .      -   \n",
       "3519268   Y  ensembl_havana  mRNA   25622117.0  25624902.0     .      +   \n",
       "\n",
       "        Unknown                                               Tags  \\\n",
       "396           .  ID=transcript:ENST00000426406;Parent=gene:ENSG...   \n",
       "542           .  ID=transcript:ENST00000332831;Parent=gene:ENSG...   \n",
       "1162          .  ID=transcript:ENST00000616016;Parent=gene:ENSG...   \n",
       "1525          .  ID=transcript:ENST00000327044;Parent=gene:ENSG...   \n",
       "1622          .  ID=transcript:ENST00000338591;Parent=gene:ENSG...   \n",
       "...         ...                                                ...   \n",
       "3518468       .  ID=transcript:ENST00000382392;Parent=gene:ENSG...   \n",
       "3518527       .  ID=transcript:ENST00000382365;Parent=gene:ENSG...   \n",
       "3518777       .  ID=transcript:ENST00000682740;Parent=gene:ENSG...   \n",
       "3519057       .  ID=transcript:ENST00000382287;Parent=gene:ENSG...   \n",
       "3519268       .  ID=transcript:ENST00000306609;Parent=gene:ENSG...   \n",
       "\n",
       "                                    ID                       Parent  \\\n",
       "396      ID=transcript:ENST00000426406  Parent=gene:ENSG00000284733   \n",
       "542      ID=transcript:ENST00000332831  Parent=gene:ENSG00000284662   \n",
       "1162     ID=transcript:ENST00000616016  Parent=gene:ENSG00000187634   \n",
       "1525     ID=transcript:ENST00000327044  Parent=gene:ENSG00000188976   \n",
       "1622     ID=transcript:ENST00000338591  Parent=gene:ENSG00000187961   \n",
       "...                                ...                          ...   \n",
       "3518468  ID=transcript:ENST00000382392  Parent=gene:ENSG00000183795   \n",
       "3518527  ID=transcript:ENST00000382365  Parent=gene:ENSG00000187191   \n",
       "3518777  ID=transcript:ENST00000682740  Parent=gene:ENSG00000205916   \n",
       "3519057  ID=transcript:ENST00000382287  Parent=gene:ENSG00000185894   \n",
       "3519268  ID=transcript:ENST00000306609  Parent=gene:ENSG00000172288   \n",
       "\n",
       "               Name Gene Symbol  \n",
       "396      OR4F29-201      OR4F29  \n",
       "542      OR4F16-201      OR4F16  \n",
       "1162     SAMD11-209      SAMD11  \n",
       "1525      NOC2L-201       NOC2L  \n",
       "1622     KLHL17-201      KLHL17  \n",
       "...             ...         ...  \n",
       "3518468   BPY2B-201       BPY2B  \n",
       "3518527    DAZ3-202        DAZ3  \n",
       "3518777    DAZ4-208        DAZ4  \n",
       "3519057   BPY2C-201       BPY2C  \n",
       "3519268    CDY1-201        CDY1  \n",
       "\n",
       "[16735 rows x 13 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tanscripts_df_ManeSelect_DGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "488b32a0-f425-4432-931a-ba7742ea5dc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df_ManeSelect_DGs2 = tanscripts_df_ManeSelect[tanscripts_df_ManeSelect['Name'].str.contains(str(DGs_df['Gene Symbol']), case=False).any()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d9d5d84-c37f-4ba6-9595-a7d2b8757ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tanscripts_df_ManeSelect_DGs2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60191185-3c4f-4319-9422-42af1266ecd3",
   "metadata": {},
   "source": [
    "### Obtenção de todas as linhas do dataframe de transcritos que derivam dos transcritos principais (\"Mane_select\"). Linhas que tenham a anotação: 'Parent=transcript:.............' id do transcrito pretendido."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a5c3c633-e489-4447-941c-b42b57cca639",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RegionsByTranscript(main_transcripts_df):\n",
    "    counter = 0\n",
    "    for row in main_transcripts_df.index:\n",
    "        #print(\"transcripto:\", main_transcripts_df.values[row])\n",
    "        ID_transcript = main_transcripts_df['ID'][row]\n",
    "        transcript = ID_transcript.split(\"=\")[1]\n",
    "        Parent_Tag = \"Parent=\" + transcript\n",
    "        #Parent=transcript:.............'\n",
    "        transcriptRegions_df = tanscripts_df.loc[tanscripts_df[\"Parent\"] == Parent_Tag]\n",
    "        if counter == 0:\n",
    "            #print(\"sou a primeira iteração do novo df\")\n",
    "            MainTranscripts_df = transcriptRegions_df\n",
    "            counter += 1\n",
    "        else:\n",
    "            #print(\"index: \", row)\n",
    "            MainTranscripts_df = pd.concat([MainTranscripts_df, transcriptRegions_df])\n",
    "    return MainTranscripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9c5fbc-af75-416a-bc7d-79de89e689f6",
   "metadata": {},
   "source": [
    "## OBTER AS REGIÕES DERIVADAS DOS TRANSCRITOS QUE SÃO DISEASES GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23e21f81-b968-4fab-8edc-5145e83c87b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_dg_trscpts_df = RegionsByTranscript(tanscripts_df_ManeSelect_DGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26cb86c1-9329-4a74-9f87-810f43440988",
   "metadata": {},
   "outputs": [],
   "source": [
    "### demora mais de 1 hora\n",
    "#main_transcripts_df = RegionsByTranscript()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d35a663e-3288-4fbf-995a-6117e8674ca2",
   "metadata": {},
   "source": [
    "### Problema do espaço que o dataframe ocupa, experimentar analisar e apagar algumas colunas. Resolvido, a coluna other tags estava a ocupar muito espaço."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6831d166-8bab-423f-9c7a-d64fe95efcd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000426406;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000426406</td>\n",
       "      <td>ENSE00003989331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000409316;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000409316</td>\n",
       "      <td>Parent=transcript:ENST00000426406</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000332831;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000332831</td>\n",
       "      <td>ENSE00004001351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000329982;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000329982</td>\n",
       "      <td>Parent=transcript:ENST00000332831</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>924431.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519270</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>ENSE00003978110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519271</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519272</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624527.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519273</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>ENSE00001666303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519274</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25624528.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406711 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome             Type  First_index  Last_index Value  \\\n",
       "397       1  ensembl_havana             exon     450740.0    451678.0     .   \n",
       "398       1  ensembl_havana              CDS     450740.0    451678.0     .   \n",
       "543       1  ensembl_havana             exon     685716.0    686654.0     .   \n",
       "544       1  ensembl_havana              CDS     685716.0    686654.0     .   \n",
       "1163      1          havana   five_prime_UTR     923923.0    924431.0     .   \n",
       "...      ..             ...              ...          ...         ...   ...   \n",
       "3519270   Y  ensembl_havana             exon   25622117.0  25624034.0     .   \n",
       "3519271   Y  ensembl_havana              CDS   25622443.0  25624034.0     .   \n",
       "3519272   Y  ensembl_havana              CDS   25624455.0  25624527.0     .   \n",
       "3519273   Y  ensembl_havana             exon   25624455.0  25624902.0     .   \n",
       "3519274   Y  ensembl_havana  three_prime_UTR   25624528.0  25624902.0     .   \n",
       "\n",
       "        Signal Unknown                                               Tags  \\\n",
       "397          -       .  Parent=transcript:ENST00000426406;Name=ENSE000...   \n",
       "398          -       0  ID=CDS:ENSP00000409316;Parent=transcript:ENST0...   \n",
       "543          -       .  Parent=transcript:ENST00000332831;Name=ENSE000...   \n",
       "544          -       0  ID=CDS:ENSP00000329982;Parent=transcript:ENST0...   \n",
       "1163         +       .                  Parent=transcript:ENST00000616016   \n",
       "...        ...     ...                                                ...   \n",
       "3519270      +       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "3519271      +       0  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519272      +       1  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519273      +       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "3519274      +       .                  Parent=transcript:ENST00000306609   \n",
       "\n",
       "                             ID                             Parent  \\\n",
       "397                         NaN  Parent=transcript:ENST00000426406   \n",
       "398      ID=CDS:ENSP00000409316  Parent=transcript:ENST00000426406   \n",
       "543                         NaN  Parent=transcript:ENST00000332831   \n",
       "544      ID=CDS:ENSP00000329982  Parent=transcript:ENST00000332831   \n",
       "1163                        NaN  Parent=transcript:ENST00000616016   \n",
       "...                         ...                                ...   \n",
       "3519270                     NaN  Parent=transcript:ENST00000306609   \n",
       "3519271  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609   \n",
       "3519272  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609   \n",
       "3519273                     NaN  Parent=transcript:ENST00000306609   \n",
       "3519274                     NaN  Parent=transcript:ENST00000306609   \n",
       "\n",
       "                    Name  \n",
       "397      ENSE00003989331  \n",
       "398                  NaN  \n",
       "543      ENSE00004001351  \n",
       "544                  NaN  \n",
       "1163                 NaN  \n",
       "...                  ...  \n",
       "3519270  ENSE00003978110  \n",
       "3519271              NaN  \n",
       "3519272              NaN  \n",
       "3519273  ENSE00001666303  \n",
       "3519274              NaN  \n",
       "\n",
       "[406711 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_dg_trscpts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f314b424-6bfc-4cad-a5cd-e2847bc528bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "trscpts_main_DG = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/transcrpts_main_DGs.gff3\"\n",
    "writeNewtxt(trscpts_main_DG, main_dg_trscpts_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5de61062-f76f-4767-b985-c790c9cc4745",
   "metadata": {},
   "source": [
    "De todos os elementos são agrupados os elementos que pertencem aos transcritos principais \"Mane_Select\".\n",
    "A densidade de cada tipo de elemeto é feita considerando o intervalo desde o indice inicial mais baixo até ao indice final de maior valor, sendo portanto uma densidade conservativa.\n",
    "Relembrar forma de verificar a densidade dos intrões"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "34da5156-4ffe-43c4-b11b-7d2f42631d0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19',\n",
       "       '2', '20', '21', '22', '3', '4', '5', '6', '7', '8', '9', 'X', 'Y'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_dg_trscpts_df[\"Chr\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cd443f7-e722-4bb0-81e9-e637640c2a26",
   "metadata": {},
   "outputs": [],
   "source": [
    "#main_transcripts_df = main_transcripts_df.drop(\"Other_tags\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98ef0631-85c5-42e6-8df7-2de8eb7ce6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#main_transcripts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc91121f-b857-4129-a9f4-51db8cb4be16",
   "metadata": {},
   "source": [
    "## Methods for calculating the densities"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36fb16c1-9405-4eab-aefe-c8e5a966ec8c",
   "metadata": {},
   "source": [
    "Uma vez que a representação dos cromossomas e da ocorrência dos elementos no cromossoma é feito cromossoma a cromossoma, o calculo tambem e feito da mesma cromossoma a cromossoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9fa85c33-ac44-4276-a7d0-a8e595258d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Função para considerar todo o intervalo entre o primeiro indice até ao ultimo indice dos transcritos\n",
    "def element_chr_array(chromosomes_df, sequence_df, chr_n):\n",
    "    \"\"\"\n",
    "    An array all basePairs of a chromosome, the interval from first instance to the last instance of the sequence type\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    chr_n: String only number ex: \"1\"\n",
    "    \"\"\"\n",
    "    #print(\"searching cytobands in chromossome: \", chr_n)\n",
    "    chromosome_index = chromosomes_df.index[chromosomes_df[\"Chr\"] == str(chr_n)]\n",
    "    chr_size = int(chromosomes_df[\"Last_index\"][chromosome_index].iloc[0])\n",
    "    chr_sequence = sequence_df.loc[sequence_df['Chr'] == str(chr_n)]\n",
    "    chr_array = np.zeros((chr_size), dtype=int)\n",
    "    #elements region\n",
    "    sequence_indexes = []\n",
    "    for row in chr_sequence.index:\n",
    "        sequence_indexes.append(int(float(chr_sequence[\"First_index\"][row])))\n",
    "        sequence_indexes.append(int(float(chr_sequence[\"Last_index\"][row])))\n",
    "        sequence_indexes = [min(sequence_indexes), max(sequence_indexes)]\n",
    "    #print(\"element's initial and final indexes:\", sequence_indexes)\n",
    "    chr_array[sequence_indexes[0] : sequence_indexes[1]] += 1\n",
    "    #print(chr_array[sequence_indexes[0]-20 : sequence_indexes[1]]+20)\n",
    "    return chr_array\n",
    "\n",
    "#Juntar os arrays talvez não seja boa ideia, memória\n",
    "\n",
    "\n",
    "def element_array(chromosomes_df, sequence_df, chr_n):\n",
    "    \"\"\"\n",
    "    An array representing all intances of a sequence type in a chromosome available from the original transcripts dataframe\n",
    "    sequence_df = Genes, CDSs, 5 or 3 prime UTR, Exons\n",
    "    \"\"\"\n",
    "    chromosome_index = chromosomes_df.index[chromosomes_df[\"Chr\"] == str(chr_n)]\n",
    "    chr_size = int(chromosomes_df[\"Last_index\"][chromosome_index].iloc[0])\n",
    "    chr_sequence = sequence_df.loc[sequence_df['Chr'] == str(chr_n)]\n",
    "    chr_array = np.zeros((chr_size), dtype=int)\n",
    "    #elements region\n",
    "    for row in chr_sequence.index:\n",
    "        chr_array[int(float(chr_sequence[\"First_index\"][row])) : int(float(chr_sequence[\"Last_index\"][row]))] += 1\n",
    "    return chr_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "90359246-06df-4f39-a506-0c7e2e782cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def element_chr_trnscpt_array(chromosomes_df, main_trcpts_df, chr_n, element_type=\"CDS\"):\n",
    "    \"\"\"\n",
    "    The array representing all intances first instance until the last instance of the sequence type from the original transcripts dataframe\n",
    "    main_trcpts_df = Genes, includes: CDSs, 5 or 3 prime UTR, Exons\n",
    "    chr_n: String only number ex: \"1\"\n",
    "    element_type can be: CDS, five_prime_UTR or three_prime_UTR\n",
    "    \"\"\"\n",
    "    print(\"searching cytobands in chromossome: \", chr_n)\n",
    "    chromosome = chromosomes_df.loc[chromosomes_df[\"Chr\"] == chr_n]\n",
    "    chr_size = int(chromosome[\"Last_index\"].iloc[0])\n",
    "    chr_array = np.zeros((chr_size), dtype=int)\n",
    "    #elements region\n",
    "    chr_trcpts = main_trcpts_df.loc[main_trcpts_df[\"Chr\"] == chr_n]\n",
    "    chr_transcripts = np.unique_all(chr_trcpts[\"Parent\"].to_numpy())[0]\n",
    "    for transcript in chr_transcripts:\n",
    "        chr_onetrcpt = chr_trcpts.loc[chr_trcpts[\"Parent\"] == str(transcript)]\n",
    "        element_trnscrpt = chr_trcpts.loc[(chr_trcpts[\"Type\"] == str(element_type)) & (chr_trcpts[\"Parent\"] == str(transcript))]\n",
    "        #print(\"element_trnscrpt: \", element_trnscrpt)\n",
    "        first_element = element_trnscrpt.index[0]\n",
    "        last_element = element_trnscrpt.index[-1]\n",
    "        chr_array[int(float(element_trnscrpt[\"First_index\"][first_element])) : int(float(element_trnscrpt[\"Last_index\"][last_element]))] += 1\n",
    "    \n",
    "    #print(chr_array[sequence_indexes[0]-20 : sequence_indexes[1]]+20)\n",
    "    return chr_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6e0fa18e-d098-4d9a-86a8-cf8c4447f421",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_elementDensity_chr(cytobands_df, chr_array, chr_n):\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    chr_densities = []\n",
    "    chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == \"chr\" + str(chr_n)]\n",
    "    #print(\"cytobands of chr\", chr_n, \" indexes: \", chr_cytobands.index)\n",
    "    #Calculo da densidade para cada cytoband\n",
    "    for row in chr_cytobands.index:\n",
    "        cytoband_array = chr_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "        dr_length = np.count_nonzero(cytoband_array)\n",
    "        cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "        density = dr_length/cytoband_length\n",
    "        chr_densities.append(round(density, 10))\n",
    "    return chr_densities   \n",
    "    #cytobands_df.insert(len(cytobands_df.columns), col_name, densities, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "93ab2952-9ba2-4935-b7f0-c0121d2ebaba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_ExonIntronDensity_chr(cytobands_df, chr_sequence_array, chr_exons_array, chr_n):\n",
    "    \"\"\"In a given chromosome calculate de exon and intron densities within a given sequence type (e.g CDSs) for \n",
    "    every cytoband\n",
    "    \"\"\"\n",
    "    #cytobands_df = cytobands_df.drop(cytobands_df[cytobands_df['Chromossome'] == 'chrM'].index)\n",
    "    intron_densities = []\n",
    "    exon_densities = []\n",
    "    #print()\n",
    "    chr_cytobands = cytobands_df.loc[cytobands_df['Chromossome'] == \"chr\" + str(chr_n)]\n",
    "    #Calculo das densidades para cada cytoband\n",
    "    for row in chr_cytobands.index:\n",
    "        exon_length, intron_length = get_lengths(chr_cytobands, chr_sequence_array, chr_exons_array, row)\n",
    "        cytoband_length = chr_cytobands[\"Size\"][row]\n",
    "        exon_density = exon_length/cytoband_length\n",
    "        intron_density = intron_length/cytoband_length\n",
    "        exon_densities.append(round(exon_density, 10))\n",
    "        intron_densities.append(round(intron_density, 10))   #max 6    \n",
    "    return exon_densities, intron_densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9e36d9b6-76ef-45eb-9f8c-e4dae0ebe6c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7da4d400-65c9-4e41-a5cb-ce2af0e63fd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([5, 6, 7, 8, 9])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where(np.array(a) > 10)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e334f87c-7f8b-4ae6-ba27-62a1ff350910",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=[1, 0, 1, 1, 1, 0, 0, 1, 0]\n",
    "b=[1, 1, 0, 1, 0, 1, 0, 1, 0]\n",
    "np.intersect1d(a, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5fb137c0-9cbd-4e37-a043-9f9b7abc41ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lengths(chr_cytobands, chr_sequence_array, chr_exons_array, row):\n",
    "    region_cytoband_array = chr_sequence_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "    exon_cytoband_array = chr_exons_array[int(chr_cytobands[\"First_index\"][row]):int(chr_cytobands[\"Last_index\"][row])]\n",
    "    region_indexes = np.where(region_cytoband_array > 0)[0].tolist()\n",
    "    #print(\"region_indexes: \", region_indexes)\n",
    "    exon_indexes = np.where(exon_cytoband_array > 0)[0].tolist()\n",
    "    #print(\"exon_indexes: \", exon_indexes)\n",
    "    exon_length = len(np.intersect1d(region_indexes, exon_indexes))   #indices que estam na região e são exões\n",
    "    intron_length = len(np.setdiff1d(region_indexes, exon_indexes))    #indices que estam na região mas não são exões\n",
    "    return exon_length, intron_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "739924fa-f2c6-43b3-a577-4bafe5b554f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Não é possivel correr esta função de uma vez devido a restrições de memória\n",
    "def chr_process(chromosomes_df, main_trnscpts_df, cytobands_df, chr_n):\n",
    "    #for chr in chromosomes_df[\"Chr\"]:\n",
    "    cds_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"CDS\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    utr5_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"five_prime_UTR\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    utr3_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"three_prime_UTR\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    exons = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"exon\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    ### criar já o array do chromossoma correspondente à presença de exons\n",
    "    chr_exons_array = element_chr_array(chromosomes_df, exons, chr_n)\n",
    "    ### Para cada um dos 3 tipos de sequencias calcular as densidades necessárias\n",
    "    ###CDS\n",
    "    chr_cds_array = element_chr_array(chromosomes_df, cds_df, chr_n)\n",
    "    chr_cds_dens = calc_elementDensity_chr(cytobands_df, chr_cds_array, chr_n)\n",
    "    cds_exon_dens, cds_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_cds_array, chr_exons_array, chr_n)\n",
    "    ###five_prime_UTR\n",
    "    #print(\"5prime utr dataframe: \", utr5_df)\n",
    "    chr_utr5_array = element_chr_array(chromosomes_df, utr5_df, chr_n)\n",
    "    chr_utr5_dens = calc_elementDensity_chr(cytobands_df, chr_utr5_array, chr_n)\n",
    "    utr5_exon_dens, utr5_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_utr5_array, chr_exons_array, chr_n)\n",
    "    ###three_prime_UTR\n",
    "    chr_utr3_array = element_chr_array(chromosomes_df, utr3_df, chr_n)\n",
    "    chr_utr3_dens = calc_elementDensity_chr(cytobands_df, chr_utr3_array, chr_n)\n",
    "    utr3_exon_dens, utr3_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_utr3_array, chr_exons_array, chr_n)\n",
    "    ###exon\n",
    "    chr_exon_dens = calc_elementDensity_chr(cytobands_df, chr_exons_array, chr_n)\n",
    "    ###Guardar as variáveis em arrays separados dentro ou fora da função\n",
    "    return chr_cds_dens, cds_exon_dens, cds_intron_dens, chr_utr5_dens, utr5_exon_dens, utr5_intron_dens, chr_utr3_dens, utr3_exon_dens, utr3_intron_dens, chr_exon_dens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "53dde56a-1a3b-40dd-ae73-3b816c1006ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chr_process1(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chr_n):\n",
    "    cds_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"CDS\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    ###CDS\n",
    "    chr_cds_array = element_chr_trnscpt_array(chromosomes_df, main_trnscpts_df, chr_n, \"CDS\")\n",
    "    chr_cds_dens = calc_elementDensity_chr(cytobands_df, chr_cds_array, chr_n)\n",
    "    cds_exon_dens, cds_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_cds_array, chr_exons_array, chr_n)\n",
    "    return chr_cds_dens, cds_exon_dens, cds_intron_dens\n",
    "    \n",
    "def chr_process2(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chr_n):\n",
    "    utr5_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"five_prime_UTR\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    ###five_prime_UTR\n",
    "    chr_utr5_array = element_array(chromosomes_df, utr5_df, chr_n)\n",
    "    chr_utr5_dens = calc_elementDensity_chr(cytobands_df, chr_utr5_array, chr_n)\n",
    "    # It is assumed 5 prime UTR is only introns, because the region as no coding role\n",
    "    #utr5_exon_dens, utr5_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_utr5_array, chr_exons_array, chr_n)\n",
    "    return chr_utr5_dens#, utr5_exon_dens, utr5_intron_dens\n",
    "    \n",
    "def chr_process3(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chr_n):\n",
    "    utr3_df = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"three_prime_UTR\") & (main_trnscpts_df[\"Chr\"] == str(chr_n))]\n",
    "    ###three_prime_UTR\n",
    "    chr_utr3_array = element_array(chromosomes_df, utr3_df, chr_n)\n",
    "    chr_utr3_dens = calc_elementDensity_chr(cytobands_df, chr_utr3_array, chr_n)\n",
    "    # It is assumed 3 prime UTR is only introns, because the region as no coding role\n",
    "    #utr3_exon_dens, utr3_intron_dens = calc_ExonIntronDensity_chr(cytobands_df, chr_utr3_array, chr_exons_array, chr_n)\n",
    "    return chr_utr3_dens#, utr3_exon_dens, utr3_intron_dens\n",
    "\n",
    "def chr_process4(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chr_n):\n",
    "    ###exons\n",
    "    chr_exon_dens = calc_elementDensity_chr(cytobands_df, chr_exons_array, chr_n)\n",
    "    return chr_exon_dens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7a5f6119-40f3-44b3-a0df-ceb7dc08f0ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_allChr(chromosomes_df, main_trnscpts_df, cytobands_df):\n",
    "    all_chrs_cdsDens = []\n",
    "    all_chrs_cdsExonDens = []\n",
    "    all_chrs_cdsIntronDens = []\n",
    "    all_chrs_utr5Dens = []\n",
    "    #all_chrs_utr5ExonDens = []\n",
    "    #all_chrs_utr5IntronDens = []\n",
    "    all_chrs_utr3Dens = []\n",
    "    #all_chrs_utr3ExonDens = []\n",
    "    #all_chrs_utr3IntronDens = []\n",
    "    all_chrs_exon_dens = []\n",
    "    #for chrom in range(len(cytobands_df[\"Chromossome\"].unique())):\n",
    "    for chrom in main_trnscpts_df[\"Chr\"].unique():\n",
    "        print(\"processing chromosome: chr\", chrom)\n",
    "        ### criar já o array do chromossoma correspondente à presença de exons\n",
    "        exons = main_trnscpts_df.loc[(main_trnscpts_df[\"Type\"] == \"exon\") & (main_trnscpts_df[\"Chr\"] == str(chrom))]\n",
    "        chr_exons_array = element_array(chromosomes_df, exons, chrom)\n",
    "        ###\n",
    "        chr_cds_dens, cds_exon_dens, cds_intron_dens = chr_process1(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        #chr_utr5_dens, utr5_exon_dens, utr5_intron_dens = chr_process2(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        chr_utr5_dens = chr_process2(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        #chr_utr3_dens, utr3_exon_dens, utr3_intron_dens = chr_process3(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        chr_utr3_dens = chr_process3(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        chr_exon_dens = chr_process4(chromosomes_df, main_trnscpts_df, cytobands_df, chr_exons_array, chrom)\n",
    "        ###\n",
    "        all_chrs_cdsDens = all_chrs_cdsDens + chr_cds_dens\n",
    "        all_chrs_cdsExonDens = all_chrs_cdsExonDens + cds_exon_dens\n",
    "        all_chrs_cdsIntronDens = all_chrs_cdsIntronDens + cds_intron_dens\n",
    "        all_chrs_utr5Dens = all_chrs_utr5Dens + chr_utr5_dens\n",
    "        #all_chrs_utr5ExonDens = all_chrs_utr5ExonDens + utr5_exon_dens\n",
    "        #all_chrs_utr5IntronDens = all_chrs_utr5IntronDens + utr5_intron_dens\n",
    "        all_chrs_utr3Dens = all_chrs_utr3Dens + chr_utr3_dens\n",
    "        #all_chrs_utr3ExonDens = all_chrs_utr3ExonDens + utr3_exon_dens\n",
    "        #all_chrs_utr3IntronDens = all_chrs_utr3IntronDens + utr3_intron_dens\n",
    "        all_chrs_exon_dens = all_chrs_exon_dens + chr_exon_dens\n",
    "        #print(\"length of UTR5 densities so far: \", len(all_chrs_cdsExonDens))\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"CDS_dens\", all_chrs_cdsDens, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"CDS_exon_dens\", all_chrs_cdsExonDens, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"CDS_intron_dens\", all_chrs_cdsIntronDens, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"5P UTR_dens\", all_chrs_utr5Dens, True)\n",
    "    #cytobands_df.insert(len(cytobands_df.columns), \"5P UTR_exon_dens\", all_chrs_utr5ExonDens, True)\n",
    "    #cytobands_df.insert(len(cytobands_df.columns), \"5P UTR_intron_dens\", all_chrs_utr5IntronDens, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"3P UTR_dens\", all_chrs_utr3Dens, True)\n",
    "    #cytobands_df.insert(len(cytobands_df.columns), \"3P UTR_exon_dens\", all_chrs_utr3ExonDens, True)\n",
    "    #cytobands_df.insert(len(cytobands_df.columns), \"3P UTR_intron_dens\", all_chrs_utr3IntronDens, True)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"exon_dens\", all_chrs_exon_dens, True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ad709e75-109f-4580-bf1a-313301638a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeNewtxt(file_path, df):\n",
    "    \"\"\"Writes a pandas dataframe into txt, columns separated by tabs (\"\\t\")\n",
    "    \"\"\"\n",
    "    f = open(file_path, 'w')\n",
    "    df_cols = \"\"\n",
    "    for col in df.columns[:-1]:\n",
    "        df_cols += str(col) + \"\\t\"\n",
    "    df_cols += df.columns.values[-1]+ \"\\n\"\n",
    "    f.write(str(df_cols))\n",
    "    for row in df.index:\n",
    "        df_row = \"\"\n",
    "        for col in df.columns[:-1]:\n",
    "            df_row += str(df[col][row]) + \"\\t\"\n",
    "        df_row += str(df[df.columns[-1]][row])+ \"\\n\"\n",
    "        f.write(df_row)\n",
    "    f.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9554e4d7-7d8e-436d-b3ec-d651d27d5e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#main_transcripts_write = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/Homo_sapiens_GRCh38_mainTranscripts.gff3\"\n",
    "#writeNewtxt(main_transcripts_write, main_transcripts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "43e0db42-2cd9-452a-885e-094cd0259047",
   "metadata": {},
   "outputs": [],
   "source": [
    "#trscpts_main_DG = \"/home/andre/Documents/trabalho/datasets/Disease_genes/mane/transcrpts_main_DGs.gff3\"\n",
    "#main_transcripts_df = pd.read_csv(main_transcripts_write, sep=\"\\t\", dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "aef9ea4c-c7fc-455a-9a1f-e9f73e24c286",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Genome</th>\n",
       "      <th>Type</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Value</th>\n",
       "      <th>Signal</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Tags</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000426406;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000426406</td>\n",
       "      <td>ENSE00003989331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740.0</td>\n",
       "      <td>451678.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000409316;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000409316</td>\n",
       "      <td>Parent=transcript:ENST00000426406</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000332831;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000332831</td>\n",
       "      <td>ENSE00004001351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716.0</td>\n",
       "      <td>686654.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000329982;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000329982</td>\n",
       "      <td>Parent=transcript:ENST00000332831</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>923923.0</td>\n",
       "      <td>924431.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000616016</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519270</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25622117.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>ENSE00003978110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519271</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25622443.0</td>\n",
       "      <td>25624034.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519272</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624527.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=CDS:ENSP00000302968;Parent=transcript:ENST0...</td>\n",
       "      <td>ID=CDS:ENSP00000302968</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519273</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>25624455.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609;Name=ENSE000...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>ENSE00001666303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519274</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25624528.0</td>\n",
       "      <td>25624902.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Parent=transcript:ENST00000306609</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406711 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr          Genome             Type  First_index  Last_index Value  \\\n",
       "397       1  ensembl_havana             exon     450740.0    451678.0     .   \n",
       "398       1  ensembl_havana              CDS     450740.0    451678.0     .   \n",
       "543       1  ensembl_havana             exon     685716.0    686654.0     .   \n",
       "544       1  ensembl_havana              CDS     685716.0    686654.0     .   \n",
       "1163      1          havana   five_prime_UTR     923923.0    924431.0     .   \n",
       "...      ..             ...              ...          ...         ...   ...   \n",
       "3519270   Y  ensembl_havana             exon   25622117.0  25624034.0     .   \n",
       "3519271   Y  ensembl_havana              CDS   25622443.0  25624034.0     .   \n",
       "3519272   Y  ensembl_havana              CDS   25624455.0  25624527.0     .   \n",
       "3519273   Y  ensembl_havana             exon   25624455.0  25624902.0     .   \n",
       "3519274   Y  ensembl_havana  three_prime_UTR   25624528.0  25624902.0     .   \n",
       "\n",
       "        Signal Unknown                                               Tags  \\\n",
       "397          -       .  Parent=transcript:ENST00000426406;Name=ENSE000...   \n",
       "398          -       0  ID=CDS:ENSP00000409316;Parent=transcript:ENST0...   \n",
       "543          -       .  Parent=transcript:ENST00000332831;Name=ENSE000...   \n",
       "544          -       0  ID=CDS:ENSP00000329982;Parent=transcript:ENST0...   \n",
       "1163         +       .                  Parent=transcript:ENST00000616016   \n",
       "...        ...     ...                                                ...   \n",
       "3519270      +       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "3519271      +       0  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519272      +       1  ID=CDS:ENSP00000302968;Parent=transcript:ENST0...   \n",
       "3519273      +       .  Parent=transcript:ENST00000306609;Name=ENSE000...   \n",
       "3519274      +       .                  Parent=transcript:ENST00000306609   \n",
       "\n",
       "                             ID                             Parent  \\\n",
       "397                         NaN  Parent=transcript:ENST00000426406   \n",
       "398      ID=CDS:ENSP00000409316  Parent=transcript:ENST00000426406   \n",
       "543                         NaN  Parent=transcript:ENST00000332831   \n",
       "544      ID=CDS:ENSP00000329982  Parent=transcript:ENST00000332831   \n",
       "1163                        NaN  Parent=transcript:ENST00000616016   \n",
       "...                         ...                                ...   \n",
       "3519270                     NaN  Parent=transcript:ENST00000306609   \n",
       "3519271  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609   \n",
       "3519272  ID=CDS:ENSP00000302968  Parent=transcript:ENST00000306609   \n",
       "3519273                     NaN  Parent=transcript:ENST00000306609   \n",
       "3519274                     NaN  Parent=transcript:ENST00000306609   \n",
       "\n",
       "                    Name  \n",
       "397      ENSE00003989331  \n",
       "398                  NaN  \n",
       "543      ENSE00004001351  \n",
       "544                  NaN  \n",
       "1163                 NaN  \n",
       "...                  ...  \n",
       "3519270  ENSE00003978110  \n",
       "3519271              NaN  \n",
       "3519272              NaN  \n",
       "3519273  ENSE00001666303  \n",
       "3519274              NaN  \n",
       "\n",
       "[406711 rows x 12 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_dg_trscpts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "efe34810-22a5-4854-8df4-0525670dd149",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['exon', 'CDS', 'five_prime_UTR', 'three_prime_UTR'], dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_dg_trscpts_df[\"Type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "16cc0e8f-e5ab-4698-bbd4-b19fcd4b93a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chr', 'Genome', 'Type', 'First_index', 'Last_index', 'Value', 'Signal',\n",
       "       'Unknown', 'Tags', 'ID', 'Parent', 'Name'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_dg_trscpts_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e10e0386-1af0-48d8-a914-b169a7862502",
   "metadata": {},
   "source": [
    "# Code running"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "34bbbd5e-34a3-4a16-9391-8c6b42f189af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>863 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  \n",
       "0              0.087516   0.006384   0.084476     0.326328  \n",
       "1              0.084077   0.013902   0.079223     0.281556  \n",
       "2              0.056236   0.012909   0.051421     0.275410  \n",
       "3              0.049082   0.013229   0.042217     0.404844  \n",
       "4              0.055881   0.011072   0.048928     0.404047  \n",
       "..                  ...        ...        ...          ...  \n",
       "858            0.036956   0.014211   0.029356     0.015158  \n",
       "859            0.061662   0.015902   0.052790     0.012965  \n",
       "860            0.082234   0.005789   0.078347     0.006441  \n",
       "861            0.026278   0.009164   0.020721     0.001822  \n",
       "862            0.005530   0.001643   0.004251     0.000594  \n",
       "\n",
       "[863 rows x 10 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytoband_read = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr.txt\"\n",
    "cytoband_write = \"/home/andre/Documents/trabalho/cytobands/cytobandFiltered_processed_str_dr_dgsMainTrcpts.txt\"\n",
    "\n",
    "cytobands_df = pd.read_csv(cytoband_read, sep=\"\\t\")#, header=None))\n",
    "#cytobands_df2 = pd.read_csv(cytoband_write, sep=\"\\t\")\n",
    "\n",
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f485d6da-cce1-4ed4-a7b0-6f052acd261d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cytobands_df = cytobands_df.loc[cytobands_df[\"Chromossome\"] != \"chrM\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ceed4aff-3d75-4cd6-88d2-c1c0867899eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing chromosome: chr 1\n",
      "searching cytobands in chromossome:  1\n",
      "processing chromosome: chr 10\n",
      "searching cytobands in chromossome:  10\n",
      "processing chromosome: chr 11\n",
      "searching cytobands in chromossome:  11\n",
      "processing chromosome: chr 12\n",
      "searching cytobands in chromossome:  12\n",
      "processing chromosome: chr 13\n",
      "searching cytobands in chromossome:  13\n",
      "processing chromosome: chr 14\n",
      "searching cytobands in chromossome:  14\n",
      "processing chromosome: chr 15\n",
      "searching cytobands in chromossome:  15\n",
      "processing chromosome: chr 16\n",
      "searching cytobands in chromossome:  16\n",
      "processing chromosome: chr 17\n",
      "searching cytobands in chromossome:  17\n",
      "processing chromosome: chr 18\n",
      "searching cytobands in chromossome:  18\n",
      "processing chromosome: chr 19\n",
      "searching cytobands in chromossome:  19\n",
      "processing chromosome: chr 2\n",
      "searching cytobands in chromossome:  2\n",
      "processing chromosome: chr 20\n",
      "searching cytobands in chromossome:  20\n",
      "processing chromosome: chr 21\n",
      "searching cytobands in chromossome:  21\n",
      "processing chromosome: chr 22\n",
      "searching cytobands in chromossome:  22\n",
      "processing chromosome: chr 3\n",
      "searching cytobands in chromossome:  3\n",
      "processing chromosome: chr 4\n",
      "searching cytobands in chromossome:  4\n",
      "processing chromosome: chr 5\n",
      "searching cytobands in chromossome:  5\n",
      "processing chromosome: chr 6\n",
      "searching cytobands in chromossome:  6\n",
      "processing chromosome: chr 7\n",
      "searching cytobands in chromossome:  7\n",
      "processing chromosome: chr 8\n",
      "searching cytobands in chromossome:  8\n",
      "processing chromosome: chr 9\n",
      "searching cytobands in chromossome:  9\n",
      "processing chromosome: chr X\n",
      "searching cytobands in chromossome:  X\n",
      "processing chromosome: chr Y\n",
      "searching cytobands in chromossome:  Y\n"
     ]
    }
   ],
   "source": [
    "#run_allChr(chromosomes, main_transcripts_df, cytobands_df)\n",
    "run_allChr(chromosomes, main_dg_trscpts_df, cytobands_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f541ebd1-79b4-44ed-a3b8-c5ef71d78640",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDS_dens</th>\n",
       "      <th>CDS_exon_dens</th>\n",
       "      <th>CDS_intron_dens</th>\n",
       "      <th>5P UTR_dens</th>\n",
       "      <th>3P UTR_dens</th>\n",
       "      <th>exon_dens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.320123</td>\n",
       "      <td>0.033313</td>\n",
       "      <td>0.286810</td>\n",
       "      <td>0.003850</td>\n",
       "      <td>0.017207</td>\n",
       "      <td>0.054411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.319537</td>\n",
       "      <td>0.012185</td>\n",
       "      <td>0.307352</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.011664</td>\n",
       "      <td>0.025051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.513454</td>\n",
       "      <td>0.019712</td>\n",
       "      <td>0.493741</td>\n",
       "      <td>0.001799</td>\n",
       "      <td>0.017184</td>\n",
       "      <td>0.038717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.599978</td>\n",
       "      <td>0.011923</td>\n",
       "      <td>0.588055</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.010427</td>\n",
       "      <td>0.023335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.503717</td>\n",
       "      <td>0.028345</td>\n",
       "      <td>0.475372</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.022889</td>\n",
       "      <td>0.051908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.130477</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.126828</td>\n",
       "      <td>0.000374</td>\n",
       "      <td>0.002092</td>\n",
       "      <td>0.006119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002646</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.002702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.074550</td>\n",
       "      <td>0.004770</td>\n",
       "      <td>0.069780</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.002394</td>\n",
       "      <td>0.007705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.045305</td>\n",
       "      <td>0.002629</td>\n",
       "      <td>0.042676</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0.001974</td>\n",
       "      <td>0.005223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDS_dens  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328  0.320123   \n",
       "1              0.084077   0.013902   0.079223     0.281556  0.319537   \n",
       "2              0.056236   0.012909   0.051421     0.275410  0.513454   \n",
       "3              0.049082   0.013229   0.042217     0.404844  0.599978   \n",
       "4              0.055881   0.011072   0.048928     0.404047  0.503717   \n",
       "..                  ...        ...        ...          ...       ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158  0.130477   \n",
       "859            0.061662   0.015902   0.052790     0.012965  0.002646   \n",
       "860            0.082234   0.005789   0.078347     0.006441  0.074550   \n",
       "861            0.026278   0.009164   0.020721     0.001822  0.045305   \n",
       "862            0.005530   0.001643   0.004251     0.000594  0.000000   \n",
       "\n",
       "     CDS_exon_dens  CDS_intron_dens  5P UTR_dens  3P UTR_dens  exon_dens  \n",
       "0         0.033313         0.286810     0.003850     0.017207   0.054411  \n",
       "1         0.012185         0.307352     0.001189     0.011664   0.025051  \n",
       "2         0.019712         0.493741     0.001799     0.017184   0.038717  \n",
       "3         0.011923         0.588055     0.000973     0.010427   0.023335  \n",
       "4         0.028345         0.475372     0.002091     0.022889   0.051908  \n",
       "..             ...              ...          ...          ...        ...  \n",
       "858       0.003649         0.126828     0.000374     0.002092   0.006119  \n",
       "859       0.002263         0.000383     0.000335     0.000100   0.002702  \n",
       "860       0.004770         0.069780     0.000534     0.002394   0.007705  \n",
       "861       0.002629         0.042676     0.000616     0.001974   0.005223  \n",
       "862       0.000000         0.000000     0.000000     0.000000   0.000000  \n",
       "\n",
       "[862 rows x 16 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6267a32e-52da-4d24-935e-c15dc1cf0d7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_intron_dens(cytobands_df):\n",
    "    introns_dens = []\n",
    "    for dens in cytobands_df[\"exon_dens\"]:\n",
    "        intron_dens = 1.000000 - float(dens)\n",
    "        introns_dens.append(intron_dens)\n",
    "    cytobands_df.insert(len(cytobands_df.columns), \"intron_dens\", introns_dens, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "48746dd1-ded2-42ca-a244-df4058533458",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_intron_dens(cytobands_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2fa8138d-0b32-4856-88d4-912d0bd7195e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>CDS_dens</th>\n",
       "      <th>CDS_exon_dens</th>\n",
       "      <th>CDS_intron_dens</th>\n",
       "      <th>5P UTR_dens</th>\n",
       "      <th>3P UTR_dens</th>\n",
       "      <th>exon_dens</th>\n",
       "      <th>intron_dens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>0.320123</td>\n",
       "      <td>0.033313</td>\n",
       "      <td>0.286810</td>\n",
       "      <td>0.003850</td>\n",
       "      <td>0.017207</td>\n",
       "      <td>0.054411</td>\n",
       "      <td>0.945589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>0.319537</td>\n",
       "      <td>0.012185</td>\n",
       "      <td>0.307352</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.011664</td>\n",
       "      <td>0.025051</td>\n",
       "      <td>0.974949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>0.513454</td>\n",
       "      <td>0.019712</td>\n",
       "      <td>0.493741</td>\n",
       "      <td>0.001799</td>\n",
       "      <td>0.017184</td>\n",
       "      <td>0.038717</td>\n",
       "      <td>0.961283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>0.599978</td>\n",
       "      <td>0.011923</td>\n",
       "      <td>0.588055</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.010427</td>\n",
       "      <td>0.023335</td>\n",
       "      <td>0.976665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>0.503717</td>\n",
       "      <td>0.028345</td>\n",
       "      <td>0.475372</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.022889</td>\n",
       "      <td>0.051908</td>\n",
       "      <td>0.948092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.130477</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.126828</td>\n",
       "      <td>0.000374</td>\n",
       "      <td>0.002092</td>\n",
       "      <td>0.006119</td>\n",
       "      <td>0.993881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>0.002646</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.997298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>0.074550</td>\n",
       "      <td>0.004770</td>\n",
       "      <td>0.069780</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.002394</td>\n",
       "      <td>0.007705</td>\n",
       "      <td>0.992295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>0.045305</td>\n",
       "      <td>0.002629</td>\n",
       "      <td>0.042676</td>\n",
       "      <td>0.000616</td>\n",
       "      <td>0.001974</td>\n",
       "      <td>0.005223</td>\n",
       "      <td>0.994777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  CDS_dens  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328  0.320123   \n",
       "1              0.084077   0.013902   0.079223     0.281556  0.319537   \n",
       "2              0.056236   0.012909   0.051421     0.275410  0.513454   \n",
       "3              0.049082   0.013229   0.042217     0.404844  0.599978   \n",
       "4              0.055881   0.011072   0.048928     0.404047  0.503717   \n",
       "..                  ...        ...        ...          ...       ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158  0.130477   \n",
       "859            0.061662   0.015902   0.052790     0.012965  0.002646   \n",
       "860            0.082234   0.005789   0.078347     0.006441  0.074550   \n",
       "861            0.026278   0.009164   0.020721     0.001822  0.045305   \n",
       "862            0.005530   0.001643   0.004251     0.000594  0.000000   \n",
       "\n",
       "     CDS_exon_dens  CDS_intron_dens  5P UTR_dens  3P UTR_dens  exon_dens  \\\n",
       "0         0.033313         0.286810     0.003850     0.017207   0.054411   \n",
       "1         0.012185         0.307352     0.001189     0.011664   0.025051   \n",
       "2         0.019712         0.493741     0.001799     0.017184   0.038717   \n",
       "3         0.011923         0.588055     0.000973     0.010427   0.023335   \n",
       "4         0.028345         0.475372     0.002091     0.022889   0.051908   \n",
       "..             ...              ...          ...          ...        ...   \n",
       "858       0.003649         0.126828     0.000374     0.002092   0.006119   \n",
       "859       0.002263         0.000383     0.000335     0.000100   0.002702   \n",
       "860       0.004770         0.069780     0.000534     0.002394   0.007705   \n",
       "861       0.002629         0.042676     0.000616     0.001974   0.005223   \n",
       "862       0.000000         0.000000     0.000000     0.000000   0.000000   \n",
       "\n",
       "     intron_dens  \n",
       "0       0.945589  \n",
       "1       0.974949  \n",
       "2       0.961283  \n",
       "3       0.976665  \n",
       "4       0.948092  \n",
       "..           ...  \n",
       "858     0.993881  \n",
       "859     0.997298  \n",
       "860     0.992295  \n",
       "861     0.994777  \n",
       "862     1.000000  \n",
       "\n",
       "[862 rows x 17 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "40a0f4f6-b5c1-4f16-9526-783bf429aa72",
   "metadata": {},
   "outputs": [],
   "source": [
    "writeNewtxt(cytoband_write, cytobands_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aec56a34-91b9-4e6e-af9b-2f2826d561ce",
   "metadata": {},
   "source": [
    "### Restultados obtidos considerando o intervalo inteiro desde o primeiro ao ultimo indices de cada tipo de elemento. ERRADO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "991b26b5-15a1-4718-bc62-dcf23daea448",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromossome</th>\n",
       "      <th>First_index</th>\n",
       "      <th>Last_index</th>\n",
       "      <th>Citoband</th>\n",
       "      <th>Unkown</th>\n",
       "      <th>Size</th>\n",
       "      <th>TandemRepeats_dens</th>\n",
       "      <th>STRs_dens</th>\n",
       "      <th>LTRs_dens</th>\n",
       "      <th>DRs_density</th>\n",
       "      <th>...</th>\n",
       "      <th>CDS_dens</th>\n",
       "      <th>CDS_exon_dens</th>\n",
       "      <th>CDS_intron_dens</th>\n",
       "      <th>5P UTR_dens</th>\n",
       "      <th>5P UTR_exon_dens</th>\n",
       "      <th>5P UTR_intron_dens</th>\n",
       "      <th>3P UTR_dens</th>\n",
       "      <th>3P UTR_exon_dens</th>\n",
       "      <th>3P UTR_intron_dens</th>\n",
       "      <th>exon_dens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2299999</td>\n",
       "      <td>p36.33</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2299999</td>\n",
       "      <td>0.087516</td>\n",
       "      <td>0.006384</td>\n",
       "      <td>0.084476</td>\n",
       "      <td>0.326328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.971493</td>\n",
       "      <td>0.971493</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.971557</td>\n",
       "      <td>0.971557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.969561</td>\n",
       "      <td>0.969561</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.971557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2300000</td>\n",
       "      <td>5299999</td>\n",
       "      <td>p36.32</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>2999999</td>\n",
       "      <td>0.084077</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.079223</td>\n",
       "      <td>0.281556</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5300000</td>\n",
       "      <td>7099999</td>\n",
       "      <td>p36.31</td>\n",
       "      <td>gneg</td>\n",
       "      <td>1799999</td>\n",
       "      <td>0.056236</td>\n",
       "      <td>0.012909</td>\n",
       "      <td>0.051421</td>\n",
       "      <td>0.275410</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7100000</td>\n",
       "      <td>9099999</td>\n",
       "      <td>p36.23</td>\n",
       "      <td>gpos25</td>\n",
       "      <td>1999999</td>\n",
       "      <td>0.049082</td>\n",
       "      <td>0.013229</td>\n",
       "      <td>0.042217</td>\n",
       "      <td>0.404844</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9100000</td>\n",
       "      <td>12499999</td>\n",
       "      <td>p36.22</td>\n",
       "      <td>gneg</td>\n",
       "      <td>3399999</td>\n",
       "      <td>0.055881</td>\n",
       "      <td>0.011072</td>\n",
       "      <td>0.048928</td>\n",
       "      <td>0.404047</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>chrY</td>\n",
       "      <td>12400000</td>\n",
       "      <td>17099999</td>\n",
       "      <td>q11.221</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4699999</td>\n",
       "      <td>0.036956</td>\n",
       "      <td>0.014211</td>\n",
       "      <td>0.029356</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>chrY</td>\n",
       "      <td>17100000</td>\n",
       "      <td>19599999</td>\n",
       "      <td>q11.222</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2499999</td>\n",
       "      <td>0.061662</td>\n",
       "      <td>0.015902</td>\n",
       "      <td>0.052790</td>\n",
       "      <td>0.012965</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19600000</td>\n",
       "      <td>23799999</td>\n",
       "      <td>q11.223</td>\n",
       "      <td>gpos50</td>\n",
       "      <td>4199999</td>\n",
       "      <td>0.082234</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.078347</td>\n",
       "      <td>0.006441</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23800000</td>\n",
       "      <td>26599999</td>\n",
       "      <td>q11.23</td>\n",
       "      <td>gneg</td>\n",
       "      <td>2799999</td>\n",
       "      <td>0.026278</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.020721</td>\n",
       "      <td>0.001822</td>\n",
       "      <td>...</td>\n",
       "      <td>0.651617</td>\n",
       "      <td>0.651617</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.650872</td>\n",
       "      <td>0.650872</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.651751</td>\n",
       "      <td>0.651751</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.651751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26600000</td>\n",
       "      <td>57227414</td>\n",
       "      <td>q12</td>\n",
       "      <td>gvar</td>\n",
       "      <td>30627414</td>\n",
       "      <td>0.005530</td>\n",
       "      <td>0.001643</td>\n",
       "      <td>0.004251</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>862 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromossome  First_index  Last_index Citoband  Unkown      Size  \\\n",
       "0          chr1            0     2299999   p36.33    gneg   2299999   \n",
       "1          chr1      2300000     5299999   p36.32  gpos25   2999999   \n",
       "2          chr1      5300000     7099999   p36.31    gneg   1799999   \n",
       "3          chr1      7100000     9099999   p36.23  gpos25   1999999   \n",
       "4          chr1      9100000    12499999   p36.22    gneg   3399999   \n",
       "..          ...          ...         ...      ...     ...       ...   \n",
       "858        chrY     12400000    17099999  q11.221  gpos50   4699999   \n",
       "859        chrY     17100000    19599999  q11.222    gneg   2499999   \n",
       "860        chrY     19600000    23799999  q11.223  gpos50   4199999   \n",
       "861        chrY     23800000    26599999   q11.23    gneg   2799999   \n",
       "862        chrY     26600000    57227414      q12    gvar  30627414   \n",
       "\n",
       "     TandemRepeats_dens  STRs_dens  LTRs_dens  DRs_density  ...  CDS_dens  \\\n",
       "0              0.087516   0.006384   0.084476     0.326328  ...  0.971493   \n",
       "1              0.084077   0.013902   0.079223     0.281556  ...  1.000000   \n",
       "2              0.056236   0.012909   0.051421     0.275410  ...  1.000000   \n",
       "3              0.049082   0.013229   0.042217     0.404844  ...  1.000000   \n",
       "4              0.055881   0.011072   0.048928     0.404047  ...  1.000000   \n",
       "..                  ...        ...        ...          ...  ...       ...   \n",
       "858            0.036956   0.014211   0.029356     0.015158  ...  1.000000   \n",
       "859            0.061662   0.015902   0.052790     0.012965  ...  1.000000   \n",
       "860            0.082234   0.005789   0.078347     0.006441  ...  1.000000   \n",
       "861            0.026278   0.009164   0.020721     0.001822  ...  0.651617   \n",
       "862            0.005530   0.001643   0.004251     0.000594  ...  0.000000   \n",
       "\n",
       "     CDS_exon_dens  CDS_intron_dens  5P UTR_dens  5P UTR_exon_dens  \\\n",
       "0         0.971493              0.0     0.971557          0.971557   \n",
       "1         1.000000              0.0     1.000000          1.000000   \n",
       "2         1.000000              0.0     1.000000          1.000000   \n",
       "3         1.000000              0.0     1.000000          1.000000   \n",
       "4         1.000000              0.0     1.000000          1.000000   \n",
       "..             ...              ...          ...               ...   \n",
       "858       1.000000              0.0     1.000000          1.000000   \n",
       "859       1.000000              0.0     1.000000          1.000000   \n",
       "860       1.000000              0.0     1.000000          1.000000   \n",
       "861       0.651617              0.0     0.650872          0.650872   \n",
       "862       0.000000              0.0     0.000000          0.000000   \n",
       "\n",
       "     5P UTR_intron_dens  3P UTR_dens  3P UTR_exon_dens  3P UTR_intron_dens  \\\n",
       "0                   0.0     0.969561          0.969561                 0.0   \n",
       "1                   0.0     1.000000          1.000000                 0.0   \n",
       "2                   0.0     1.000000          1.000000                 0.0   \n",
       "3                   0.0     1.000000          1.000000                 0.0   \n",
       "4                   0.0     1.000000          1.000000                 0.0   \n",
       "..                  ...          ...               ...                 ...   \n",
       "858                 0.0     1.000000          1.000000                 0.0   \n",
       "859                 0.0     1.000000          1.000000                 0.0   \n",
       "860                 0.0     1.000000          1.000000                 0.0   \n",
       "861                 0.0     0.651751          0.651751                 0.0   \n",
       "862                 0.0     0.000000          0.000000                 0.0   \n",
       "\n",
       "     exon_dens  \n",
       "0     0.971557  \n",
       "1     1.000000  \n",
       "2     1.000000  \n",
       "3     1.000000  \n",
       "4     1.000000  \n",
       "..         ...  \n",
       "858   1.000000  \n",
       "859   1.000000  \n",
       "860   1.000000  \n",
       "861   0.651751  \n",
       "862   0.000000  \n",
       "\n",
       "[862 rows x 23 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytobands_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c8d06c6e-8646-448d-ad60-dc9b6fa931c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_df_1 = main_transcripts_df.loc[(main_transcripts_df[\"Type\"] == \"exon\") & (main_transcripts_df[\"Chr\"] == str(1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f070da9d-0e29-4c43-aeaf-b8520956de5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_exons_array = element_array(chromosomes, exons_df_1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5b55ec5f-1c62-4a8e-a20d-0a2e9a52b602",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "241670035"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.where(chr_exons_array == 0)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "83e060db-9679-4e82-9cf2-a992e800ccea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7286387"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.where(chr_exons_array > 0)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1c61a4cf-b335-43cd-992c-4ad5bf17091e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_df_1 = main_transcripts_df.loc[(main_transcripts_df[\"Type\"] == \"CDS\") & (main_transcripts_df[\"Chr\"] == str(1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "3397eb61-c8c5-4462-bd81-f05e9e1c3573",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_cds_array = element_array(chromosomes, cds_df_1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6b1e05a4-74da-458b-b407-0bdc663d5a91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "245561881"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.where(chr_cds_array == 0)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "2e16fee3-19a0-4ae9-a09a-6d04f9de96de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.013635081082584004"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.where(chr_cds_array > 0)[0])/ len(chr_cds_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e330a720-7c8b-4250-97b0-d7422e026e62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248956422"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr_cds_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f37ad7d1-94ef-40b4-97e6-33ecb39df045",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
